{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "from PyCUB import pyCUB "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "working on session: ensembl2\n",
      "doing all 5818 homologies\n",
      "computing tRNA copy numbers\n",
      "species: _candida_glabrata\n",
      "we got 413 datapoints and managed to extract 345\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'list' object has no attribute 'sum'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-10-2b8a0f21da65>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      3\u001b[0m cub.get_data(From='ensembl', kingdom='compara=fungi', sequence='cdna',\n\u001b[1;32m      4\u001b[0m         \u001b[0madditional\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'type=orthologues'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msaveonfiles\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnormalized\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msetnans\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mFalse\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m         by=\"entropy\", using=\"computejerem\", inpar=20)\n\u001b[0m",
      "\u001b[0;32m/home/ec2-user/PyCUB/CodonUsageBias/PyCUB/pyCUB.pyc\u001b[0m in \u001b[0;36mget_data\u001b[0;34m(self, From, homonames, kingdom, sequence, additional, saveonfiles, normalized, setnans, by, using, inpar)\u001b[0m\n\u001b[1;32m    204\u001b[0m             \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mspece\u001b[0m \u001b[0;32min\u001b[0m \u001b[0menumerate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mspecies\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    205\u001b[0m                 \u001b[0mespece_val\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mspe\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mEspece\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mspece\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtaxonid\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtaxons\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 206\u001b[0;31m                 \u001b[0mespece_val\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_tRNAcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    207\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mspecies\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0mspece\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mespece_val\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    208\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mall_homoset\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloadhashomo\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py\u001b[0m in \u001b[0;36mget_tRNAcopy\u001b[0;34m(self, getentropy, setnans)\u001b[0m\n\u001b[1;32m    204\u001b[0m         \u001b[0;31m# we find probabilities of tRNA\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    205\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0m_\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mv\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mcopynumber\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miteritems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 206\u001b[0;31m             \u001b[0mn\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mv\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    207\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mn\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    208\u001b[0m                 \u001b[0;32mfor\u001b[0m \u001b[0m_\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mval\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mv\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miteritems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'list' object has no attribute 'sum'"
     ]
    }
   ],
   "source": [
    "cub = pyCUB.PyCUB(session = 'ensembl2')\n",
    "cub.get_mymetadata()\n",
    "cub.get_data(From='ensembl', kingdom='compara=fungi', sequence='cdna',\n",
    "        additional='type=orthologues', saveonfiles=False, normalized=True, setnans=False,\n",
    "        by=\"entropy\", using=\"computejerem\", inpar=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "species: exidia_glandulosa_hhb12029\n",
      "we got 300 datapoints and managed to extract 141\n",
      "species: penicillium_roqueforti_fm164\n",
      "we got 179 datapoints and managed to extract 174\n",
      "species: gymnopus_luxurians_fd_317_m1\n",
      "we got 566 datapoints and managed to extract 296\n",
      "species: pachysolen_tannophilus_nrrl_y_2460\n",
      "we got 128 datapoints and managed to extract 1\n",
      "species: daedalea_quercina_l_15889\n",
      "we got 238 datapoints and managed to extract 112\n",
      "species: schizosaccharomyces_cryophilus\n",
      "we got 742 datapoints and managed to extract 461\n",
      "species: galerina_marginata_cbs_339_88\n",
      "we got 292 datapoints and managed to extract 127\n",
      "species: fonsecaea_erecta\n",
      "we got 56 datapoints and managed to extract 26\n",
      "species: fusarium_culmorum\n",
      "we got 283 datapoints and managed to extract 273\n",
      "species: acremonium_chrysogenum_atcc_11550\n",
      "we got 240 datapoints and managed to extract 107\n",
      "species: coccidioides_posadasii_str_silveira\n",
      "we got 233 datapoints and managed to extract 111\n",
      "species: fusarium_verticillioides\n",
      "we got 547 datapoints and managed to extract 521\n",
      "species: dacryopinax_primogenitus\n",
      "we got 114 datapoints and managed to extract 53\n",
      "species: magnaporthe_poae\n",
      "we got 315 datapoints and managed to extract 294\n",
      "species: tuber_melanosporum\n",
      "we got 281 datapoints and managed to extract 261\n",
      "species: erysiphe_necator\n",
      "we got 183 datapoints and managed to extract 172\n",
      "species: kluyveromyces_lactis\n",
      "we got 323 datapoints and managed to extract 152\n",
      "species: wallemia_mellicola_cbs_633_66\n",
      "we got 230 datapoints and managed to extract 109\n",
      "species: zygosaccharomyces_parabailii\n",
      "we got 517 datapoints and managed to extract 1\n",
      "species: escovopsis_weberi\n",
      "we got 213 datapoints and managed to extract 202\n",
      "species: spizellomyces_punctatus_daom_br117\n",
      "we got 411 datapoints and managed to extract 193\n",
      "species: pseudogymnoascus_sp_vkm_f_4514_fw_929_\n",
      "we got 142 datapoints and managed to extract 135\n",
      "species: saccharomyces_sp_boulardii_\n",
      "we got 282 datapoints and managed to extract 266\n",
      "species: metschnikowia_bicuspidata_var_bicuspidata_nrrl_yb_4993\n",
      "we got 252 datapoints and managed to extract 4\n",
      "species: phycomyces_blakesleeanus_nrrl_1555_\n",
      "empty data\n",
      "we got 307 datapoints and managed to extract 0\n",
      "species: exophiala_dermatitidis_nih_ut8656\n",
      "we got 114 datapoints and managed to extract 58\n",
      "species: phaeosphaeria_nodorum\n",
      "we got 176 datapoints and managed to extract 136\n",
      "species: calocera_cornea_hhb12733\n",
      "we got 106 datapoints and managed to extract 48\n",
      "species: lachancea_thermotolerans_cbs_6340\n",
      "we got 456 datapoints and managed to extract 215\n",
      "species: penicillium_subrubescens\n",
      "we got 166 datapoints and managed to extract 9\n",
      "species: nosema_bombycis_cq1\n",
      "we got 158 datapoints and managed to extract 140\n",
      "species: penicillium_arizonense\n",
      "we got 173 datapoints and managed to extract 167\n",
      "species: lichtheimia_corymbifera_jmrc_fsu_9682\n",
      "empty data\n",
      "we got 0 datapoints and managed to extract 0\n",
      "species: dichomitus_squalens_lyad_421_ss1\n",
      "we got 353 datapoints and managed to extract 196\n",
      "species: coniophora_puteana_rwd_64_598_ss2\n",
      "we got 144 datapoints and managed to extract 127\n",
      "species: ustilago_bromivora\n",
      "we got 252 datapoints and managed to extract 114\n",
      "species: penicillium_expansum\n",
      "we got 41 datapoints and managed to extract 41\n",
      "species: exophiala_sideris\n",
      "we got 42 datapoints and managed to extract 19\n",
      "species: cryptococcus_depauperatus_cbs_7841\n",
      "we got 257 datapoints and managed to extract 120\n",
      "species: verruconis_gallopava\n",
      "we got 75 datapoints and managed to extract 34\n",
      "species: tilletia_walkeri\n",
      "we got 80 datapoints and managed to extract 76\n",
      "species: nosema_apis_brl_01\n",
      "we got 53 datapoints and managed to extract 52\n",
      "species: laccaria_amethystina_laam_08_1\n",
      "we got 183 datapoints and managed to extract 172\n",
      "species: grifola_frondosa\n",
      "we got 199 datapoints and managed to extract 183\n",
      "species: penicillium_vulpinum\n",
      "we got 195 datapoints and managed to extract 186\n",
      "species: _candida_glabrata\n",
      "we got 413 datapoints and managed to extract 345\n",
      "species: tulasnella_calospora_mut_4182\n",
      "we got 148 datapoints and managed to extract 69\n",
      "species: aureobasidium_melanogenum_cbs_110374\n",
      "we got 233 datapoints and managed to extract 219\n",
      "species: aspergillus_oryzae\n",
      "empty data\n",
      "we got 285 datapoints and managed to extract 0\n",
      "species: xylona_heveae_tc161\n",
      "empty data\n",
      "we got 144 datapoints and managed to extract 0\n",
      "species: tetrapisispora_phaffii_cbs_4417\n",
      "we got 409 datapoints and managed to extract 193\n",
      "species: ascoidea_rubescens_dsm_1968\n",
      "empty data\n",
      "we got 175 datapoints and managed to extract 0\n",
      "species: zygosaccharomyces_bailii_isa1307\n",
      "we got 1022 datapoints and managed to extract 961\n",
      "species: penicillium_antarcticum\n",
      "we got 180 datapoints and managed to extract 171\n",
      "species: cordyceps_militaris_cm01\n",
      "we got 110 datapoints and managed to extract 104\n",
      "species: nannizzia_gypsea_cbs_118893\n",
      "we got 174 datapoints and managed to extract 85\n",
      "species: aspergillus_lentulus\n",
      "we got 207 datapoints and managed to extract 4\n",
      "species: paracoccidioides_sp_lutzii_pb01\n",
      "we got 200 datapoints and managed to extract 92\n",
      "species: metarhizium_brunneum_arsef_3297\n",
      "we got 143 datapoints and managed to extract 134\n",
      "species: aspergillus_glaucus_cbs_516_65\n",
      "we got 173 datapoints and managed to extract 2\n",
      "species: saccharomyces_kudriavzevii_ifo_1802\n",
      "we got 263 datapoints and managed to extract 250\n",
      "species: coccidioides_immitis_h538_4\n",
      "we got 118 datapoints and managed to extract 5\n",
      "species: pseudogymnoascus_sp_vkm_f_4246\n",
      "we got 117 datapoints and managed to extract 110\n",
      "species: verticillium_alfalfae_vams_102\n",
      "we got 470 datapoints and managed to extract 212\n",
      "species: aureobasidium_namibiae_cbs_147_97\n",
      "we got 233 datapoints and managed to extract 208\n",
      "species: enterocytozoon_bieneusi_h348\n",
      "we got 147 datapoints and managed to extract 71\n",
      "species: saccharomyces_arboricola_h_6\n",
      "we got 362 datapoints and managed to extract 265\n",
      "species: moesziomyces_antarcticus\n",
      "we got 207 datapoints and managed to extract 110\n",
      "species: ophiocordyceps_sinensis_co18\n",
      "we got 87 datapoints and managed to extract 80\n",
      "species: sistotremastrum_niveocremeum_hhb9708\n",
      "we got 155 datapoints and managed to extract 74\n",
      "species: verticillium_dahliae\n",
      "we got 475 datapoints and managed to extract 235\n",
      "species: colletotrichum_sublineola\n",
      "we got 284 datapoints and managed to extract 265\n",
      "species: pseudozyma_hubeiensis_sy62\n",
      "we got 229 datapoints and managed to extract 103\n",
      "species: cyphellophora_europaea_cbs_101466\n",
      "we got 82 datapoints and managed to extract 36\n",
      "species: wickerhamomyces_anomalus_nrrl_y_366_8\n",
      "we got 154 datapoints and managed to extract 1\n",
      "species: colletotrichum_gloeosporioides\n",
      "we got 733 datapoints and managed to extract 665\n",
      "species: babjeviella_inositovora_nrrl_y_12698\n",
      "we got 398 datapoints and managed to extract 5\n",
      "species: nematocida_displodere\n",
      "we got 44 datapoints and managed to extract 42\n",
      "species: kalmanozyma_brasiliensis_ghg001\n",
      "we got 228 datapoints and managed to extract 103\n",
      "species: kluyveromyces_marxianus_dmku3_1042\n",
      "we got 273 datapoints and managed to extract 67\n",
      "species: aspergillus_sydowii_cbs_593_65\n",
      "empty data\n",
      "we got 136 datapoints and managed to extract 0\n",
      "species: pseudocercospora_fijiensis_cirad86\n",
      "we got 85 datapoints and managed to extract 37\n",
      "species: bipolaris_zeicola_26_r_13\n",
      "we got 100 datapoints and managed to extract 95\n",
      "species: pneumocystis_murina_b123\n",
      "we got 93 datapoints and managed to extract 45\n",
      "species: exophiala_aquamarina_cbs_119918\n",
      "we got 36 datapoints and managed to extract 35\n",
      "species: trichosporon_asahii_var_asahii_cbs_8904\n",
      "we got 1101 datapoints and managed to extract 513\n",
      "species: fusarium_mangiferae\n",
      "empty data\n",
      "we got 289 datapoints and managed to extract 0\n",
      "species: penicillium_brasilianum\n",
      "we got 177 datapoints and managed to extract 167\n",
      "species: trichoderma_virens\n",
      "we got 393 datapoints and managed to extract 372\n",
      "species: aspergillus_ustus\n",
      "we got 160 datapoints and managed to extract 154\n",
      "species: lachancea_nothofagi_cbs_11611\n",
      "empty data\n",
      "we got 203 datapoints and managed to extract 0\n",
      "species: pestalotiopsis_fici_w106_1\n",
      "we got 227 datapoints and managed to extract 213\n",
      "species: puccinia_graminis\n",
      "we got 709 datapoints and managed to extract 658\n",
      "species: cylindrobasidium_torrendii_fp15055_ss_10\n",
      "we got 617 datapoints and managed to extract 287\n",
      "species: colletotrichum_orbiculare\n",
      "we got 856 datapoints and managed to extract 724\n",
      "species: schizosaccharomyces_octosporus\n",
      "we got 827 datapoints and managed to extract 517\n",
      "species: cryptococcus_gattii_wm276\n",
      "we got 140 datapoints and managed to extract 132\n",
      "species: cryptococcus_neoformans\n",
      "empty data\n",
      "we got 226 datapoints and managed to extract 0\n",
      "species: pseudocercospora_musae\n",
      "we got 36 datapoints and managed to extract 36\n",
      "species: tolypocladium_ophioglossoides_cbs_100239\n",
      "we got 70 datapoints and managed to extract 70\n",
      "species: neolecta_irregularis_dah_3\n",
      "we got 76 datapoints and managed to extract 73\n",
      "species: aspergillus_tubingensis_cbs_134_48\n",
      "empty data\n",
      "we got 267 datapoints and managed to extract 0\n",
      "species: schizosaccharomyces_pombe\n",
      "we got 378 datapoints and managed to extract 171\n",
      "species: fusarium_graminearum\n",
      "we got 303 datapoints and managed to extract 294\n",
      "species: aspergillus_bombycis\n",
      "we got 241 datapoints and managed to extract 231\n",
      "species: nadsonia_fulvescens_var_elongata_dsm_6958\n",
      "we got 584 datapoints and managed to extract 1\n",
      "species: hydnomerulius_pinastri_md_312\n",
      "we got 191 datapoints and managed to extract 123\n",
      "species: metarhizium_majus_arsef_297\n",
      "we got 138 datapoints and managed to extract 134\n",
      "species: metarhizium_anisopliae_arsef_23\n",
      "we got 134 datapoints and managed to extract 125\n",
      "species: fonsecaea_multimorphosa_cbs_102226\n",
      "we got 58 datapoints and managed to extract 28\n",
      "species: penicillium_oxalicum_114_2\n",
      "we got 180 datapoints and managed to extract 169\n",
      "species: emmonsia_crescens_uamh_3008\n",
      "we got 199 datapoints and managed to extract 94\n",
      "species: chaetomium_thermophilum_var_thermophilum_dsm_1495\n",
      "we got 467 datapoints and managed to extract 217\n",
      "species: penicillium_flavigenum\n",
      "we got 200 datapoints and managed to extract 195\n",
      "species: fonsecaea_nubica\n",
      "we got 58 datapoints and managed to extract 21\n",
      "species: zygosaccharomyces_rouxii\n",
      "we got 543 datapoints and managed to extract 257\n",
      "species: candida_dubliniensis_cd36\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) dat\n",
      "{u'end': u'87762', u'name': u'tRNA-Ser (CAG)', u'start': u'87678', u'strand': u'-1', u'coord_system': {u'version': u'ASM2694v1', u'name': u'chromosome'}, u'seq_region_name': u'5', u'synonyms': [u'CD36_tRNA_0079'], u'biotype': u'tRNA', u'transcripts': [{u'strand': u'-1', u'xrefs': None, u'end': u'87762', u'name': u'tRNA-Ser (CAG)', u'start': u'87678', u'translations': None, u'coord_system': {u'version': u'ASM2694v1', u'name': u'chromosome'}, u'seq_region_name': u'5', u'biotype': u'tRNA', u'id': u'CD36_tRNA_0079-1', u'seq_region_synonyms': [{u'db': u'INSDC', u'id': u'FM992692.1'}], u'description': None}], u'id': u'CD36_tRNA_0079', u'seq_region_synonyms': [{u'db': u'INSDC', u'id': u'FM992692.1'}], u'description': None}\n",
      "(Pdb) continue\n",
      "we got 241 datapoints and managed to extract 218\n",
      "species: ustilago_hordei\n",
      "we got 233 datapoints and managed to extract 217\n",
      "species: parasitella_parasitica\n",
      "we got 141 datapoints and managed to extract 26\n",
      "species: metarhizium_album_arsef_1941\n",
      "we got 117 datapoints and managed to extract 110\n",
      "species: phialocephala_subalpina\n",
      "we got 92 datapoints and managed to extract 2\n",
      "species: cryptococcus_gattii_vgii_r265\n",
      "we got 308 datapoints and managed to extract 143\n",
      "species: penicillium_digitatum_phi26\n",
      "we got 207 datapoints and managed to extract 196\n",
      "species: enterocytozoon_hepatopenaei\n",
      "we got 40 datapoints and managed to extract 38\n",
      "species: penicillium_griseofulvum\n",
      "we got 200 datapoints and managed to extract 195\n",
      "species: mucor_circinelloides_f_circinelloides_1006phl\n",
      "empty data\n",
      "we got 180 datapoints and managed to extract 0\n",
      "species: torulaspora_delbrueckii\n",
      "we got 384 datapoints and managed to extract 181\n",
      "species: peniophora_sp_cont\n",
      "we got 472 datapoints and managed to extract 217\n",
      "species: trichoderma_guizhouense\n",
      "we got 202 datapoints and managed to extract 191\n",
      "species: hanseniaspora_guilliermondii\n",
      "we got 131 datapoints and managed to extract 124\n",
      "species: botrytis_cinerea\n",
      "we got 344 datapoints and managed to extract 128\n",
      "species: fistulina_hepatica_atcc_64428\n",
      "we got 252 datapoints and managed to extract 126\n",
      "species: histoplasma_capsulatum_nam1\n",
      "we got 157 datapoints and managed to extract 76\n",
      "species: nematocida_sp_1_ertm2\n",
      "we got 103 datapoints and managed to extract 49\n",
      "species: tremella_mesenterica_dsm_1558\n",
      "we got 41 datapoints and managed to extract 39\n",
      "species: botryobasidium_botryosum_fd_172_ss1\n",
      "we got 300 datapoints and managed to extract 139\n",
      "species: sclerotinia_sclerotiorum\n",
      "we got 285 datapoints and managed to extract 273\n",
      "species: umbilicaria_pustulata\n",
      "we got 93 datapoints and managed to extract 86\n",
      "species: nematocida_parisii_ertm3\n",
      "we got 103 datapoints and managed to extract 49\n",
      "species: plicaturopsis_crispa_fd_325_ss_3\n",
      "we got 122 datapoints and managed to extract 98\n",
      "species: talaromyces_islandicus\n",
      "we got 188 datapoints and managed to extract 82\n",
      "species: penicillium_freii\n",
      "we got 174 datapoints and managed to extract 167\n",
      "species: pseudogymnoascus_sp_03vt05\n",
      "we got 144 datapoints and managed to extract 5\n",
      "species: _candida_tenuis_atcc_10573\n",
      "we got 112 datapoints and managed to extract 107\n",
      "species: grosmannia_clavigera_kw1407\n",
      "we got 184 datapoints and managed to extract 174\n",
      "species: torrubiella_hemipterigena\n",
      "we got 97 datapoints and managed to extract 91\n",
      "species: schizosaccharomyces_japonicus\n",
      "we got 965 datapoints and managed to extract 566\n",
      "species: emergomyces_pasteuriana_ep9510\n",
      "empty data\n",
      "we got 102 datapoints and managed to extract 0\n",
      "species: debaryomyces_hansenii_cbs767\n",
      "we got 399 datapoints and managed to extract 189\n",
      "species: kwoniella_heveanensis_bcc8398\n",
      "we got 165 datapoints and managed to extract 77\n",
      "species: puccinia_sorghi\n",
      "we got 676 datapoints and managed to extract 223\n",
      "species: blastomyces_percursus\n",
      "empty data\n",
      "we got 88 datapoints and managed to extract 0\n",
      "species: colletotrichum_salicis\n",
      "we got 299 datapoints and managed to extract 278\n",
      "species: sporothrix_schenckii_atcc_58251\n",
      "we got 217 datapoints and managed to extract 101\n",
      "species: hypholoma_sublateritium_fd_334_ss_4\n",
      "we got 447 datapoints and managed to extract 186\n",
      "species: metarhizium_anisopliae_brip_53293\n",
      "we got 129 datapoints and managed to extract 123\n",
      "species: baudoinia_panamericana_uamh_10762\n",
      "we got 68 datapoints and managed to extract 26\n",
      "species: aspergillus_brasiliensis_cbs_101740\n",
      "we got 270 datapoints and managed to extract 2\n",
      "species: phaeoacremonium_minimum_ucrpa7\n",
      "we got 79 datapoints and managed to extract 75\n",
      "species: arthrobotrys_oligospora_atcc_24927\n",
      "we got 118 datapoints and managed to extract 108\n",
      "species: termitomyces_sp_j132\n",
      "we got 332 datapoints and managed to extract 311\n",
      "species: aspergillus_carbonarius_item_5010\n",
      "we got 258 datapoints and managed to extract 3\n",
      "species: fonsecaea_monophora\n",
      "we got 58 datapoints and managed to extract 20\n",
      "species: moniliophthora_roreri_mca_2997\n",
      "we got 247 datapoints and managed to extract 226\n",
      "species: thielaviopsis_punctulata\n",
      "we got 239 datapoints and managed to extract 225\n",
      "species: cyberlindnera_fabianii\n",
      "we got 158 datapoints and managed to extract 147\n",
      "species: aspergillus_ochraceoroseus\n",
      "we got 152 datapoints and managed to extract 143\n",
      "species: candida_orthopsilosis_co_90_125\n",
      "we got 163 datapoints and managed to extract 77\n",
      "species: hirsutella_minnesotensis_3608\n",
      "we got 104 datapoints and managed to extract 97\n",
      "species: aureobasidium_subglaciale_exf_2481\n",
      "we got 196 datapoints and managed to extract 187\n",
      "species: anthracocystis_flocculosa_pf_1\n",
      "we got 172 datapoints and managed to extract 165\n",
      "species: kwoniella_mangroviensis_cbs_10435\n",
      "we got 163 datapoints and managed to extract 75\n",
      "species: tilletia_indica\n",
      "we got 81 datapoints and managed to extract 77\n",
      "species: dactylellina_haptotyla_cbs_200_50\n",
      "we got 120 datapoints and managed to extract 107\n",
      "species: trichophyton_interdigitale_mr816\n",
      "we got 153 datapoints and managed to extract 71\n",
      "species: aspergillus_clavatus\n",
      "empty data\n",
      "we got 234 datapoints and managed to extract 0\n",
      "species: clavispora_lusitaniae_atcc_42720\n",
      "we got 415 datapoints and managed to extract 191\n",
      "species: candida_albicans_sc5314\n",
      "we got 261 datapoints and managed to extract 124\n",
      "species: millerozyma_farinosa_cbs_7064\n",
      "we got 575 datapoints and managed to extract 272\n",
      "species: aspergillus_cristatus\n",
      "we got 160 datapoints and managed to extract 153\n",
      "species: penicillium_polonicum\n",
      "we got 198 datapoints and managed to extract 193\n",
      "species: xanthophyllomyces_dendrorhous\n",
      "we got 89 datapoints and managed to extract 83\n",
      "species: candida_parapsilosis_cdc317\n",
      "empty data\n",
      "we got 58 datapoints and managed to extract 0\n",
      "species: rozella_allomycis_csf55\n",
      "we got 145 datapoints and managed to extract 139\n",
      "species: heterobasidion_irregulare_tc_32_1\n",
      "we got 120 datapoints and managed to extract 114\n",
      "species: pseudogymnoascus_sp_vkm_f_4513_fw_928_\n",
      "we got 115 datapoints and managed to extract 107\n",
      "species: colletotrichum_graminicola\n",
      "we got 306 datapoints and managed to extract 286\n",
      "species: talaromyces_stipitatus_atcc_10500\n",
      "we got 226 datapoints and managed to extract 112\n",
      "species: allomyces_macrogynus_atcc_38327\n",
      "we got 979 datapoints and managed to extract 431\n",
      "species: pseudogymnoascus_sp_vkm_f_4515_fw_2607_\n",
      "we got 122 datapoints and managed to extract 115\n",
      "species: candida_tropicalis_mya_3404\n",
      "we got 384 datapoints and managed to extract 186\n",
      "species: jaapia_argillacea_mucl_33604\n",
      "we got 307 datapoints and managed to extract 141\n",
      "species: aspergillus_ruber_cbs_135680\n",
      "we got 338 datapoints and managed to extract 158\n",
      "species: bipolaris_victoriae_fi3\n",
      "we got 94 datapoints and managed to extract 90\n",
      "species: lachancea_sp_cbs_6924\n",
      "empty data\n",
      "we got 200 datapoints and managed to extract 0\n",
      "species: aschersonia_aleyrodis_rcef_2490\n",
      "we got 90 datapoints and managed to extract 86\n",
      "species: cordyceps_confragosa_rcef_1005\n",
      "we got 114 datapoints and managed to extract 110\n",
      "species: rhizopogon_vesiculosus\n",
      "we got 129 datapoints and managed to extract 119\n",
      "species: magnaporthe_oryzae\n",
      "we got 190 datapoints and managed to extract 178\n",
      "species: gloeophyllum_trabeum_atcc_11539\n",
      "we got 246 datapoints and managed to extract 110\n",
      "species: fusarium_pseudograminearum\n",
      "we got 605 datapoints and managed to extract 589\n",
      "species: colletotrichum_nymphaeae_sa_01\n",
      "we got 325 datapoints and managed to extract 303\n",
      "species: serpula_lacrymans_var_lacrymans_s7_3\n",
      "we got 404 datapoints and managed to extract 397\n",
      "species: coprinopsis_cinerea_okayama7_130\n",
      "we got 546 datapoints and managed to extract 253\n",
      "species: phanerochaete_carnosa_hhb_10118_sp\n",
      "we got 266 datapoints and managed to extract 123\n",
      "species: sordaria_macrospora\n",
      "we got 862 datapoints and managed to extract 372\n",
      "species: pyrenophora_triticirepentis\n",
      "empty data\n",
      "we got 103 datapoints and managed to extract 0\n",
      "species: diaporthe_helianthi\n",
      "we got 36 datapoints and managed to extract 34\n",
      "species: encephalitozoon_romaleae_sj_2008\n",
      "we got 90 datapoints and managed to extract 42\n",
      "species: aspergillus_parasiticus_su_1\n",
      "we got 221 datapoints and managed to extract 214\n",
      "species: kuraishia_capsulata_cbs_1993\n",
      "we got 75 datapoints and managed to extract 70\n",
      "species: uncinocarpus_reesii_1704\n",
      "we got 197 datapoints and managed to extract 96\n",
      "species: ustilago_maydis\n",
      "we got 220 datapoints and managed to extract 104\n",
      "species: stachybotrys_chlorohalonata_ibt_40285\n",
      "we got 105 datapoints and managed to extract 103\n",
      "species: hypsizygus_marmoreus\n",
      "we got 237 datapoints and managed to extract 226\n",
      "species: stereum_hirsutum_fp_91666_ss1\n",
      "we got 339 datapoints and managed to extract 233\n",
      "species: melampsora_laricipopulina\n",
      "we got 439 datapoints and managed to extract 414\n",
      "species: microdochium_bolleyi\n",
      "empty data\n",
      "we got 170 datapoints and managed to extract 0\n",
      "species: kazachstania_naganishii_cbs_8797\n",
      "we got 317 datapoints and managed to extract 151\n",
      "species: hanseniaspora_osmophila\n",
      "we got 191 datapoints and managed to extract 179\n",
      "species: aspergillus_niger_atcc_1015\n",
      "we got 261 datapoints and managed to extract 251\n",
      "species: colletotrichum_fioriniae_pj7\n",
      "we got 302 datapoints and managed to extract 282\n",
      "species: trichoderma_harzianum\n",
      "we got 151 datapoints and managed to extract 144\n",
      "species: nosema_ceranae_brl01\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) dat\n",
      "{u'end': u'21856', u'name': u'trnN', u'start': u'21784', u'strand': u'-1', u'coord_system': {u'version': u'ASM18298v1', u'name': u'supercontig'}, u'seq_region_name': u'Nc000006', u'synonyms': [u'NCER_100675'], u'biotype': u'tRNA', u'transcripts': [{u'strand': u'-1', u'xrefs': None, u'end': u'21856', u'name': u'trnN', u'start': u'21784', u'translations': None, u'coord_system': {u'version': u'ASM18298v1', u'name': u'supercontig'}, u'seq_region_name': u'Nc000006', u'biotype': u'tRNA', u'id': u'NCER_100171-1', u'seq_region_synonyms': [{u'db': u'INSDC', u'id': u'ACOL01000006.1'}], u'description': None}], u'id': u'NCER_100171', u'seq_region_synonyms': [{u'db': u'INSDC', u'id': u'ACOL01000006.1'}], u'description': None}\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) continue\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) dat\n",
      "{u'end': u'598', u'name': u'trnL', u'start': u'515', u'strand': u'1', u'coord_system': {u'version': u'ASM18298v1', u'name': u'supercontig'}, u'seq_region_name': u'Nc001007', u'synonyms': [u'NCER_100039'], u'biotype': u'tRNA', u'transcripts': [{u'strand': u'1', u'xrefs': None, u'end': u'598', u'name': u'trnL', u'start': u'515', u'translations': None, u'coord_system': {u'version': u'ASM18298v1', u'name': u'supercontig'}, u'seq_region_name': u'Nc001007', u'biotype': u'tRNA', u'id': u'NCER_102352-1', u'seq_region_synonyms': [{u'db': u'INSDC', u'id': u'ACOL01001007.1'}], u'description': None}], u'id': u'NCER_102352', u'seq_region_synonyms': [{u'db': u'INSDC', u'id': u'ACOL01001007.1'}], u'description': None}\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) continue\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) dat\n",
      "{u'end': u'3626', u'name': u'trnP', u'start': u'3556', u'strand': u'1', u'coord_system': {u'version': u'ASM18298v1', u'name': u'supercontig'}, u'seq_region_name': u'Nc000017', u'synonyms': [u'NCER_102654'], u'biotype': u'tRNA', u'transcripts': [{u'strand': u'1', u'xrefs': None, u'end': u'3626', u'name': u'trnP', u'start': u'3556', u'translations': None, u'coord_system': {u'version': u'ASM18298v1', u'name': u'supercontig'}, u'seq_region_name': u'Nc000017', u'biotype': u'tRNA', u'id': u'NCER_100350-1', u'seq_region_synonyms': [{u'db': u'INSDC', u'id': u'ACOL01000017.1'}], u'description': None}], u'id': u'NCER_100350', u'seq_region_synonyms': [{u'db': u'INSDC', u'id': u'ACOL01000017.1'}], u'description': None}\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) \n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) \n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) \n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) \n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) \n",
      "(Pdb) \n",
      "(Pdb) \n",
      "(Pdb) \n",
      "(Pdb) \n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) \n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) \n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) \n",
      "(Pdb) \n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) \n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) \n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) \n",
      "(Pdb) \n",
      "(Pdb) \n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) \n",
      "(Pdb) \n",
      "(Pdb) \n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) dat\n",
      "{u'end': u'7655', u'name': u'trnI', u'start': u'7553', u'strand': u'-1', u'coord_system': {u'version': u'ASM18298v1', u'name': u'supercontig'}, u'seq_region_name': u'Nc000109', u'synonyms': [u'NCER_101211'], u'biotype': u'tRNA', u'transcripts': [{u'strand': u'-1', u'xrefs': None, u'end': u'7655', u'name': u'trnI', u'start': u'7553', u'translations': None, u'coord_system': {u'version': u'ASM18298v1', u'name': u'supercontig'}, u'seq_region_name': u'Nc000109', u'biotype': u'tRNA', u'id': u'NCER_101211-1', u'seq_region_synonyms': [{u'db': u'INSDC', u'id': u'ACOL01000109.1'}], u'description': None}], u'id': u'NCER_101211', u'seq_region_synonyms': [{u'db': u'INSDC', u'id': u'ACOL01000109.1'}], u'description': None}\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) \n",
      "(Pdb) c\n",
      "KeyError\n",
      "> /home/ec2-user/PyCUB/CodonUsageBias/PyCUB/espece.py(134)get_tRNAcopy()\n",
      "-> for j, dat in enumerate(data):\n",
      "(Pdb) c\n",
      "we got 127 datapoints and managed to extract 60\n",
      "species: pochonia_chlamydosporia_170\n",
      "we got 112 datapoints and managed to extract 2\n",
      "species: aspergillus_terreus\n",
      "empty data\n",
      "we got 146 datapoints and managed to extract 0\n",
      "species: hanseniaspora_uvarum_dsm_2768\n",
      "we got 233 datapoints and managed to extract 111\n",
      "species: fusarium_langsethiae\n",
      "we got 255 datapoints and managed to extract 248\n",
      "species: aspergillus_calidoustus\n",
      "we got 157 datapoints and managed to extract 155\n",
      "species: wallemia_ichthyophaga_exf_994\n",
      "we got 250 datapoints and managed to extract 119\n",
      "species: gelatoporia_subvermispora_b\n",
      "we got 380 datapoints and managed to extract 179\n",
      "species: ascochyta_rabiei\n",
      "we got 109 datapoints and managed to extract 101\n",
      "species: hebeloma_cylindrosporum_h7\n",
      "we got 279 datapoints and managed to extract 122\n",
      "species: eremothecium_sinecaudum\n",
      "we got 172 datapoints and managed to extract 3\n",
      "species: ceratocystis_platani\n",
      "we got 642 datapoints and managed to extract 351\n",
      "species: pseudogymnoascus_sp_24mn13\n",
      "empty data\n",
      "we got 98 datapoints and managed to extract 0\n",
      "species: mycosphaerella_eumusae\n",
      "we got 69 datapoints and managed to extract 66\n",
      "species: emmonsia_sp_cac_2015a\n",
      "empty data\n",
      "we got 80 datapoints and managed to extract 0\n",
      "species: pseudogymnoascus_verrucosus\n",
      "we got 140 datapoints and managed to extract 5\n",
      "species: pseudogymnoascus_sp_vkm_f_4519_fw_2642_\n",
      "we got 141 datapoints and managed to extract 131\n",
      "species: blastomyces_gilchristii_slh14081\n",
      "we got 245 datapoints and managed to extract 111\n",
      "species: encephalitozoon_cuniculi_gb_m1\n",
      "we got 90 datapoints and managed to extract 42\n",
      "species: coniochaeta_ligniaria_nrrl_30616\n",
      "we got 267 datapoints and managed to extract 12\n",
      "species: rhodotorula_sp_jg_1b\n",
      "we got 220 datapoints and managed to extract 105\n",
      "species: neofusicoccum_parvum_ucrnp2\n",
      "we got 103 datapoints and managed to extract 98\n",
      "species: scedosporium_apiospermum\n",
      "we got 254 datapoints and managed to extract 238\n",
      "species: rhodotorula_graminis_wp1\n",
      "we got 298 datapoints and managed to extract 140\n",
      "species: sugiyamaella_lignohabitans\n",
      "empty data\n",
      "we got 225 datapoints and managed to extract 0\n",
      "species: vanderwaltozyma_polyspora_dsm_70294\n",
      "we got 463 datapoints and managed to extract 220\n",
      "species: pseudogymnoascus_sp_vkm_f_3775\n",
      "we got 131 datapoints and managed to extract 123\n",
      "species: aspergillus_versicolor_cbs_583_65\n",
      "empty data\n",
      "we got 140 datapoints and managed to extract 0\n",
      "species: ogataea_polymorpha\n",
      "empty data\n",
      "we got 79 datapoints and managed to extract 0\n",
      "species: laetiporus_sulphureus_93_53\n",
      "we got 197 datapoints and managed to extract 91\n",
      "species: lodderomyces_elongisporus_nrrl_yb_4239\n",
      "we got 217 datapoints and managed to extract 106\n",
      "species: mixia_osmundae_iam_14324\n",
      "we got 41 datapoints and managed to extract 39\n",
      "species: _candida_intermedia\n",
      "we got 200 datapoints and managed to extract 189\n",
      "species: fusarium_poae\n",
      "we got 292 datapoints and managed to extract 284\n",
      "species: aspergillus_nidulans\n",
      "empty data\n",
      "we got 151 datapoints and managed to extract 0\n",
      "species: trichoderma_reesei\n",
      "we got 385 datapoints and managed to extract 364\n",
      "species: rhizopus_microsporus\n",
      "empty data\n",
      "we got 49 datapoints and managed to extract 0\n",
      "species: aspergillus_luchuensis\n",
      "we got 256 datapoints and managed to extract 246\n",
      "species: lachancea_lanzarotensis\n",
      "we got 398 datapoints and managed to extract 188\n",
      "species: geotrichum_candidum\n",
      "we got 353 datapoints and managed to extract 3\n",
      "species: exophiala_spinifera\n",
      "we got 79 datapoints and managed to extract 34\n",
      "species: pseudogymnoascus_sp_wsf_3629\n",
      "we got 151 datapoints and managed to extract 2\n",
      "species: paxillus_involutus_atcc_200175\n",
      "we got 174 datapoints and managed to extract 111\n",
      "species: trametes_versicolor_fp_101664_ss1\n",
      "we got 231 datapoints and managed to extract 212\n",
      "species: komagataella_phaffii_gs115\n",
      "empty data\n",
      "we got 122 datapoints and managed to extract 0\n",
      "species: hyphopichia_burtonii_nrrl_y_1933\n",
      "we got 175 datapoints and managed to extract 5\n",
      "species: cutaneotrichosporon_oleaginosus\n",
      "we got 548 datapoints and managed to extract 252\n",
      "species: choanephora_cucurbitarum\n",
      "empty data\n",
      "we got 0 datapoints and managed to extract 0\n",
      "species: pyrenophora_teres\n",
      "we got 155 datapoints and managed to extract 139\n",
      "species: arthroderma_otae_cbs_113480\n",
      "we got 168 datapoints and managed to extract 82\n",
      "species: encephalitozoon_hellem_atcc_50504\n",
      "we got 90 datapoints and managed to extract 42\n",
      "species: puccinia_triticina\n",
      "we got 1981 datapoints and managed to extract 1857\n",
      "species: pisolithus_tinctorius_marx_270\n",
      "we got 364 datapoints and managed to extract 172\n",
      "species: colletotrichum_tofieldiae\n",
      "we got 371 datapoints and managed to extract 347\n",
      "species: pseudogymnoascus_sp_vkm_f_4520_fw_2644_\n",
      "we got 128 datapoints and managed to extract 119\n",
      "species: trichophyton_equinum_cbs_127_97\n",
      "empty data\n",
      "we got 81 datapoints and managed to extract 0\n",
      "species: brettanomyces_bruxellensis_awri1499\n",
      "we got 100 datapoints and managed to extract 91\n",
      "species: conidiobolus_coronatus_nrrl_28638\n",
      "we got 1309 datapoints and managed to extract 619\n",
      "species: lentinula_edodes\n",
      "we got 245 datapoints and managed to extract 230\n",
      "species: gaeumannomyces_graminis\n",
      "empty data\n",
      "we got 197 datapoints and managed to extract 0\n",
      "species: piloderma_croceum_f_1598\n",
      "we got 141 datapoints and managed to extract 62\n",
      "species: pyrenochaeta_sp_ds3say3a\n",
      "empty data\n",
      "we got 82 datapoints and managed to extract 0\n",
      "species: nematocida_sp_ertm5\n",
      "we got 50 datapoints and managed to extract 48\n",
      "species: penicillium_decumbens\n",
      "we got 167 datapoints and managed to extract 158\n",
      "species: emmonsia_parva_uamh_139\n",
      "we got 219 datapoints and managed to extract 104\n",
      "species: pseudogymnoascus_sp_vkm_f_3808\n",
      "we got 132 datapoints and managed to extract 125\n",
      "species: aureobasidium_pullulans_exf_150\n",
      "we got 235 datapoints and managed to extract 220\n",
      "species: oidiodendron_maius_zn\n",
      "we got 68 datapoints and managed to extract 26\n",
      "species: paracoccidioides_brasiliensis_pb18\n",
      "we got 202 datapoints and managed to extract 94\n",
      "species: hanseniaspora_opuntiae\n",
      "we got 0 datapoints and managed to extract 1\n",
      "species: rachicladosporium_antarcticum\n",
      "we got 107 datapoints and managed to extract 32\n",
      "species: beauveria_bassiana\n",
      "we got 105 datapoints and managed to extract 100\n",
      "species: rhizopus_delemar_ra_99_880\n",
      "empty data\n",
      "we got 237 datapoints and managed to extract 0\n",
      "species: glarea_lozoyensis_74030\n",
      "we got 140 datapoints and managed to extract 101\n",
      "species: bipolaris_sorokiniana_nd90pr\n",
      "we got 183 datapoints and managed to extract 86\n",
      "species: wickerhamomyces_ciferrii\n",
      "we got 160 datapoints and managed to extract 149\n",
      "species: tetrapisispora_blattae_cbs_6284\n",
      "we got 656 datapoints and managed to extract 311\n",
      "species: trichophyton_rubrum_cbs_118892\n",
      "we got 170 datapoints and managed to extract 84\n",
      "species: tilletia_caries\n",
      "we got 92 datapoints and managed to extract 87\n",
      "species: rhynchosporium_commune\n",
      "we got 113 datapoints and managed to extract 1\n",
      "species: neurospora_crassa\n",
      "empty data\n",
      "we got 426 datapoints and managed to extract 0\n",
      "species: leucoagaricus_sp_symc_cos\n",
      "we got 182 datapoints and managed to extract 167\n",
      "species: pseudogymnoascus_sp_23342_1_i1\n",
      "we got 153 datapoints and managed to extract 3\n",
      "species: mitosporidium_daphniae\n",
      "we got 90 datapoints and managed to extract 85\n",
      "species: penicillium_solitum\n",
      "we got 197 datapoints and managed to extract 193\n",
      "species: neurospora_tetrasperma_fgsc_2509\n",
      "we got 393 datapoints and managed to extract 368\n",
      "species: pseudogymnoascus_sp_vkm_f_103\n",
      "we got 142 datapoints and managed to extract 134\n",
      "species: ordospora_colligata_oc4\n",
      "we got 86 datapoints and managed to extract 40\n",
      "species: sporisorium_scitamineum\n",
      "we got 117 datapoints and managed to extract 111\n",
      "species: pleurotus_ostreatus_pc15\n",
      "we got 317 datapoints and managed to extract 148\n",
      "species: rasamsonia_emersonii_cbs_393_64\n",
      "we got 98 datapoints and managed to extract 93\n",
      "species: talaromyces_cellulolyticus\n",
      "we got 99 datapoints and managed to extract 87\n",
      "species: meyerozyma_guilliermondii_atcc_6260\n",
      "we got 269 datapoints and managed to extract 125\n",
      "species: pseudogymnoascus_sp_vkm_f_4516_fw_969_\n",
      "we got 156 datapoints and managed to extract 147\n",
      "species: rhizophagus_irregularis_daom_181602_gca_000439145\n",
      "empty data\n",
      "we got 86 datapoints and managed to extract 0\n",
      "species: sporothrix_insectorum_rcef_264\n",
      "we got 110 datapoints and managed to extract 102\n",
      "species: neonectria_ditissima\n",
      "we got 171 datapoints and managed to extract 166\n",
      "species: cladophialophora_psammophila_cbs_110553\n",
      "we got 30 datapoints and managed to extract 29\n",
      "species: cladophialophora_carrionii\n",
      "we got 62 datapoints and managed to extract 2\n",
      "species: candida_maltosa_xu316\n",
      "we got 46 datapoints and managed to extract 42\n",
      "species: yarrowia_lipolytica\n",
      "we got 1519 datapoints and managed to extract 948\n",
      "species: komagataella_pastoris\n",
      "we got 242 datapoints and managed to extract 230\n",
      "species: drechmeria_coniospora\n",
      "we got 114 datapoints and managed to extract 1\n",
      "species: capronia_coronata_cbs_617_96\n",
      "we got 38 datapoints and managed to extract 37\n",
      "species: sporisorium_reilianum\n",
      "we got 209 datapoints and managed to extract 198\n",
      "species: metarhizium_acridum_cqma_102\n",
      "we got 232 datapoints and managed to extract 106\n",
      "species: pyronema_omphalodes_cbs_100304\n",
      "we got 662 datapoints and managed to extract 42\n",
      "species: tilletia_controversa\n",
      "we got 93 datapoints and managed to extract 88\n",
      "species: endocarpon_pusillum_z07020\n",
      "we got 47 datapoints and managed to extract 43\n",
      "species: pseudogymnoascus_sp_05ny08\n",
      "we got 152 datapoints and managed to extract 4\n",
      "species: fomitopsis_pinicola_fp_58527_ss1\n",
      "we got 276 datapoints and managed to extract 129\n",
      "species: penicillium_nalgiovense\n",
      "we got 192 datapoints and managed to extract 183\n",
      "species: eutypa_lata_ucrel1\n",
      "we got 125 datapoints and managed to extract 113\n",
      "species: cladophialophora_immunda\n",
      "we got 58 datapoints and managed to extract 27\n",
      "species: serendipita_indica_dsm_11827\n",
      "we got 106 datapoints and managed to extract 93\n",
      "species: penicilliopsis_zonata_cbs_506_65\n",
      "we got 156 datapoints and managed to extract 2\n",
      "species: lipomyces_starkeyi_nrrl_y_11557\n",
      "empty data\n",
      "we got 93 datapoints and managed to extract 0\n",
      "species: aspergillus_wentii_dto_134e9\n",
      "empty data\n",
      "we got 220 datapoints and managed to extract 0\n",
      "species: aspergillus_fumigatus\n",
      "empty data\n",
      "we got 228 datapoints and managed to extract 0\n",
      "species: saitoella_complicata_nrrl_y_17804\n",
      "we got 166 datapoints and managed to extract 158\n",
      "species: _candida_auris\n",
      "we got 223 datapoints and managed to extract 122\n",
      "species: malassezia_sympodialis_atcc_42132\n",
      "we got 75 datapoints and managed to extract 71\n",
      "species: kazachstania_africana_cbs_2517\n",
      "we got 532 datapoints and managed to extract 251\n",
      "species: drechslerella_stenobrocha_248\n",
      "we got 51 datapoints and managed to extract 49\n",
      "species: phaeomoniella_chlamydospora\n",
      "we got 52 datapoints and managed to extract 50\n",
      "species: capronia_epimyces_cbs_606_96\n",
      "we got 40 datapoints and managed to extract 38\n",
      "species: metarhizium_rileyi_rcef_4871\n",
      "we got 108 datapoints and managed to extract 102\n",
      "species: verticillium_longisporum\n",
      "we got 462 datapoints and managed to extract 435\n",
      "species: penicillium_coprophilum\n",
      "we got 206 datapoints and managed to extract 194\n",
      "species: kwoniella_pini_cbs_10737\n",
      "we got 152 datapoints and managed to extract 72\n",
      "species: stemphylium_lycopersici\n",
      "we got 94 datapoints and managed to extract 8\n",
      "species: fibulorhizoctonia_sp_cbs_109695\n",
      "we got 359 datapoints and managed to extract 157\n",
      "species: ophiocordyceps_unilateralis\n",
      "we got 73 datapoints and managed to extract 70\n",
      "species: trichophyton_violaceum\n",
      "we got 80 datapoints and managed to extract 75\n",
      "species: punctularia_strigosozonata_hhb_11173_ss5\n",
      "we got 118 datapoints and managed to extract 98\n",
      "species: scleroderma_citrinum_foug_a\n",
      "we got 197 datapoints and managed to extract 179\n",
      "species: acidomyces_richmondensis_bfw\n",
      "we got 28 datapoints and managed to extract 28\n",
      "species: edhazardia_aedis_usnm_41457\n",
      "we got 119 datapoints and managed to extract 56\n",
      "species: rhynchosporium_secalis\n",
      "we got 89 datapoints and managed to extract 81\n",
      "species: lachancea_mirantina\n",
      "empty data\n",
      "we got 204 datapoints and managed to extract 0\n",
      "species: cladophialophora_yegresii_cbs_114405\n",
      "we got 38 datapoints and managed to extract 33\n",
      "species: candida_arabinofermentans_nrrl_yb_2248\n",
      "empty data\n",
      "we got 118 datapoints and managed to extract 0\n",
      "species: debaryomyces_fabryi\n",
      "we got 104 datapoints and managed to extract 93\n",
      "species: vavraia_culicis_subsp_floridensis\n",
      "we got 99 datapoints and managed to extract 47\n",
      "species: candida_tanzawaensis_nrrl_y_17324\n",
      "empty data\n",
      "we got 256 datapoints and managed to extract 0\n",
      "species: penicillium_italicum\n",
      "we got 184 datapoints and managed to extract 181\n",
      "species: aspergillus_flavus\n",
      "empty data\n",
      "we got 0 datapoints and managed to extract 0\n",
      "species: pisolithus_microcarpus_441\n",
      "we got 249 datapoints and managed to extract 112\n",
      "species: exophiala_oligosperma\n",
      "we got 76 datapoints and managed to extract 34\n",
      "species: malassezia_pachydermatis\n",
      "we got 82 datapoints and managed to extract 77\n",
      "species: metarhizium_guizhouense_arsef_977\n",
      "we got 132 datapoints and managed to extract 126\n",
      "species: neolentinus_lepideus_hhb14362_ss_1\n",
      "we got 278 datapoints and managed to extract 132\n",
      "species: tortispora_caseinolytica_nrrl_y_17796\n",
      "empty data\n",
      "we got 53 datapoints and managed to extract 0\n",
      "species: lichtheimia_ramosa\n",
      "empty data\n",
      "we got 0 datapoints and managed to extract 0\n",
      "species: bipolaris_oryzae_atcc_44560\n",
      "we got 92 datapoints and managed to extract 86\n",
      "species: ascosphaera_apis_arsef_7405\n",
      "we got 225 datapoints and managed to extract 211\n",
      "species: cryptococcus_gattii_ca1873\n",
      "we got 284 datapoints and managed to extract 133\n",
      "species: fibroporia_radiculosa\n",
      "we got 124 datapoints and managed to extract 115\n",
      "species: diplodia_corticola\n",
      "we got 130 datapoints and managed to extract 125\n",
      "species: vittaforma_corneae_atcc_50505\n",
      "we got 93 datapoints and managed to extract 44\n",
      "species: stagonospora_sp_src1lsm3a\n",
      "we got 90 datapoints and managed to extract 1\n",
      "species: cyberlindnera_jadinii\n",
      "we got 154 datapoints and managed to extract 146\n",
      "species: fusarium_oxysporum\n",
      "we got 568 datapoints and managed to extract 549\n",
      "species: tsuchiyaea_wingfieldii_cbs_7118\n",
      "we got 342 datapoints and managed to extract 162\n",
      "species: colletotrichum_simmondsii\n",
      "we got 318 datapoints and managed to extract 298\n",
      "species: trichoderma_gamsii\n",
      "we got 215 datapoints and managed to extract 202\n",
      "species: phlebiopsis_gigantea_11061_1_cr5_6\n",
      "we got 314 datapoints and managed to extract 152\n",
      "species: moesziomyces_aphidis_dsm_70725\n",
      "we got 24 datapoints and managed to extract 23\n",
      "species: spathaspora_passalidarum_nrrl_y_27907\n",
      "we got 165 datapoints and managed to extract 157\n",
      "species: gonapodya_prolifera_jel478\n",
      "we got 189 datapoints and managed to extract 101\n",
      "species: saccharomyces_eubayanus\n",
      "empty data\n",
      "we got 304 datapoints and managed to extract 0\n",
      "species: phialocephala_scopiformis\n",
      "empty data\n",
      "we got 81 datapoints and managed to extract 0\n",
      "species: penicillium_steckii\n",
      "we got 159 datapoints and managed to extract 155\n",
      "species: talaromyces_marneffei_atcc_18224\n",
      "we got 244 datapoints and managed to extract 126\n",
      "species: fonsecaea_pedrosoi_cbs_271_37\n",
      "we got 53 datapoints and managed to extract 24\n",
      "species: macrophomina_phaseolina_ms6\n",
      "we got 288 datapoints and managed to extract 136\n",
      "species: calocera_viscosa_tufc12733\n",
      "we got 120 datapoints and managed to extract 57\n",
      "species: pseudogymnoascus_sp_vkm_f_4281_fw_2241_\n",
      "we got 112 datapoints and managed to extract 105\n",
      "species: cordyceps_brongniartii_rcef_3172\n",
      "we got 123 datapoints and managed to extract 113\n",
      "species: suillus_luteus_uh_slu_lm8_n1\n",
      "we got 231 datapoints and managed to extract 108\n",
      "species: pseudogymnoascus_destructans_20631_21\n",
      "we got 245 datapoints and managed to extract 115\n",
      "species: thermothelomyces_thermophila_atcc_42464\n",
      "we got 383 datapoints and managed to extract 180\n",
      "species: schizopora_paradoxa\n",
      "we got 160 datapoints and managed to extract 75\n",
      "species: agaricus_bisporus_var_burnettii_jb137_s8\n",
      "we got 385 datapoints and managed to extract 193\n",
      "species: penicillium_chrysogenum\n",
      "empty data\n",
      "we got 213 datapoints and managed to extract 0\n",
      "species: pseudoloma_neurophilia\n",
      "we got 29 datapoints and managed to extract 29\n",
      "species: naumovozyma_castellii_cbs_4309\n",
      "we got 539 datapoints and managed to extract 254\n",
      "species: enterospora_canceri\n",
      "we got 44 datapoints and managed to extract 42\n",
      "species: fomitiporia_mediterranea_mf3_22\n",
      "we got 82 datapoints and managed to extract 77\n",
      "species: sclerotinia_borealis_f_4128\n",
      "we got 358 datapoints and managed to extract 167\n",
      "species: saccharomycetaceae_sp_ashbya_aceri_\n",
      "we got 401 datapoints and managed to extract 186\n",
      "species: kwoniella_bestiolae_cbs_10118\n",
      "we got 152 datapoints and managed to extract 71\n",
      "species: colletotrichum_higginsianum\n",
      "we got 778 datapoints and managed to extract 395\n",
      "species: schizophyllum_commune_h4_8\n",
      "we got 275 datapoints and managed to extract 259\n",
      "species: amanita_muscaria_koide_bx008\n",
      "we got 101 datapoints and managed to extract 94\n",
      "species: penicillium_camemberti_fm_013\n",
      "we got 182 datapoints and managed to extract 178\n",
      "species: valsa_mali_var_pyri\n",
      "we got 154 datapoints and managed to extract 146\n",
      "species: puccinia_striiformis\n",
      "we got 284 datapoints and managed to extract 256\n",
      "species: kwoniella_dejecticola_cbs_10117\n",
      "we got 151 datapoints and managed to extract 71\n",
      "species: cladophialophora_bantiana_cbs_173_52\n",
      "we got 60 datapoints and managed to extract 26\n",
      "species: lachancea_dasiensis_cbs_10888\n",
      "empty data\n",
      "we got 203 datapoints and managed to extract 0\n",
      "species: saccharomyces_cerevisiae_x_saccharomyces_kudriavzevii_vin7\n",
      "we got 547 datapoints and managed to extract 512\n",
      "species: penicillium_nordicum\n",
      "we got 165 datapoints and managed to extract 158\n",
      "species: sistotremastrum_suecicum_hhb10207_ss_3\n",
      "we got 155 datapoints and managed to extract 74\n",
      "species: blastomyces_dermatitidis_er_3\n",
      "we got 216 datapoints and managed to extract 102\n",
      "species: rhizoctonia_solani_ag_1_ib\n",
      "we got 316 datapoints and managed to extract 150\n",
      "species: tilletiaria_anomala_ubc_951\n",
      "we got 124 datapoints and managed to extract 57\n",
      "species: diaporthe_ampelina\n",
      "we got 120 datapoints and managed to extract 116\n",
      "species: pichia_membranifaciens_nrrl_y_2026\n",
      "empty data\n",
      "we got 158 datapoints and managed to extract 0\n",
      "species: eremothecium_cymbalariae_dbvpg_7215\n",
      "we got 289 datapoints and managed to extract 136\n",
      "species: dothistroma_septosporum\n",
      "we got 84 datapoints and managed to extract 45\n",
      "species: aspergillus_aculeatus_atcc_16872\n",
      "empty data\n",
      "we got 320 datapoints and managed to extract 0\n",
      "species: eremothecium_gossypii_fdag1\n",
      "we got 425 datapoints and managed to extract 193\n",
      "species: amphiamblys_sp_wsbs2006\n",
      "we got 93 datapoints and managed to extract 43\n",
      "species: madurella_mycetomatis\n",
      "we got 147 datapoints and managed to extract 137\n",
      "species: paxillus_rubicundulus_ve08_2h10\n",
      "we got 548 datapoints and managed to extract 255\n",
      "species: colletotrichum_orchidophilum\n",
      "we got 335 datapoints and managed to extract 313\n",
      "species: fusarium_solani\n",
      "we got 520 datapoints and managed to extract 495\n",
      "species: sphaerobolus_stellatus_ss14\n",
      "we got 1289 datapoints and managed to extract 653\n",
      "species: cryptococcus_amylolentus_cbs_6039\n",
      "we got 393 datapoints and managed to extract 185\n",
      "species: exophiala_xenobiotica\n",
      "we got 74 datapoints and managed to extract 35\n",
      "species: purpureocillium_lilacinum\n",
      "we got 88 datapoints and managed to extract 2\n",
      "species: cryptococcus_gattii_vgiv_ind107\n",
      "we got 282 datapoints and managed to extract 132\n",
      "species: colletotrichum_incanum\n",
      "we got 355 datapoints and managed to extract 332\n",
      "species: ophiostoma_piceae_uamh_11346\n",
      "we got 243 datapoints and managed to extract 230\n",
      "species: alternaria_alternata\n",
      "we got 106 datapoints and managed to extract 1\n",
      "species: aspergillus_udagawae\n",
      "empty data\n",
      "we got 179 datapoints and managed to extract 0\n",
      "species: fusarium_proliferatum_et1\n",
      "empty data\n",
      "we got 310 datapoints and managed to extract 0\n",
      "species: microbotryum_violaceum\n",
      "we got 226 datapoints and managed to extract 168\n",
      "species: moniliophthora_perniciosa_fa553\n",
      "we got 214 datapoints and managed to extract 96\n",
      "species: aspergillus_rambellii\n",
      "we got 152 datapoints and managed to extract 143\n",
      "species: anncaliia_algerae_pra339\n",
      "we got 108 datapoints and managed to extract 52\n",
      "species: serendipita_vermifera_maff_305830\n",
      "we got 130 datapoints and managed to extract 54\n",
      "species: naumovozyma_dairenensis_cbs_421\n",
      "we got 527 datapoints and managed to extract 250\n",
      "species: pneumocystis_carinii_b80\n",
      "empty data\n",
      "we got 44 datapoints and managed to extract 0\n",
      "species: mortierella_elongata_ag_77\n",
      "empty data\n",
      "we got 249 datapoints and managed to extract 0\n",
      "species: phlebia_centrifuga\n",
      "we got 387 datapoints and managed to extract 180\n",
      "species: trachipleistophora_hominis\n",
      "we got 84 datapoints and managed to extract 42\n",
      "species: spraguea_lophii_42_110\n",
      "we got 106 datapoints and managed to extract 51\n",
      "species: pichia_kudriavzevii\n",
      "we got 258 datapoints and managed to extract 242\n",
      "species: mucor_ambiguus\n",
      "empty data\n",
      "we got 169 datapoints and managed to extract 0\n",
      "species: trichoderma_atroviride_imi_206040\n",
      "we got 215 datapoints and managed to extract 204\n",
      "species: pneumocystis_jirovecii_ru7\n",
      "empty data\n",
      "we got 45 datapoints and managed to extract 0\n",
      "species: rhinocladiella_mackenziei_cbs_650_93\n",
      "we got 46 datapoints and managed to extract 17\n",
      "species: blumeria_graminis\n",
      "we got 461 datapoints and managed to extract 449\n",
      "species: ogataea_parapolymorpha_dl_1\n",
      "we got 159 datapoints and managed to extract 74\n",
      "species: laccaria_bicolor_s238n_h82\n",
      "we got 225 datapoints and managed to extract 213\n",
      "species: paraphaeosphaeria_sporulosa\n",
      "we got 107 datapoints and managed to extract 4\n",
      "species: bipolaris_maydis_c5\n",
      "we got 229 datapoints and managed to extract 98\n",
      "species: phialophora_americana\n",
      "we got 59 datapoints and managed to extract 23\n",
      "species: zymoseptoria_brevis\n",
      "we got 40 datapoints and managed to extract 40\n",
      "species: rhynchosporium_agropyri\n",
      "we got 107 datapoints and managed to extract 1\n",
      "species: colletotrichum_chlorophyti\n",
      "we got 360 datapoints and managed to extract 339\n",
      "species: hepatospora_eriocheir\n",
      "we got 57 datapoints and managed to extract 54\n",
      "species: podospora_anserina_s_mat_\n",
      "we got 705 datapoints and managed to extract 324\n",
      "species: postia_placenta_mad_698_r\n",
      "we got 263 datapoints and managed to extract 237\n",
      "species: aspergillus_nomius_nrrl_13137\n",
      "we got 236 datapoints and managed to extract 228\n",
      "species: claviceps_purpurea_20_1\n",
      "we got 219 datapoints and managed to extract 206\n",
      "species: pseudogymnoascus_sp_vkm_f_3557\n",
      "we got 141 datapoints and managed to extract 134\n",
      "species: trametes_pubescens\n",
      "we got 566 datapoints and managed to extract 272\n",
      "species: encephalitozoon_intestinalis_atcc_50506\n",
      "we got 90 datapoints and managed to extract 42\n",
      "species: trichophyton_tonsurans_cbs_112818\n",
      "we got 168 datapoints and managed to extract 82\n",
      "species: phialophora_attae\n",
      "we got 47 datapoints and managed to extract 44\n",
      "species: batrachochytrium_salamandrivorans\n"
     ]
    }
   ],
   "source": [
    "cub.save('entropy_cerevisiae')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'cub' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-b72417153b4f>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mcub\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'cub' is not defined"
     ]
    }
   ],
   "source": [
    "cub\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ALA': {'CGA': 4, 'CGC': 1, 'CGG': 0, 'CGU': 2},\n",
       " 'ARG': {'GCA': 1, 'GCC': 1, 'GCG': 0, 'GCU': 0, 'UCC': 1, 'UCU': 5},\n",
       " 'ASN': {'UUA': 0, 'UUG': 5},\n",
       " 'ASP': {'CUA': 0, 'CUG': 5},\n",
       " 'CYS': {'ACA': 0, 'ACG': 3},\n",
       " 'GLN': {'GUC': 1, 'GUU': 3},\n",
       " 'GLU': {'CUC': 1, 'CUU': 5},\n",
       " 'GLY': {'CCA': 0, 'CCC': 0, 'CCG': 5, 'CCU': 3},\n",
       " 'HIS': {'GUA': 0, 'GUG': 3},\n",
       " 'ILE': {'UAA': 4, 'UAG': 0, 'UAU': 2},\n",
       " 'LEU': {'AAC': 5, 'AAU': 2, 'GAA': 1, 'GAC': 1, 'GAG': 0, 'GAU': 0},\n",
       " 'LYS': {'UUC': 4, 'UUU': 3},\n",
       " 'PHE': {'AAA': 0, 'AAG': 5},\n",
       " 'PRO': {'GGA': 1, 'GGC': 0, 'GGG': 0, 'GGU': 3},\n",
       " 'SER': {'AGA': 4, 'AGC': 1, 'AGG': 0, 'AGU': 2, 'UCA': 0, 'UCG': 3},\n",
       " 'THR': {'UGA': 3, 'UGC': 1, 'UGG': 0, 'UGU': 2},\n",
       " 'TYR': {'AUA': 0, 'AUG': 4},\n",
       " 'VAL': {'CAA': 4, 'CAC': 2, 'CAG': 0, 'CAU': 1}}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cub.species[\"_candida_tenuis_atcc_10573\"].copynumbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 356,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "working on session: ensembl2\n",
      "you already have a session here (just a warning)\n",
      "doing all 1 homologies\n",
      "homology: YPR004C : 0\n",
      "> /Users/jeremie/Documents/Projects/MasterProj/PyCUB/utils.py(720)computepartition_normal_approximation()\n",
      "-> proba = np.ones(nbcod) / nbcod if probavector is None else probavector\n",
      "(Pdb) q\n"
     ]
    },
    {
     "ename": "BdbQuit",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mBdbQuit\u001b[0m                                   Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-356-1005a6cf209d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      2\u001b[0m cub.get_data(From='ensembl', homonames=['YPR004C'], kingdom='compara=fungi', sequence='cdna',\n\u001b[1;32m      3\u001b[0m         \u001b[0madditional\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'type=orthologues'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msaveonfiles\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnormalized\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msetnans\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mFalse\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m         by=\"entropyLocation\", inpar=False)\n\u001b[0m",
      "\u001b[0;32m/Users/jeremie/Documents/Projects/MasterProj/PyCUB/pyCUB.pyc\u001b[0m in \u001b[0;36mget_data\u001b[0;34m(self, From, homonames, kingdom, sequence, additional, saveonfiles, normalized, setnans, by, using, inpar)\u001b[0m\n\u001b[1;32m    185\u001b[0m                                              utils.loadfromensembl(name, kingdom, sequence,\n\u001b[1;32m    186\u001b[0m                                                                    \u001b[0madditional\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msaveonfiles\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 187\u001b[0;31m                                                                    normalized, setnans, i, by, using)})\n\u001b[0m\u001b[1;32m    188\u001b[0m             \u001b[0;31m# TODO: test full pipeline with frequency/entropy/entropylocation\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    189\u001b[0m             \u001b[0mtaxons\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mspecies\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mall_homoset\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpreprocessing\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mwithtaxons\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/jeremie/Documents/Projects/MasterProj/PyCUB/utils.py\u001b[0m in \u001b[0;36mloadfromensembl\u001b[0;34m(homology, kingdom, sequence, additional, saveonfiles, normalized, setnans, number, by, using)\u001b[0m\n\u001b[1;32m    236\u001b[0m         data, normalized=normalized, setnans=setnans, by=by)\n\u001b[1;32m    237\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mby\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'entropyLocation'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 238\u001b[0;31m         \u001b[0mH\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetloc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mH\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlenmat\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0musing\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0musing\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    239\u001b[0m     \u001b[0;31m# here we add two things into names but only as a temporary saving measures removed by the\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    240\u001b[0m     \u001b[0;31m# application fo preprocessing in homoset.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/jeremie/Documents/Projects/MasterProj/PyCUB/utils.py\u001b[0m in \u001b[0;36mgetloc\u001b[0;34m(valH, geneleng, using)\u001b[0m\n\u001b[1;32m    383\u001b[0m                     \u001b[0mvalHloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    384\u001b[0m                     \u001b[0;32mcontinue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 385\u001b[0;31m                 \u001b[0mref\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcomputepartition\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnbcod\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mleng\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m1000000\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0musing\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0musing\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    386\u001b[0m                 \u001b[0mref\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdivide\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlog\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdivide\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mEg\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mref\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mleng\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    387\u001b[0m                 \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/jeremie/Documents/Projects/MasterProj/PyCUB/utils.py\u001b[0m in \u001b[0;36mcomputepartition\u001b[0;34m(nbcod, leng, MAXITR, using)\u001b[0m\n\u001b[1;32m    441\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mcomputepartition_sorted_full\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnbcod\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mleng\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    442\u001b[0m     \u001b[0;32melif\u001b[0m \u001b[0musing\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'normal'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 443\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mcomputepartition_normal_approximation\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnbcod\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mleng\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    444\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    445\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mNameError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\" give a Using from [permutation, random, normal, full]\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/jeremie/Documents/Projects/MasterProj/PyCUB/utils.py\u001b[0m in \u001b[0;36mcomputepartition_normal_approximation\u001b[0;34m(nbcod, leng, probavector)\u001b[0m\n\u001b[1;32m    718\u001b[0m     \u001b[0;31m# TODO: debug the singular matrix problem\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    719\u001b[0m     \u001b[0mpdb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_trace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 720\u001b[0;31m     \u001b[0mproba\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mones\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnbcod\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m/\u001b[0m \u001b[0mnbcod\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mprobavector\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0mNone\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0mprobavector\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    721\u001b[0m     \u001b[0mmean\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mproba\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0mleng\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    722\u001b[0m     \u001b[0mker\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkron\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mproba\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mproba\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreshape\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnbcod\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnbcod\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0mleng\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/jeremie/Documents/Projects/MasterProj/PyCUB/utils.py\u001b[0m in \u001b[0;36mcomputepartition_normal_approximation\u001b[0;34m(nbcod, leng, probavector)\u001b[0m\n\u001b[1;32m    718\u001b[0m     \u001b[0;31m# TODO: debug the singular matrix problem\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    719\u001b[0m     \u001b[0mpdb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_trace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 720\u001b[0;31m     \u001b[0mproba\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mones\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnbcod\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m/\u001b[0m \u001b[0mnbcod\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mprobavector\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0mNone\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0mprobavector\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    721\u001b[0m     \u001b[0mmean\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mproba\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0mleng\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    722\u001b[0m     \u001b[0mker\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkron\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mproba\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mproba\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreshape\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnbcod\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnbcod\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0mleng\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/jeremie/anaconda2/envs/tensorflow1/lib/python2.7/bdb.pyc\u001b[0m in \u001b[0;36mtrace_dispatch\u001b[0;34m(self, frame, event, arg)\u001b[0m\n\u001b[1;32m     47\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0;31m# None\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     48\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mevent\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'line'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 49\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdispatch_line\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mframe\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     50\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mevent\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'call'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     51\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdispatch_call\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mframe\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0marg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/jeremie/anaconda2/envs/tensorflow1/lib/python2.7/bdb.pyc\u001b[0m in \u001b[0;36mdispatch_line\u001b[0;34m(self, frame)\u001b[0m\n\u001b[1;32m     66\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstop_here\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mframe\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbreak_here\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mframe\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     67\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0muser_line\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mframe\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 68\u001b[0;31m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mquitting\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;32mraise\u001b[0m \u001b[0mBdbQuit\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     69\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtrace_dispatch\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     70\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mBdbQuit\u001b[0m: "
     ]
    }
   ],
   "source": [
    "cub = pyCUB.PyCUB(session = 'ensembl2')\n",
    "cub.get_data(From='ensembl', homonames=['YPR004C'], kingdom='compara=fungi', sequence='cdna',\n",
    "        additional='type=orthologues', saveonfiles=False, normalized=True, setnans=False,\n",
    "        by=\"entropyLocation\", inpar=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "Image data can not convert to float",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-19-a8e17693a964>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mget_ipython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmagic\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mu'matplotlib inline'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mcub\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mall_homoset\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mplot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msize\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m100\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/Users/jeremie/Documents/Projects/MasterProj/PyCUB/homoset.pyc\u001b[0m in \u001b[0;36mplot\u001b[0;34m(self, invert, size, interactive)\u001b[0m\n\u001b[1;32m    631\u001b[0m         \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfigure\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfigsize\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msize\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m200\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    632\u001b[0m         \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtitle\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'the regular matrix'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 633\u001b[0;31m         \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mimshow\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhashomo_matrix\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mT\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0minvert\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhashomo_matrix\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    634\u001b[0m         \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msavefig\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"utils/lastplot.pdf\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    635\u001b[0m         \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshow\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python2.7/site-packages/matplotlib/pyplot.pyc\u001b[0m in \u001b[0;36mimshow\u001b[0;34m(X, cmap, norm, aspect, interpolation, alpha, vmin, vmax, origin, extent, shape, filternorm, filterrad, imlim, resample, url, hold, data, **kwargs)\u001b[0m\n\u001b[1;32m   3155\u001b[0m                         \u001b[0mfilternorm\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mfilternorm\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfilterrad\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mfilterrad\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3156\u001b[0m                         \u001b[0mimlim\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mimlim\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mresample\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mresample\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0murl\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3157\u001b[0;31m                         **kwargs)\n\u001b[0m\u001b[1;32m   3158\u001b[0m     \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3159\u001b[0m         \u001b[0max\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_hold\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mwashold\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python2.7/site-packages/matplotlib/__init__.pyc\u001b[0m in \u001b[0;36minner\u001b[0;34m(ax, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1896\u001b[0m                     warnings.warn(msg % (label_namer, func.__name__),\n\u001b[1;32m   1897\u001b[0m                                   RuntimeWarning, stacklevel=2)\n\u001b[0;32m-> 1898\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0max\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1899\u001b[0m         \u001b[0mpre_doc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0minner\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__doc__\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1900\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mpre_doc\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python2.7/site-packages/matplotlib/axes/_axes.pyc\u001b[0m in \u001b[0;36mimshow\u001b[0;34m(self, X, cmap, norm, aspect, interpolation, alpha, vmin, vmax, origin, extent, shape, filternorm, filterrad, imlim, resample, url, **kwargs)\u001b[0m\n\u001b[1;32m   5122\u001b[0m                               resample=resample, **kwargs)\n\u001b[1;32m   5123\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5124\u001b[0;31m         \u001b[0mim\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_data\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5125\u001b[0m         \u001b[0mim\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_alpha\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0malpha\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5126\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mim\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_clip_path\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python2.7/site-packages/matplotlib/image.pyc\u001b[0m in \u001b[0;36mset_data\u001b[0;34m(self, A)\u001b[0m\n\u001b[1;32m    594\u001b[0m         if (self._A.dtype != np.uint8 and\n\u001b[1;32m    595\u001b[0m                 not np.can_cast(self._A.dtype, np.float)):\n\u001b[0;32m--> 596\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Image data can not convert to float\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    597\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    598\u001b[0m         if (self._A.ndim not in (2, 3) or\n",
      "\u001b[0;31mTypeError\u001b[0m: Image data can not convert to float"
     ]
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "cub.all_homoset.plot(size=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cubias.all_homoset.plot_speciesperhomo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cubias.all_homoset.order_from_matrix(clustering='kmodes', homogroupnb=4, \n",
    "        plot_ordering=True, findnb=False, reducedim=False, byspecie=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cubias.all_homoset.order_from_matrix(clustering='kmodes', homogroupnb=4, \n",
    "        plot_ordering=True, findnb=False, reducedim=False, byspecie=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cubias.all_homoset.order_from_matrix(clustering='kmodes', homogroupnb=4, \n",
    "        plot_ordering=True, findnb=False, reducedim=False, byspecie=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cubias.all_homoset[\"YFL039C\"].nb_unique_species()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cubias.get_working_homoset(per_cluster=True, clusternb=1, species=None, homologies=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cubias.working_homoset.find_clusters(clustering='dbscan',homogroupnb=2, assess= True,eps=0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cubias.working_homoset[300].plot(reducer = 'pca', interactive=True,D=2,size=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%output size=150\n",
    "cubias.all_homoset.plot_all('PCA',interactive = True)\n",
    "# it is super computationally expensive so below you can find a version with only 4000 genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cubias.save(\"ensembl-processed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cubias.working_homoset.get_clusterstats(sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from PyCUB import utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 324,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100 loops, best of 3: 2.62 ms per loop\n"
     ]
    }
   ],
   "source": [
    "%timeit a = utils.computepartition_sorted_full(3,50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 335,
   "metadata": {},
   "outputs": [],
   "source": [
    "b = utils.computepartition_without_permutation(2,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 355,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1.08420217e-19, 6.83047369e-18, 2.11744684e-16, 4.30547525e-15,\n",
       "       6.45821287e-14, 7.62069119e-13, 7.36666815e-12, 5.99857263e-11,\n",
       "       4.19900084e-10, 2.56605607e-09, 1.38567028e-08, 6.67641134e-08,\n",
       "       2.89311158e-07, 1.13498993e-06, 4.05353546e-06, 1.32415492e-05,\n",
       "       3.97246475e-05, 1.09826967e-04, 2.80668915e-04, 6.64742166e-04,\n",
       "       1.46243277e-03, 2.99450519e-03, 5.71678263e-03, 1.01907864e-02,\n",
       "       1.69846440e-02, 2.64960447e-02, 3.87249884e-02, 5.30675767e-02,\n",
       "       6.82297415e-02, 8.23462397e-02, 9.33257384e-02, 9.93467537e-02,\n",
       "       9.93467537e-02, 9.33257384e-02, 8.23462397e-02, 6.82297415e-02,\n",
       "       5.30675767e-02, 3.87249884e-02, 2.64960447e-02, 1.69846440e-02,\n",
       "       1.01907864e-02, 5.71678263e-03, 2.99450519e-03, 1.46243277e-03,\n",
       "       6.64742166e-04, 2.80668915e-04, 1.09826967e-04, 3.97246475e-05,\n",
       "       1.32415492e-05, 4.05353546e-06, 1.13498993e-06, 2.89311158e-07,\n",
       "       6.67641134e-08, 1.38567028e-08, 2.56605607e-09, 4.19900084e-10,\n",
       "       5.99857263e-11, 7.36666815e-12, 7.62069119e-13, 6.45821287e-14,\n",
       "       4.30547525e-15, 2.11744684e-16, 6.83047369e-18, 1.08420217e-19])"
      ]
     },
     "execution_count": 355,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "utils.computejerem(2,63)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 353,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.0"
      ]
     },
     "execution_count": 353,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "c.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 337,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0.25, 0.5 , 0.25])"
      ]
     },
     "execution_count": 337,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 271,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(6, 0, 0, 3),\n",
       " (3, 0, 6, 0),\n",
       " (0, 6, 3, 0),\n",
       " (3, 6, 0, 0),\n",
       " (0, 0, 3, 6),\n",
       " (6, 3, 0, 0),\n",
       " (0, 6, 0, 3),\n",
       " (6, 0, 3, 0),\n",
       " (0, 3, 0, 6),\n",
       " (0, 3, 6, 0),\n",
       " (0, 0, 6, 3),\n",
       " (3, 0, 0, 6)]"
      ]
     },
     "execution_count": 271,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "metadata": {},
   "outputs": [],
   "source": [
    "b = set([])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "nan"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.NaN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[False, False, False],\n",
       "       [ True,  True, False]])"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np \n",
    "a = np.array([[1,2,3],[4,5,3]])\n",
    "b = np.array([2,3,4])\n",
    "pos = a>b\n",
    "pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1, 2, 3],\n",
       "       [4, 5, 3]])"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a = a.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "457"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "data = pd.read_csv(\"utils/meta/metaphylo/Yun_Species_phylodistMat.csv\",sep=' ')\n",
    "len(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'conidiobolus_coronatus_nrrl_28638'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-88-2e7445ad090e>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mnames\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'NematocidaparisiiERTm3'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'VittaformacorneaeATCC50505'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mdata\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'conidiobolus_coronatus_nrrl_28638'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/Users/jeremie/anaconda2/envs/tensorflow1/lib/python2.7/site-packages/pandas/core/frame.pyc\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   2137\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_multilevel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2138\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2139\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_column\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2140\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2141\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_getitem_column\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/jeremie/anaconda2/envs/tensorflow1/lib/python2.7/site-packages/pandas/core/frame.pyc\u001b[0m in \u001b[0;36m_getitem_column\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   2144\u001b[0m         \u001b[0;31m# get column\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2145\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_unique\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2146\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_item_cache\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2147\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2148\u001b[0m         \u001b[0;31m# duplicate columns & possible reduce dimensionality\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/jeremie/anaconda2/envs/tensorflow1/lib/python2.7/site-packages/pandas/core/generic.pyc\u001b[0m in \u001b[0;36m_get_item_cache\u001b[0;34m(self, item)\u001b[0m\n\u001b[1;32m   1840\u001b[0m         \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcache\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1841\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mres\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1842\u001b[0;31m             \u001b[0mvalues\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1843\u001b[0m             \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_box_item_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalues\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1844\u001b[0m             \u001b[0mcache\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mres\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/jeremie/anaconda2/envs/tensorflow1/lib/python2.7/site-packages/pandas/core/internals.pyc\u001b[0m in \u001b[0;36mget\u001b[0;34m(self, item, fastpath)\u001b[0m\n\u001b[1;32m   3841\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3842\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0misna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3843\u001b[0;31m                 \u001b[0mloc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3844\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3845\u001b[0m                 \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0misna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/jeremie/anaconda2/envs/tensorflow1/lib/python2.7/site-packages/pandas/core/indexes/base.pyc\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   2525\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2526\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2527\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_maybe_cast_indexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2528\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2529\u001b[0m         \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_indexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmethod\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmethod\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtolerance\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtolerance\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'conidiobolus_coronatus_nrrl_28638'"
     ]
    }
   ],
   "source": [
    "names = ['NematocidaparisiiERTm3','VittaformacorneaeATCC50505']\n",
    "data['conidiobolus_coronatus_nrrl_28638']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.image.AxesImage at 0x11181cbd0>"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAQUAAAD8CAYAAAB+fLH0AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAIABJREFUeJztnX2MHdWV4H+nTbubXmi/bgNpfzTYMWRtK2b8GRpnpEUh8QYmGjxZwhB7gSRIrKIgJWw0EzIrZ7LKKEoWBWdnWWUWLZGBYYY4YWJQlChmSdBoBGQM2GuTbhO3g8PrhcYx7g9bnW78cfePqnrvVr1bH6/d3a/ee+cntbrq1r3nnqr3+vQ95557S4wxKIqiBLTUWgFFUfKFGgVFUUKoUVAUJYQaBUVRQqhRUBQlhBoFRVFCzIpREJGPi8jrIjIoIvfPRh+KoswOMtN5CiIyD/gN8DFgCNgHfNoY0z+jHSmKMivMxkjhQ8CgMea3xpj3gCeBW2ahH0VRZoGLZkHmEqBonQ8B1yU1mC9t5uo1hg5p4d1znkp/OD+fi1veC9U7efbfcHHLGZa2TgAwYc7TIfF2bcKcB0isM912rr7fPXcRC+edTW0fXI/WGTrTUbq3QF4WFs47W6FD9D7sPuN0i3ueWcqj92T3betkt7X1cPXhkh/ITqobh+t5uI5tOXHPLe7eo3LS9E4i+lm69Ir2kXTvA4fOnDDGXJ7W72wYBXGUVfgoInIPcA/AoiXzWPLan/HAsZd4/OT1AAyM97CqczjUZu+bK+ktjPLNZT8G4MDkUta2D8UqcmByKUBinem2c/X9+MnruaP7xdT2wfVonb869melewvkZeGO7hcrdIjeh91nnG5xzzNLefSe7L5tney2th6uPlzyA9lJdeNwPQ/XsS0n7rnF3XtUTpreSUQ/S5de0T6S7v2z//al32XpdzaMwhDQa50vBd6KVjLGPAw8DLDxj9rNAy+8xF8s62P74fIgY01HMdRmoOAZimvnt5e6Kh/HqUJKnem2q+x7TUfRL0trX/7w7Drhe6u8/ziibeznU74+FNvG1mv65dF7Kvcd1olQ/eTPslJ+IDu5bhyVz8N1HJYT99zc914pJ03veKKfpUuvbN+3Sh2TmA2jsA+4RkSWA/8PuB3YltTg3XMX8fjJ69l+uMgTK5dyZNcGZKSVp7rWh+q1H21joPsq1tzk/bEcmuhN/MM5NOHZpqx/XNW0c/W9e3gTsC+1fXA9WmfvmytD5568LOyr0MHuZ01HMdRnnG5xzzNLefSe7L5tney2th6uPlzyA9lJdeNwPQ/XsS0n7rnF3XtUTpreSUQ/S5de0T6S7/2Hmfqd8UCjMeYscC/wc2AA2G2M+XXW9kd2beCaz7wy02opipKR2RgpYIz5KfDTrPX/cH4+A+M9AMhIK4M7+7j6vpcY3NkXqtc2AtBSsoxBmzjSrl9IO1ed4miBQ52909br1GhH6L9OcbSQqgfAoc5wm+Dc7me695S1PFrHPrd1ylI/TX5a3TiyyMuiVzX6VPsZ2MR9ltORW03fs2IUquXilvdY1TnMmo5iyWUIDENxx2YmV0wBMNXVxlT3+dAwKctwrFr3IWs7V8zjQnTbWwi7DwOFbB9ktE20X5cecbrNRHn0Gbh0iqubJj84r+aesvaZJDurvnFyssiOkuX7VM337WcZ+811mnNxx2Z6v/FCrdVQlKYi10ZhcsWUxhgUZY7JtVEIUMOgKHPHjK99mA4LVr7PXPnf/hO9hVF++9xywAsqTnWF6/V+4wWO37uZJbe+AXiBmN7CaKzcIFCXVGe67Vx99x9bzOplb6W2twOIdp2B/Vexal05v6T/2OJM+q5eVk4DCXSI3kc0aOnSLe55ZimP3pPdt62T3dbWw9WHS34gO6luHK7n4Tq25cQ9t7h7j8pJ0zuJ6Gfp0ivaR9K9v3bLN14xxmxM6zcngcYz9BZGWdU5zED3VX5pC1Pd4fTR4/du5oqHXmDV58pJk9GsRxdZ6kynXfR6sVAIlVWrW3/X4tB5sZBt9iHaJtqvS4843WaiPPoMXDrF1U2TH5xXc09Z+0ySnVXfODlZZEfJ8n2q5vv2WsZ+c2EUlrZO8M1lP/ay3xISk3YPb2LV54SD6w3bD3sZWvU++/BAz/7Y86x639l5ItQmOD/YPQl4mW7BcYAr++1g9+S0y235dn9BRqOtY1S3uD5c8gPZSXXjcD0P17EtJ+65xd17VE6a3klEP0uXXtE+ku69IWYf4th+eIgnVi5Nr6goStXkYqQwYc77izqGEhOTbN/t2lfh4HoDr/ZW1Auoh+Ql+7/MwHhP6DyaPhvHQWshzKGJ3tJ5eaFO5YIoVy588BlMpzwsfyjUt61TuG1ZD1cfbvlDqXXjcD8P13H8giiXvq46wfU0vZOo/Cxdeg1VPEsX1fSdC6PQIS2sbR8qDTUD4hZElcpf7eXgekNxx+ZSnfu3766Qn2f3QRdE6YKoOGq1IKou3QcbTXBSlJml7o0CqGFQlJmkIYwCqGFQlJkiFzGFCyEUQ9hOaT+Gr/c9XTulFKWOaZiRQoCmRCvKhVH3I4UoX+97Gg57I4bhPSsBWLNyerMPitKM5MIoTCdPIa5OwPCelfRsHaBl7WoOfT/bfH8WuUl1NE+hLFvzFOo3T6Hh3AeblrWrOX9A30GjKNWQi5HCtJOXHHVK5SuL/gjBWyvR+vyizPpk2QbNVefUaAcDhR6KowW2XHlYk5fQ5CVNXsoprc8v4swNb9daDUWpC5rCKIAaBkXJStMYBVDDoChZaCqjAGoYFCWNpjMKoIZBUZJoSqMAahgUJY5cTEnWisAwHNm1IVR+aWEipoWiND5NO1II0LUSihKm6Y0CqGFQFBs1Cj6BYWg/2lZrVRSlpqhRsNCNWhQlJ4HG2VglaV9PWssQCipumGB4zyp6tg5wZNcGklYenBrtqLguI60UC4XSGog0vUBXSeoqyXh0lWSO0BiD0szkYqQwG6skL4QihEYMLmSkld5lkXdJdnV47/aDCj3j0FWSukoyDl0lmUN0xKA0I2oUUlDDoDQbahQyoIZBaSbUKGREDYPSLKhRqAI1DEozoEahStQwKI1O6pSkiHwf+ARw3BjzQb+sG/gBsAw4BtxmjBkREQH+O3AzMAF8xhjzaloftUxecnFqtAPwph1dDO7s4+r7XqK4Z1WoXJOXyrI1eal+k5ey5CnsAh4CHrPK7geeM8Z8S0Tu98+/AtwEXOP/XAd8z/+dSB7zFICKPIRQHT+Pwd4lWvMUbNmap9CweQrGmH8GTkaKbwEe9Y8fBbZa5Y8Zj5eAgohk31u9zgj2Y+g/tpj+Y4tpP9pG/7HFnH5jQa1VU5RpM92YwvuMMW8D+L+v8MuXQGhJwJBfVoGI3CMiL4vIy79/99w01ag9GmNQGo2ZDjSKo8y4KhpjHjbGbDTGbLx84bwZVmNuUcOgNBLTXfvwjogsMsa87bsHx/3yIcCOji0F3koT9u65i3j85PWs6Siye3gT4AUHo8G6/mOLKRbK5a46NkGAMQgcZiUIMPZ3xbeTkVb6uzpK+y8Ey66LOzbTNgJ7l6+M1c0OfNpxg71vrgydB88inX2RNt55EKhc01GsCFq64hWHJnqnXW7Lt/srf6b7Ktraerj6cMkPZCfVjcP1PFzHtpy45xZ371E5aXonEf0sXXpF+0i+9x9m6ne6RuEZ4C7gW/7vp63ye0XkSbwA41jgZiSxcN5Z7uh+0Q+U+A+hs/IB7mZTKIDnqmNzqNOfyUgwHC6KBe+PtreQEGgsFOgtjNLP4nKZbxgGd/bxyYTXxgV6AdzZeaJcfuXh0Hn0DymOaJvgPIhcXzu/vSL67wpKHWx3B8KylNvy7f6CzzSsI5Hr7j5c8gPZSXXjcD0P17EtJ+65xd17VE6a3smEP0uXXpXPOv7eH8zYa5YpyX8EbgAuE5Eh4K/xjMFuEbkbeBP4lF/9p3jTkYN4U5KfzaJEXqcks+ynYE9bto2UpysHEt5dqVOSOiWZhdxOSRpjPh1z6UZHXQN8IVPPc8iWKw9XVT8YWazqHE6ss6pzGK4sl+1dvpJPXnmYAX9WYm8kjyHg1GgHq5elelWKUhOaYj+F6eYypLVz6ReU7fXzGMa39QEwtqIc0710w0jJ4GieguYpxJHbPAVl+oxv66PzH16qtRqKUhVqFGYZNQxKvaFGYRYZW9HC2IqW0qzEgqPnWXD0fK3VUpRE1CjMETpiUOoFNQpziBoGpR7IxezDbOcpVEuWdnH6HersZWC8h1OjHVy6YaR0bcLfFHriNm9WIshj0DwFzVOII7d5Csr0WL3srdg8hyCPYXxbH3xzjhVTlBTUfagR6kooeSUXI4Vqk5ce6NkPeEPvxHxvfyheXb55tnauvtd0FLmz80Rq+4Pdk94I4ZvwF8v6GN/Wx+e/9iOKo4XQUDSr+3OgI9wmOLfdjyyuSFydLOXROva5rVOW+mny0+rGkUVeFr2q0afaz8Am7rNMkxtdZ+KhyUt1g44YlLyhRqHGfP5rP2L74SGeWBkNVilKbRBvDVNt6Vndbdb/3X9kVecwe99cCXiLhkJvhAZOv7EA03WG/3CttxfswHhP8qKl8fSFTdNt5+p775sr2XLl4dT2tltgz6j0bB1g2FpElXVbt0uWj4XaBOfBas9LCxMVe0pEn21Qf7rltny7v0sLEyGd7La2Hq4+XPID2Ul143A9D9exLSfuucXde1ROmt5JRD9Ll17RPsC9AHBgvIef/bv/8YoxZmNav7mIKSxtneCby34ciim4NqTYXdjUODEFHzuG8J09H6Nn6wDbD3v+3+5Ctk1WbuuxNlkpbCqd52aTFUsnu62th26yUkn0s8y6yYorpnCwe5KfZew3F0ZBKRO4EuPb+lh4z+9qrY7ShGhMIYdo8FGpJbmIKSz/4CXmk39/c8UejdHt0PqPLebSwgRfXvkskD5kzLJ33XTbOd2bYW+4l9Y+Ol1XOt5/FavWlUcHZ254m5a1q1P1ff1LF4fOgw1cgnhFb2G0IhvUtdWc65lnLbfl2/31FkbpP7a4YlMZ+3pcHy75geykunG4nofr2JYT99zi7j0qJ03vJOznFqdXtA+gwlUD7zv34LofZoop6Eghx7SsXc35A/21VkNpMtQo5Bw1DMpco0ahDlDDoMwlahTqBDUMylyhRqGOUMOgzAVqFOoMNQzKbKNGoQ5Rw6DMJmoU6hQ1DMpsoUahjlHDoMwGuVj7UO0LZuNeuhkly4s3p9vO3bf3QtC09vZ+hfbmI99ZXghlo339S7dk0HYD13zmFY7s8jaBXL3srfIimk5r0UxnhgVRMS/szVJuy7f7W9NRZDeOBVGdkQVRjj5c8gPZSXXjcD0P13Fo4VHMc4u796icNL2TsJ9bnF7RPiBmQVQVL5jVkUIDcGSXZxjaj7bVWhWlAVCj0CAEL5xRlAtFF0RNs91cLIjqP7Y4k772YqMzN7xdciVkpBUA03WmdBxgus5UyJGR1lD/AbogqkwzLIjSmMI0281FTGE31W+y8vVdt3DNZ16h9flFFAvWF6eQYZVkoeD+QmlMIXRvSfcelaMxBSUXtPrvlWj5P121VkWpQ3IxUqj2DVHBdmZpb73J8uac6bZz9R280Setvb0Fm+1KnBrtCJ1Hh6Vx2P+pZKSVYqFAy73LueKhFziycQNFKvcRdP2/ivYfELfVfMj1idSJ7kMZ/W+aVD9NflrdOLLIy6JXNfqk1UvCfm5JekWvRd/GBdW9IUpHCg3K+Y+OlGYlTr+xgPajbTo7oWQiFyOFal8GU/bVk2MKgWWsNqaQrV1l32s6in5ZWnu3xd5bWBm654FCtv8sdpunutaXA2DA4M4+rr7vJY7fuxmAVn/HX2dMgfjgaDXl0c9woNCTqX2aLPs8S9044vpMkp1V3zg5WWRHiT63LM8a4r532UfLuTAK6j54zKT7EMhrP9nC8Xs3c8VDLzC+rY93fDdC3Qd1H+JQ96FJ0M1glazkYqSgzB6TK6ZoLUzwzmgH72zxYgzFHZs53d1aUbf9ZAtcWwMllVyROlIQkV4R+aWIDIjIr0Xki355t4g8KyJH/N9dfrmIyN+KyKCIHBSR9bN9E0p2NPNRSSPLSOEs8GVjzKsicinwiog8C3wGeM4Y8y0RuR+4H/gKcBNwjf9zHfA9/3csGmj0mI1AI/jJS/7x6e7WUvCxuGNzSE7biAYa0+Q0Q6AxdaRgjHnbGPOqf3wKGACWALcAj/rVHgW2+se3AI8Zj5eAgogsyqyRMicEI4YFR8+z4Oj5Wquj5IiqAo0isgxYB/wKeJ8x5m3wDAdwhV9tCeHg9pBfpuQMDT4qLjIbBRG5BHgK+JIxZjypqqOsYtWViNwjIi+LyMu/f/dcVjWUGUYNgxIlk1EQkVY8g/CEMeaf/OJ3ArfA/33cLx8C7EnppUB4iRxgjHnYGLPRGLPx8oXzpqu/cgGMrWhhbEWLBh+VEKmBRhER4BFgwBhjL7R6BrgL+Jb/+2mr/F4ReRIvwDgWuBlxaPKSx2wkLwGhtQ/tJ73/A20jYRlBgtOhz2nyUha5jZy8lGX24cPAHcAhETngl/0VnjHYLSJ3A28Cn/Kv/RS4GRgEJoDPpnWgsw8eMzH7MLCup7zc1trXoWRsrq1sE3Doc70cXG9ofT4cF44zTnZ5tI59fmq0o+JekuonyQ/uTWcf4vWZ9TRnY8y/4I4TANzoqG+AL2TWQMkVwbLrYLry/Te+UWONlLlG05yVCjTG0NyoUVCcqGFoXnTtg1JByWW40XMl2JMt4Kk0BrkwCjr74DETsw9xEeus0fmKPvcU6Nk6QHHPqor6p0Y7Sllq9s5O9mxHkfCMCJH6dvtoGM0l/9BEr84+pMidi9mHWUdnHzxmZPbBsc7gQiL1AMU9qzzDsGMzkyumAG8H6SJUrLOA8FqL3sIo/V0dlbsvW9eDc1edqHxd+5Cuz6yvfVAU0BhDM6FGQcnE5IopfRNVk6BGQakKHTE0PrmIKSj5p/SGp2WUZiUWbllC/zYvABh6A9W67G9BUvJHLoyCzj545HL2AffMwMItS2jd+zJyUx8AbSfLg87iaEFnH3T24cLQ2QeP3M4+UDkz0L+tA7nJ28FpfFsfF594r3St40adfZiO7Cg6+6DUJbofQ+ORi5GCug8e9eQ+yEgrbSdbuPjEe5zZspHWvS/TsnZ1SQd1H+rXfdCRgjIjtKxdzfkD/bVWQ5kB1CgoM4YahsYgF+6DBho96inQyLpRiqMFOm70A4Elt8NLiT69s49Llo9poLEK2VFyu8nKXKAxBY96iilEy+0FUaf990oc2bVBYwoaU1AUuGT5WCklWqk/cjFSUPfBo67cB9zDeyivkhy2Vlee6uoItQ0yIE9RzogMju069nVbb3Uf4vW5UPdBvC0Va8vyD15iPvn3N7Omo8ju4U2AN3SqSJg5tphLCxN8eeWzgDeUTHrI9sal1ZClnavv3cObuK1nX2p720UIDQP3X8Wqdb8rnfcfW5xJ31IKst8mOA9cgd7CaIVb4Pwjt6YSbWSkFdN1JrHcTnM2XWdK520nW+j9xgvOV9QBTHWVz4Nju459HWDJrW84vxuuMtf9Qfh5uI5tOXHPza7nqhNcT5KdRvSzdOkV7QPKm9vaHJro5cF1P3zFGLMxrd9cjBQWzjvLHd0v+hbO34m40/FHxyZWdQ5zZ+cJwPOdkv6bB75VtSOFLO3cfe/jzs4Tqe1tn+9ARzm+8J3lhdAHuptNmfSNtint5txp7eYc2WLduZtzZ69zdFIsuL/MdrkdM+gtjJa3me/qKC2iGt/Wx9iKFiZXTDHlG43AqEw5DI9dFtS/r2ef87vhKnPdX3DvSce2nLjnZtdz1QmuJ8lOI/pZuvSK9gGU/j5sDrYP8WBFqRuNKShzgmY+1g9qFJQ5Qd9EVT+oUVDmhMkVU6GNWjqPttB5VL9+eSQXMQXNU/BolDyFonUuI61ewNAKRAavqBvf1se4FZCcwtqTIVIWLM3WPIX0/jVPQalLNMaQX3IxUtA8BY9Gy1MALw9hqqstNLMwthHGR1o5vim8H8OxbeHp8Sms2Qc0T0H3U1CaAh0x5I9cjBQ0puDRLDEFKMcI7P0Ygq3d7DoaU9Dt2NR9aAL3AcruQOAyBFu72ZmPbSMwuULdh6T+1H1QGhrNY8gHahSUXBEYhgVHz9P1m/fSGygzjhoFJXdo8LG25CKmoIFGj2YMNNqBxGAl5MUnzlYEHzXQqIFGQAONzRBotPMQgqDimL+41/VeiTXbNdCYpo8GGpWGRl2JuScXIwV1Hzya3X1wrX2IvldC3Yf0/nXtg9IU6Pbxc0cuRgoaU/DIQ0whbisv505NVrk9wlnTUQxtSbd7eBP3ReRGt6xz9XFoorcUQ/DqCwfXG4b3VI5msoyq7DpZjpPkJrUJnmG9xhRSjYKItAP/DLT59X9kjPlrEVkOPAl0A68Cdxhj3hORNuAxYAPwLvDnxphjSX2o++CRB/dhuuVJw1tbpyz1k+QP7ynQs3WAgZ2VKdED3QsqZFyyfKx0bG8rn3Rs/+lF96y03SWX6wThGZJ6dB+yjBSmgI8YY06LSCvwLyLyM+A/AzuNMU+KyN8BdwPf83+PGGOuFpHbgW8Df57UgY4UPPIwUpiJ8uhn6NIprm4W+QPWeyUC7BWVNqHNTq2ypGNXm6i8uJkXW+d6HSmkxhSMx2n/tNX/McBHgB/55Y8CW/3jW/xz/Os3iohk1khRMhDs4NR+tI32o221VqehyBRoFJF5InIAOA48CxwFRo0xZ/0qQ8AS/3gJvvH0r48BCx0y7xGRl0Xk5d+/e+7C7kJpSuyUaGXmyGQUjDHnjDFrgaXAh4BVrmr+b9eooOLlEsaYh40xG40xGy9fOC+rvooSQvMYZp6qZh+MMaMi8jzQBxRE5CJ/NLAUCN5IMgT0AkMichGwADiZJFcDjR4aaEyXXxwtlHIaprrPc7yb0g5Og5HgI4TfR6GBxhkKNIrI5cAZ3yBcDHwUL3j4S+BWvBmIu4Cn/SbP+Ocv+td/YVJeQ6WBRg8NNGYMNHYvqAgqDu6s3I8BoHedBho9Zvat04uAR0VkHp67sdsY8xMR6QeeFJG/AfYDj/j1HwEeF5FBvBHC7Zm1UZQLwH4TVYkba6dPvZJqFIwxB4F1jvLf4sUXouWTwKdmRDtFqZIgxhAyDEpVaJqz0nBo8PHCUKOgNCRqGKZPLtY+KEo1XLJ8rPJV9IWCF1S0Ywj3LOLMDW+XMh/FX6nZ39XhPGZd5Z4RzYgaBaWhCTIfizs2l3Z2mupqcx5XRs6aE3UflIYnmJXo+s17uhlsBnIxUtDkJQ9NXkqXXxwtODd9idsIJtjkxd6opW1kPu0ng9ToltKxS3YzJi/pSEFpGnSjlmzkYqSgGY0emtGYTc9o1mFcGXiBxKmuNoavm0/byHz4481c8dALnNni7Q47tmI+wf9GzWj0yIVRUPfBQ92HdPnTdR+AkpsQ50qo++Ch7oPSlKgrEU8uRgrqPnio+zB77oOH9z9wbEWlKzHygfnqPvjkwiio++Ch7kO6/JlwH+wZh8CVaL+sT90HH3UflKZm5APzNSU6Qi5GCuo+eKj7MPPuA+tGS5mKwcirtzBacTxx2yp6tg5wemcfpyj/55fIS2yCazLSGjp26azuwwWg7oOHug/p8qt1H6J1IH7npdP+Ri1ntmzkdzd5fxrBm6wCXG+0itapd/chF0ZBRwoeOlKYhZFCpA7E77x0io5SjMFs8xZRRV9lZ78Q1z526awjBUVpAH5300WYbd4iKu9t12dD14ev82Yy2kYozWqUFlQ1CBpoVBQHzRx8zMVIQWMKHhpTSJc/mzEFGWktxQouPnE2lPkY0DYyHwimN1us4zL1HlPQkYKiJNCMmY9qFBQlhWYzDGoUFCUDzWQYchFT0ClJD52SrP2UZDDVWJ5lmF8W4K+VKM9KeH8+0RmKep+S1JGColRBM8xK5GKkoLMPHjr7kC5/rmYfKhdRedizEvgbtbQdnwjVqffZh1wYBXUfPNR9yI/7EF1uXcb/k/ENg/cmqvAqSXUfFKVJaVRXIhcjBXUfPNR9SJefB/cBApehw5ngVO/ug44UFOUCabTpSjUKijIDNJJhUKOgKDNEoxiGXMQUdPbBQ2cf0vW8rWcfhyZ6K8pdZVGCeM2ajmLisS3HjvHY+tj1wnWEg+sNgzuvor9rsfUiW++4v2txoo427UfbeGp0vSc10jYqNzgGeGDLfoe0OttPQQONHhpozCY/rW4cWeRl0SutzeDOq7j6vpcY3NlX2pVpitbQbk1ZaBuBqZHKnZ6Cc1tucAxw8L3JClmap1BCRwrR60llM1U+myOFpPn+rDkAcX1mySVI0xe8/9iD/tZugzv7AG/HJnu3pixMdbU5d3oKzm25wTE0yM5LOlLw0JFCNvl5HykEU5uBYSju2FxXIwUNNCrKLFLcsZneb7xQazWqIhcjBXUfPNR9aBz3oZQuTWtpxHBmy0aObTOZdAR1H9R9QN2HrPLrxX2wh/ZB5qPc1JeqZ4AGGnWkoCOFBh0pABzbZpCbyjGGOCZXTJWOazVSyBxTEJF5IrJfRH7iny8XkV+JyBER+YGIzPfL2/zzQf/6sszaKEqDUw8xhmoCjV8EBqzzbwM7jTHXACPA3X753cCIMeZqYKdfT1EUn7wbhkxGQUSWAn8C/G//XICPAD/yqzwKbPWPb/HP8a/f6NdXFMUnz4Yh60jhu8BfAsE60oXAqDEm2JxuCFjiHy/B38/Cvz7m1w8hIveIyMsi8vLv3z03TfUVpX7Jq2FINQoi8gnguDHmFbvYUdVkuFYuMOZhY8xGY8zGyxfOy6SsojQagWFYcPQ8C46eT28wB2QZKXwY+FMROQY8iec2fBcoiEgwe7EUeMs/HgJ6AfzrC4CTM6izojQUedvBKdUoGGO+aoxZaoxZBtwO/MIYsx34JXCrX+0u4Gn/+Bn/HP/6L4wx2TM2FKUJyZNhuJA8ha8AT4rI3wD7gUf88keAx0VkEG+EcPuFqagojYOdh1Au836/s8V723WQx+C5E5VvuQ7OAcY2zryOVRkFY8zzwPP+8W+BDzkcOUVcAAAHEklEQVTqTAKfmgHdFKXpCGIMSQlOs40uiFKUnFHrWYlcpDm/e+4iHj95PWs6iuwe3gR4+fLRNN/+Y4spFgrO3W9c2DvqVEOWdq6+Pd33pba31zfYefOn31jA7sKm0nn/sWy79Owm3CY4D9ZODBR6KtZRuFKoXc88a7kt3+5voNAT0slua+vh6sMlP5CdVDcO1/NwHdty4p5b3L2Dt2NSMNwPhvnBse0CuAhmIEoxhm3eWomLT5zlD5eV/1yDXaWjfQA8Nn5ZhVzvO/fDxL4DcmEUFs47yx3dL/o52/sAb5FPxR8dm1jVOcydnScAbyvrpPUJwVbX1a59yNLO3fc+7uw8kdre3oL7QEd5cdR3lhe4rWdf6Tz6hxRHtE1wHixCWtNRrFiQ5DJYrmeetdyWb/e3pqMY0slua+vh6sMlP5CdVDcO1/NwHYe2Y4t5bnH3DvDU6HqmRvy1DyPWGoWRLJusWEbDNwzFHZuZ7J7PVFf50mS3947LyRVToT6A0t+HzcH2IR5M6TlA3QdFyTGBK9H1m/fmLI8hFyMFXTrtoUuns8mvx6XTWXdeClwB8FyEye75oXdXBiOE4CU14yO6dFqXTjvQpdPJOmbtczaXTmfdo9GOOfzhsouY6vJ+B++unLoumJVo8eVO1W7ptKIotWFsRQsjH5g/Z7MSahQUpY6YC8OQC/dBYwoeGlPIJr/ZYgoQxBA8Wcfv3cwVD73A+La+WYkp6EhBUeqQ2VwrkYuRggYaPTTQqIFGey3DVFdbaa3E+Egrpqu8bmJ8pJXjm7w9H4/f6wUfJ1fobs7qPgQy1H1oGPfBJstuzlFXAtR9UJSmZ6ZdiVyMFNR98FD3Qd0HmyxbvE+uOBNyJYo7NmuegqIoHjM1XZmLkUIzYy9eOXTl4chiln2VDVJkBIuyILywy16EFZRFiVtglqXclm/3Fyxyiy7SiS4ac/Xhkh/ITqobh+t5uI5tOXHPLe7eAR7Ysp+D70168nz/PjiuZtT62Phl5c8y0jYqt6Tb5+HfL17L9sOV8a6sqyR1pKAoDcb2w0M8sXJpesUYdKSgKA1IYBiO7NoAeDMiOlJQlCbnyC5vz8dqUaOgKA1MYBg6j2b/U8+F+9DMyUt2nYHxntK9BfKyYAe6Ah3C/QxF+nTrFvc8s5RH78nu29Yp3Lash6sPt/yh1LpxuJ+H63jI0aZ8b/G6lesE19P0TqLys3TpVfl9C7434iczBQlOWZE8vJJBRE4Br9dajwvgMqByD6z6oJ51B9W/Gq4yxlyeVikXIwXgdWPMLOxgPzeIyMv1qn896w6q/2ygMQVFUUKoUVAUJURejMLDtVbgAqln/etZd1D9Z5xcBBoVRckPeRkpKIqSE2puFETk4yLyuogMisj9tdbHhYh8X0SOi8hrVlm3iDwrIkf8311+uYjI3/r3c1BE1tdOcxCRXhH5pYgMiMivReSL9aK/iLSLyL+KyP/1df+vfvlyEfmVr/sPRGS+X97mnw/615fVSncbEZknIvtF5Cf+ea71r6lREJF5wP8EbgJWA58WkdW11CmGXcDHI2X3A88ZY64BnvPPwbuXa/yfe4DvzZGOcZwFvmyMWQX0AV/wn3E96D8FfMQY80fAWuDjItIHfBvY6es+Atzt178bGDHGXA3s9OvlgS8CA9Z5vvU3xtTsB7ge+Ll1/lXgq7XUKUHXZcBr1vnrwCL/eBFergXA/wI+7aqXhx/gaeBj9aY/0AG8ClyHl+xzUfQ7BPwcuN4/vsivJzXWeyme0f0I8BNA8q5/rd2HJYC9fc2QX1YPvM8Y8zaA//sKvzy39+QPR9cBv6JO9PeH3geA48CzwFFg1Bhz1qFfSXf/+hiwcG41ruC7wF8CwYsgF5Jz/WttFMRRVu/TIbm8JxG5BHgK+JIxZjypqqOsZvobY84ZY9bi/cf9ELDKVc3/nSvdReQTwHFjjL1UMUnHXOhfa6MwBNirfpYCb9VIl2p5R0QWAfi/j/vlubsnEWnFMwhPGGP+yS+uG/0BjDGjwPN4cZGCiAQp+rZ+Jd396wuAk3OraYgPA38qIseAJ/FciO+Sc/1rbRT2Adf40dj5wO3AMzXWKSvPAHf5x3fh+epB+Z1+FL8PGAuG6bVARAR4BBgwxjxoXcq9/iJyuYgU/OOLgY/iBex+CdzqV4vqHtzTrcAvjO+g1wJjzFeNMUuNMcvwvtu/MMZsJ+/65yCAdDPwGzxf8b/UWp8YHf8ReBs4g2fN78bz9Z4Djvi/u/26gjejchQ4BGysse5/jDcEPQgc8H9urgf9gWuB/b7urwFf88vfD/wrMIi3nVCbX97unw/6199f6++OdS83AD+pB/01o1FRlBC1dh8URckZahQURQmhRkFRlBBqFBRFCaFGQVGUEGoUFEUJoUZBUZQQahQURQnx/wH+6+/6Xg891wAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x11161c6d0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from matplotlib.pyplot import *\n",
    "imshow(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[-1, 0, 1, 2]"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "range(-1,3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'True'"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "str(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sklearn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
