\hypertarget{class_py_c_u_b_1_1espece_1_1_espece}{}\section{Py\+C\+U\+B.\+espece.\+Espece Class Reference}
\label{class_py_c_u_b_1_1espece_1_1_espece}\index{Py\+C\+U\+B.\+espece.\+Espece@{Py\+C\+U\+B.\+espece.\+Espece}}


docstring for \mbox{\hyperlink{class_py_c_u_b_1_1espece_1_1_espece}{Espece}}  


Inheritance diagram for Py\+C\+U\+B.\+espece.\+Espece\+:\begin{figure}[H]
\begin{center}
\leavevmode
\includegraphics[height=2.000000cm]{class_py_c_u_b_1_1espece_1_1_espece}
\end{center}
\end{figure}
\subsection*{Public Member Functions}
\begin{DoxyCompactItemize}
\item 
\mbox{\Hypertarget{class_py_c_u_b_1_1espece_1_1_espece_a13a7403216d2c03d6e9ec81d6034196e}\label{class_py_c_u_b_1_1espece_1_1_espece_a13a7403216d2c03d6e9ec81d6034196e}} 
def \mbox{\hyperlink{class_py_c_u_b_1_1espece_1_1_espece_a13a7403216d2c03d6e9ec81d6034196e}{\+\_\+\+\_\+init\+\_\+\+\_\+}} (self, kwargs)
\begin{DoxyCompactList}\small\item\em can intialize the file from kwargs as a raw dictionnary for json format (output of dictify) or from regular args. \end{DoxyCompactList}\item 
\mbox{\Hypertarget{class_py_c_u_b_1_1espece_1_1_espece_ab33e33544d897c99c08074e90ca0105f}\label{class_py_c_u_b_1_1espece_1_1_espece_ab33e33544d897c99c08074e90ca0105f}} 
def \mbox{\hyperlink{class_py_c_u_b_1_1espece_1_1_espece_ab33e33544d897c99c08074e90ca0105f}{plot}} (self)
\begin{DoxyCompactList}\small\item\em will present some interesting info about this species. \end{DoxyCompactList}\item 
def \mbox{\hyperlink{class_py_c_u_b_1_1espece_1_1_espece_afa35655bb969cbc02456aa74228a1b8d}{get\+\_\+t\+R\+N\+Acopy}} (self, by=\char`\"{}entropy\char`\"{}, setnans=False, kingdom=\textquotesingle{}fungi\textquotesingle{}, base\+C\+Nvalue=2)
\begin{DoxyCompactList}\small\item\em Retrieves t\+R\+NA copy numbers from ensembl DB. \end{DoxyCompactList}\item 
def \mbox{\hyperlink{class_py_c_u_b_1_1espece_1_1_espece_af716fcd7e7e34e12cd7786469d695422}{gettaxons}} (self, kingdom=\textquotesingle{}fungi\textquotesingle{})
\begin{DoxyCompactList}\small\item\em Pars the ensemblgenomes R\+E\+ST A\+PI to retrieve the taxons id. \end{DoxyCompactList}\item 
\mbox{\Hypertarget{class_py_c_u_b_1_1espece_1_1_espece_aed93f9be40c4521c6aed536125268432}\label{class_py_c_u_b_1_1espece_1_1_espece_aed93f9be40c4521c6aed536125268432}} 
def \mbox{\hyperlink{class_py_c_u_b_1_1espece_1_1_espece_aed93f9be40c4521c6aed536125268432}{get\+\_\+epigenomes}} (self)
\begin{DoxyCompactList}\small\item\em get from ensembl all the data about the epigenome that could help asking interesting questions about the C\+UB \end{DoxyCompactList}\end{DoxyCompactItemize}
\subsection*{Public Attributes}
\begin{DoxyCompactItemize}
\item 
\mbox{\Hypertarget{class_py_c_u_b_1_1espece_1_1_espece_a3a33da4c7b0c9163b8673166b306226c}\label{class_py_c_u_b_1_1espece_1_1_espece_a3a33da4c7b0c9163b8673166b306226c}} 
{\bfseries name}
\item 
\mbox{\Hypertarget{class_py_c_u_b_1_1espece_1_1_espece_ae98c1d182e761e4c09a10f8a9994124e}\label{class_py_c_u_b_1_1espece_1_1_espece_ae98c1d182e761e4c09a10f8a9994124e}} 
{\bfseries metadata}
\item 
\mbox{\Hypertarget{class_py_c_u_b_1_1espece_1_1_espece_afbc7dbe8f3dc1f035837bb2ed610f246}\label{class_py_c_u_b_1_1espece_1_1_espece_afbc7dbe8f3dc1f035837bb2ed610f246}} 
{\bfseries is\+\_\+stored}
\item 
\mbox{\Hypertarget{class_py_c_u_b_1_1espece_1_1_espece_adb2006eee970b43f6abe9554690cb424}\label{class_py_c_u_b_1_1espece_1_1_espece_adb2006eee970b43f6abe9554690cb424}} 
{\bfseries num\+\_\+genes}
\item 
\mbox{\Hypertarget{class_py_c_u_b_1_1espece_1_1_espece_ad40242dc0031d44de4c363bd716489e2}\label{class_py_c_u_b_1_1espece_1_1_espece_ad40242dc0031d44de4c363bd716489e2}} 
{\bfseries genome\+\_\+size}
\end{DoxyCompactItemize}
\subsection*{Static Public Attributes}
\begin{DoxyCompactItemize}
\item 
\mbox{\Hypertarget{class_py_c_u_b_1_1espece_1_1_espece_a143141fe25ea32dc24abd314a0f5fdbd}\label{class_py_c_u_b_1_1espece_1_1_espece_a143141fe25ea32dc24abd314a0f5fdbd}} 
{\bfseries code} = None
\item 
\mbox{\Hypertarget{class_py_c_u_b_1_1espece_1_1_espece_ac3d8e577f2e9a0fe6853d51c4bec238d}\label{class_py_c_u_b_1_1espece_1_1_espece_ac3d8e577f2e9a0fe6853d51c4bec238d}} 
int {\bfseries num\+\_\+genes} = 0
\item 
\mbox{\Hypertarget{class_py_c_u_b_1_1espece_1_1_espece_a87ca67e0c2e560e88b72a8bdba456756}\label{class_py_c_u_b_1_1espece_1_1_espece_a87ca67e0c2e560e88b72a8bdba456756}} 
int {\bfseries genome\+\_\+size} = 0
\item 
\mbox{\Hypertarget{class_py_c_u_b_1_1espece_1_1_espece_ae066a956b3f9f29d676162fbf9773a0f}\label{class_py_c_u_b_1_1espece_1_1_espece_ae066a956b3f9f29d676162fbf9773a0f}} 
{\bfseries link} = None
\item 
dictionary {\bfseries metadata}
\item 
\mbox{\Hypertarget{class_py_c_u_b_1_1espece_1_1_espece_a8befee83c9b6ecfc8b07e66b6251afa4}\label{class_py_c_u_b_1_1espece_1_1_espece_a8befee83c9b6ecfc8b07e66b6251afa4}} 
bool {\bfseries is\+\_\+stored} = False
\item 
\mbox{\Hypertarget{class_py_c_u_b_1_1espece_1_1_espece_accf54132a0346377a2cf7a461a129dc1}\label{class_py_c_u_b_1_1espece_1_1_espece_accf54132a0346377a2cf7a461a129dc1}} 
string {\bfseries name} = \textquotesingle{}\textquotesingle{}
\item 
\mbox{\Hypertarget{class_py_c_u_b_1_1espece_1_1_espece_aae7cf0e3c6f359bd2970006817e80ed3}\label{class_py_c_u_b_1_1espece_1_1_espece_aae7cf0e3c6f359bd2970006817e80ed3}} 
{\bfseries taxonid} = None
\item 
\mbox{\Hypertarget{class_py_c_u_b_1_1espece_1_1_espece_a89cd4e211bf028a4d141202e8a885b58}\label{class_py_c_u_b_1_1espece_1_1_espece_a89cd4e211bf028a4d141202e8a885b58}} 
{\bfseries copynumbers} = None
\item 
\mbox{\Hypertarget{class_py_c_u_b_1_1espece_1_1_espece_a2dc4b3f0d2006aac69a9846ba834d209}\label{class_py_c_u_b_1_1espece_1_1_espece_a2dc4b3f0d2006aac69a9846ba834d209}} 
{\bfseries average\+\_\+entropy} = None
\item 
\mbox{\Hypertarget{class_py_c_u_b_1_1espece_1_1_espece_a2991e13a2c75e2e899f3f8f773d1d9b6}\label{class_py_c_u_b_1_1espece_1_1_espece_a2991e13a2c75e2e899f3f8f773d1d9b6}} 
{\bfseries average\+\_\+size} = None
\item 
\mbox{\Hypertarget{class_py_c_u_b_1_1espece_1_1_espece_ac7ce087a6722a531555c005c6b37ae88}\label{class_py_c_u_b_1_1espece_1_1_espece_ac7ce087a6722a531555c005c6b37ae88}} 
{\bfseries var\+\_\+entropy} = None
\item 
\mbox{\Hypertarget{class_py_c_u_b_1_1espece_1_1_espece_a27d1e4719b819c4e99e889904fbff5ed}\label{class_py_c_u_b_1_1espece_1_1_espece_a27d1e4719b819c4e99e889904fbff5ed}} 
{\bfseries fullentropy} = None
\item 
\mbox{\Hypertarget{class_py_c_u_b_1_1espece_1_1_espece_a7c21598e68d19819e523bb08af5a815d}\label{class_py_c_u_b_1_1espece_1_1_espece_a7c21598e68d19819e523bb08af5a815d}} 
{\bfseries fullvarentropy} = None
\item 
\mbox{\Hypertarget{class_py_c_u_b_1_1espece_1_1_espece_a4c199498e2938fdd8521fb724e9f4e58}\label{class_py_c_u_b_1_1espece_1_1_espece_a4c199498e2938fdd8521fb724e9f4e58}} 
{\bfseries full\+G\+Ccount} = None
\item 
\mbox{\Hypertarget{class_py_c_u_b_1_1espece_1_1_espece_a6f99cfcc3643e72bf48a2479c81b7e03}\label{class_py_c_u_b_1_1espece_1_1_espece_a6f99cfcc3643e72bf48a2479c81b7e03}} 
{\bfseries var\+G\+Ccount} = None
\item 
\mbox{\Hypertarget{class_py_c_u_b_1_1espece_1_1_espece_a32834ad6a3a42b850438789ae1e7270b}\label{class_py_c_u_b_1_1espece_1_1_espece_a32834ad6a3a42b850438789ae1e7270b}} 
{\bfseries mean\+G\+Chomo} = None
\item 
\mbox{\Hypertarget{class_py_c_u_b_1_1espece_1_1_espece_a5fdd0450239ab1bbe67a22e51caf15c6}\label{class_py_c_u_b_1_1espece_1_1_espece_a5fdd0450239ab1bbe67a22e51caf15c6}} 
{\bfseries t\+R\+N\+Aentropy} = None
\item 
\mbox{\Hypertarget{class_py_c_u_b_1_1espece_1_1_espece_a283d26f1b90fe5a4afb4cb4d950b946f}\label{class_py_c_u_b_1_1espece_1_1_espece_a283d26f1b90fe5a4afb4cb4d950b946f}} 
{\bfseries tot\+\_\+homologies} = None
\item 
\mbox{\Hypertarget{class_py_c_u_b_1_1espece_1_1_espece_af1c5763db7472083035dd6ccb6904a0c}\label{class_py_c_u_b_1_1espece_1_1_espece_af1c5763db7472083035dd6ccb6904a0c}} 
{\bfseries meanecai} = None
\end{DoxyCompactItemize}
\subsection*{Private Member Functions}
\begin{DoxyCompactItemize}
\item 
def \mbox{\hyperlink{class_py_c_u_b_1_1espece_1_1_espece_a4edfabc363a5d4be437afb94f5d03368}{\+\_\+dictify}} (self)
\begin{DoxyCompactList}\small\item\em Used by the saving function. \end{DoxyCompactList}\end{DoxyCompactItemize}


\subsection{Detailed Description}
docstring for \mbox{\hyperlink{class_py_c_u_b_1_1espece_1_1_espece}{Espece}} 

This is an object that contains all required information of a species for Py\+C\+UB and some nice functions to interact for each species


\begin{DoxyParams}{Parameters}
{\em code} & a dict from gene\+\_\+name to dna\+\_\+seq string (deprecated) \\
\hline
{\em metadata} & a dict containing different metadata information that one can gather, preferentially boolean flags to classify the species for plottings and comparings \\
\hline
{\em name} & the full scientific name of the species is\+\_\+stored, state if the data is stored in HD (deprecated) \\
\hline
{\em link} & the link to the ensembl genome \\
\hline
{\em num\+\_\+genes} & the number of gene of this species \\
\hline
{\em genome\+\_\+size} & the bp size of the coding genome \\
\hline
{\em name} & the name of the species \\
\hline
{\em taxonid} & the number assoxiated to this taxon \\
\hline
{\em copynumbers} & the approx. copynumbers if any of each t\+R\+NA known of this species \\
\hline
{\em average\+\_\+entropy} & the mean C\+UB value for each amino acids (C\+U\+BD dimension) \\
\hline
{\em average\+\_\+size} & the mean size of each homologies \\
\hline
{\em var\+\_\+entropy} & the mean of fullvarentropy \\
\hline
{\em fullentropy} & the array containing all C\+UB values from the homologies of this species \\
\hline
{\em fullvarentropy} & the variance for each amino acids of the full C\+UB values of this species \\
\hline
{\em full\+G\+Ccount} & the GC content of the full coding genome \\
\hline
{\em var\+G\+Ccount} & the variance of the GC content of the full coding genome \\
\hline
{\em t\+R\+N\+Aentropy} & the entropy values of the copynumbers of the t\+R\+N\+As if sufficient t\+R\+N\+As exist \\
\hline
{\em tot\+\_\+homologies} & the total number of homologies to cerevisiae \\
\hline
\end{DoxyParams}


\subsection{Member Function Documentation}
\mbox{\Hypertarget{class_py_c_u_b_1_1espece_1_1_espece_a4edfabc363a5d4be437afb94f5d03368}\label{class_py_c_u_b_1_1espece_1_1_espece_a4edfabc363a5d4be437afb94f5d03368}} 
\index{Py\+C\+U\+B\+::espece\+::\+Espece@{Py\+C\+U\+B\+::espece\+::\+Espece}!\+\_\+dictify@{\+\_\+dictify}}
\index{\+\_\+dictify@{\+\_\+dictify}!Py\+C\+U\+B\+::espece\+::\+Espece@{Py\+C\+U\+B\+::espece\+::\+Espece}}
\subsubsection{\texorpdfstring{\+\_\+dictify()}{\_dictify()}}
{\footnotesize\ttfamily def Py\+C\+U\+B.\+espece.\+Espece.\+\_\+dictify (\begin{DoxyParamCaption}\item[{}]{self }\end{DoxyParamCaption})\hspace{0.3cm}{\ttfamily [private]}}



Used by the saving function. 

transform the object into a dictionary that can be json serializable

\begin{DoxyReturn}{Returns}
A dict holding every element to be jsonized 
\end{DoxyReturn}


Referenced by Py\+C\+U\+B.\+py\+C\+U\+B.\+Py\+C\+U\+B.\+save().

\mbox{\Hypertarget{class_py_c_u_b_1_1espece_1_1_espece_afa35655bb969cbc02456aa74228a1b8d}\label{class_py_c_u_b_1_1espece_1_1_espece_afa35655bb969cbc02456aa74228a1b8d}} 
\index{Py\+C\+U\+B\+::espece\+::\+Espece@{Py\+C\+U\+B\+::espece\+::\+Espece}!get\+\_\+t\+R\+N\+Acopy@{get\+\_\+t\+R\+N\+Acopy}}
\index{get\+\_\+t\+R\+N\+Acopy@{get\+\_\+t\+R\+N\+Acopy}!Py\+C\+U\+B\+::espece\+::\+Espece@{Py\+C\+U\+B\+::espece\+::\+Espece}}
\subsubsection{\texorpdfstring{get\+\_\+t\+R\+N\+Acopy()}{get\_tRNAcopy()}}
{\footnotesize\ttfamily def Py\+C\+U\+B.\+espece.\+Espece.\+get\+\_\+t\+R\+N\+Acopy (\begin{DoxyParamCaption}\item[{}]{self,  }\item[{}]{by = {\ttfamily \char`\"{}entropy\char`\"{}},  }\item[{}]{setnans = {\ttfamily False},  }\item[{}]{kingdom = {\ttfamily \textquotesingle{}fungi\textquotesingle{}},  }\item[{}]{base\+C\+Nvalue = {\ttfamily 2} }\end{DoxyParamCaption})}



Retrieves t\+R\+NA copy numbers from ensembl DB. 

will print the number of t\+R\+N\+As and the number of t\+R\+N\+As with a knwon codons ( the usefull ones) will stop and set a trace for the user to inspect the data to do so\+: please write \char`\"{}dat\char`\"{} in the console. if you see something that should be corrected please do so from the console directly or from the code if there seems to be an error in the code if it is an error in the db that you can\textquotesingle{}t do anything, like a mismatched codon and amino acid, you can\textquotesingle{}t do much. resume the process by typing \char`\"{}c\char`\"{} in the console.


\begin{DoxyParams}{Parameters}
{\em species} & string, the species from which you want the Trna copy number\\
\hline
\end{DoxyParams}
\begin{DoxyReturn}{Returns}
Will populate copynumbers. And t\+R\+N\+Aentropy if by=\char`\"{}entropy\char`\"{} Or will not do anything if the species is unavailable and will print it
\end{DoxyReturn}

\begin{DoxyExceptions}{Exceptions}
{\em Attribute\+Error} & this is a wrong argument try frequency or entropy \\
\hline
\end{DoxyExceptions}


Referenced by Py\+C\+U\+B.\+espece.\+Espece.\+plot().

\mbox{\Hypertarget{class_py_c_u_b_1_1espece_1_1_espece_af716fcd7e7e34e12cd7786469d695422}\label{class_py_c_u_b_1_1espece_1_1_espece_af716fcd7e7e34e12cd7786469d695422}} 
\index{Py\+C\+U\+B\+::espece\+::\+Espece@{Py\+C\+U\+B\+::espece\+::\+Espece}!gettaxons@{gettaxons}}
\index{gettaxons@{gettaxons}!Py\+C\+U\+B\+::espece\+::\+Espece@{Py\+C\+U\+B\+::espece\+::\+Espece}}
\subsubsection{\texorpdfstring{gettaxons()}{gettaxons()}}
{\footnotesize\ttfamily def Py\+C\+U\+B.\+espece.\+Espece.\+gettaxons (\begin{DoxyParamCaption}\item[{}]{self,  }\item[{}]{kingdom = {\ttfamily \textquotesingle{}fungi\textquotesingle{}} }\end{DoxyParamCaption})}



Pars the ensemblgenomes R\+E\+ST A\+PI to retrieve the taxons id. 

for the species from which we would not have any (downloaded via Yun for example)


\begin{DoxyExceptions}{Exceptions}
{\em H\+T\+T\+Prequest\+Error} & not able to connect to the server \\
\hline
\end{DoxyExceptions}


Referenced by Py\+C\+U\+B.\+espece.\+Espece.\+get\+\_\+t\+R\+N\+Acopy().



\subsection{Member Data Documentation}
\mbox{\Hypertarget{class_py_c_u_b_1_1espece_1_1_espece_a0a27801c75339436b18b7e865b890a00}\label{class_py_c_u_b_1_1espece_1_1_espece_a0a27801c75339436b18b7e865b890a00}} 
\index{Py\+C\+U\+B\+::espece\+::\+Espece@{Py\+C\+U\+B\+::espece\+::\+Espece}!metadata@{metadata}}
\index{metadata@{metadata}!Py\+C\+U\+B\+::espece\+::\+Espece@{Py\+C\+U\+B\+::espece\+::\+Espece}}
\subsubsection{\texorpdfstring{metadata}{metadata}}
{\footnotesize\ttfamily dictionary Py\+C\+U\+B.\+espece.\+Espece.\+metadata\hspace{0.3cm}{\ttfamily [static]}}

{\bfseries Initial value\+:}
\begin{DoxyCode}
=  \{
        \textcolor{stringliteral}{"isplant\_pathogen"}: \textcolor{keyword}{False},
        \textcolor{stringliteral}{"isanimal\_pathogen"}: \textcolor{keyword}{False},
        \textcolor{stringliteral}{"isplant\_symbiotic"}: \textcolor{keyword}{False},  \textcolor{comment}{# endophyte or mycorrhizal}
        \textcolor{stringliteral}{"isbrown\_rot"}: \textcolor{keyword}{False},
        \textcolor{stringliteral}{"iswhite\_rot"}: \textcolor{keyword}{False}
    \}
\end{DoxyCode}


Referenced by Py\+C\+U\+B.\+espece.\+Espece.\+\_\+\+\_\+init\+\_\+\+\_\+(), Py\+C\+U\+B.\+espece.\+Espece.\+\_\+dictify(), and Py\+C\+U\+B.\+espece.\+Espece.\+plot().



The documentation for this class was generated from the following file\+:\begin{DoxyCompactItemize}
\item 
Py\+C\+U\+B/espece.\+py\end{DoxyCompactItemize}
