<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.14"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>PyCUB: Machine exploration of the Codon Usage Bias: PyCUB.espece.Espece Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
  $(document).ready(initResizable);
/* @license-end */</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    extensions: ["tex2jax.js"],
    jax: ["input/TeX","output/HTML-CSS"],
});
</script><script type="text/javascript" async src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.2/MathJax.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">PyCUB: Machine exploration of the Codon Usage Bias
   </div>
   <div id="projectbrief">A set of tools to explore the codon usage bias</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.14 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(document).ready(function(){initNavTree('class_py_c_u_b_1_1espece_1_1_espece.html','');});
/* @license-end */
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-attribs">Public Attributes</a> &#124;
<a href="#pub-static-attribs">Static Public Attributes</a> &#124;
<a href="#pri-methods">Private Member Functions</a> &#124;
<a href="class_py_c_u_b_1_1espece_1_1_espece-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">PyCUB.espece.Espece Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p>docstring for <a class="el" href="class_py_c_u_b_1_1espece_1_1_espece.html" title="docstring for Espece ">Espece</a>  
 <a href="class_py_c_u_b_1_1espece_1_1_espece.html#details">More...</a></p>
<div class="dynheader">
Inheritance diagram for PyCUB.espece.Espece:</div>
<div class="dyncontent">
 <div class="center">
  <img src="class_py_c_u_b_1_1espece_1_1_espece.png" usemap="#PyCUB.espece.Espece_map" alt=""/>
  <map id="PyCUB.espece.Espece_map" name="PyCUB.espece.Espece_map">
</map>
 </div></div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a13a7403216d2c03d6e9ec81d6034196e"><td class="memItemLeft" align="right" valign="top"><a id="a13a7403216d2c03d6e9ec81d6034196e"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1espece_1_1_espece.html#a13a7403216d2c03d6e9ec81d6034196e">__init__</a> (self, kwargs)</td></tr>
<tr class="memdesc:a13a7403216d2c03d6e9ec81d6034196e"><td class="mdescLeft">&#160;</td><td class="mdescRight">can intialize the file from kwargs as a raw dictionnary for json format (output of dictify) or from regular args. <br /></td></tr>
<tr class="separator:a13a7403216d2c03d6e9ec81d6034196e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4222c359e63e76ea34631a427b62fd60"><td class="memItemLeft" align="right" valign="top"><a id="a4222c359e63e76ea34631a427b62fd60"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1espece_1_1_espece.html#a4222c359e63e76ea34631a427b62fd60">__str__</a> (self)</td></tr>
<tr class="memdesc:a4222c359e63e76ea34631a427b62fd60"><td class="mdescLeft">&#160;</td><td class="mdescRight">will present some interesting info about this species. <br /></td></tr>
<tr class="separator:a4222c359e63e76ea34631a427b62fd60"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afa35655bb969cbc02456aa74228a1b8d"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1espece_1_1_espece.html#afa35655bb969cbc02456aa74228a1b8d">get_tRNAcopy</a> (self, by=&quot;entropy&quot;, setnans=False, kingdom='fungi', baseCNvalue=2)</td></tr>
<tr class="memdesc:afa35655bb969cbc02456aa74228a1b8d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Retrieves tRNA copy numbers from ensembl DB.  <a href="#afa35655bb969cbc02456aa74228a1b8d">More...</a><br /></td></tr>
<tr class="separator:afa35655bb969cbc02456aa74228a1b8d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af716fcd7e7e34e12cd7786469d695422"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1espece_1_1_espece.html#af716fcd7e7e34e12cd7786469d695422">gettaxons</a> (self, kingdom='fungi')</td></tr>
<tr class="memdesc:af716fcd7e7e34e12cd7786469d695422"><td class="mdescLeft">&#160;</td><td class="mdescRight">Pars the ensemblgenomes REST API to retrieve the taxons id.  <a href="#af716fcd7e7e34e12cd7786469d695422">More...</a><br /></td></tr>
<tr class="separator:af716fcd7e7e34e12cd7786469d695422"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aed93f9be40c4521c6aed536125268432"><td class="memItemLeft" align="right" valign="top"><a id="aed93f9be40c4521c6aed536125268432"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1espece_1_1_espece.html#aed93f9be40c4521c6aed536125268432">get_epigenomes</a> (self)</td></tr>
<tr class="memdesc:aed93f9be40c4521c6aed536125268432"><td class="mdescLeft">&#160;</td><td class="mdescRight">get from ensembl all the data about the epigenome that could help asking interesting questions about the CUB <br /></td></tr>
<tr class="separator:aed93f9be40c4521c6aed536125268432"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-attribs"></a>
Public Attributes</h2></td></tr>
<tr class="memitem:a3a33da4c7b0c9163b8673166b306226c"><td class="memItemLeft" align="right" valign="top"><a id="a3a33da4c7b0c9163b8673166b306226c"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>name</b></td></tr>
<tr class="separator:a3a33da4c7b0c9163b8673166b306226c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae98c1d182e761e4c09a10f8a9994124e"><td class="memItemLeft" align="right" valign="top"><a id="ae98c1d182e761e4c09a10f8a9994124e"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>metadata</b></td></tr>
<tr class="separator:ae98c1d182e761e4c09a10f8a9994124e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afbc7dbe8f3dc1f035837bb2ed610f246"><td class="memItemLeft" align="right" valign="top"><a id="afbc7dbe8f3dc1f035837bb2ed610f246"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>is_stored</b></td></tr>
<tr class="separator:afbc7dbe8f3dc1f035837bb2ed610f246"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adb2006eee970b43f6abe9554690cb424"><td class="memItemLeft" align="right" valign="top"><a id="adb2006eee970b43f6abe9554690cb424"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>num_genes</b></td></tr>
<tr class="separator:adb2006eee970b43f6abe9554690cb424"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad40242dc0031d44de4c363bd716489e2"><td class="memItemLeft" align="right" valign="top"><a id="ad40242dc0031d44de4c363bd716489e2"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>genome_size</b></td></tr>
<tr class="separator:ad40242dc0031d44de4c363bd716489e2"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-static-attribs"></a>
Static Public Attributes</h2></td></tr>
<tr class="memitem:a143141fe25ea32dc24abd314a0f5fdbd"><td class="memItemLeft" align="right" valign="top"><a id="a143141fe25ea32dc24abd314a0f5fdbd"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>code</b> = None</td></tr>
<tr class="separator:a143141fe25ea32dc24abd314a0f5fdbd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac3d8e577f2e9a0fe6853d51c4bec238d"><td class="memItemLeft" align="right" valign="top"><a id="ac3d8e577f2e9a0fe6853d51c4bec238d"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>num_genes</b> = 0</td></tr>
<tr class="separator:ac3d8e577f2e9a0fe6853d51c4bec238d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a87ca67e0c2e560e88b72a8bdba456756"><td class="memItemLeft" align="right" valign="top"><a id="a87ca67e0c2e560e88b72a8bdba456756"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>genome_size</b> = 0</td></tr>
<tr class="separator:a87ca67e0c2e560e88b72a8bdba456756"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae066a956b3f9f29d676162fbf9773a0f"><td class="memItemLeft" align="right" valign="top"><a id="ae066a956b3f9f29d676162fbf9773a0f"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>link</b> = None</td></tr>
<tr class="separator:ae066a956b3f9f29d676162fbf9773a0f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0a27801c75339436b18b7e865b890a00"><td class="memItemLeft" align="right" valign="top">dictionary&#160;</td><td class="memItemRight" valign="bottom"><b>metadata</b></td></tr>
<tr class="separator:a0a27801c75339436b18b7e865b890a00"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8befee83c9b6ecfc8b07e66b6251afa4"><td class="memItemLeft" align="right" valign="top"><a id="a8befee83c9b6ecfc8b07e66b6251afa4"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><b>is_stored</b> = False</td></tr>
<tr class="separator:a8befee83c9b6ecfc8b07e66b6251afa4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:accf54132a0346377a2cf7a461a129dc1"><td class="memItemLeft" align="right" valign="top"><a id="accf54132a0346377a2cf7a461a129dc1"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>name</b> = ''</td></tr>
<tr class="separator:accf54132a0346377a2cf7a461a129dc1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aae7cf0e3c6f359bd2970006817e80ed3"><td class="memItemLeft" align="right" valign="top"><a id="aae7cf0e3c6f359bd2970006817e80ed3"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>taxonid</b> = None</td></tr>
<tr class="separator:aae7cf0e3c6f359bd2970006817e80ed3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a89cd4e211bf028a4d141202e8a885b58"><td class="memItemLeft" align="right" valign="top"><a id="a89cd4e211bf028a4d141202e8a885b58"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>copynumbers</b> = None</td></tr>
<tr class="separator:a89cd4e211bf028a4d141202e8a885b58"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2dc4b3f0d2006aac69a9846ba834d209"><td class="memItemLeft" align="right" valign="top"><a id="a2dc4b3f0d2006aac69a9846ba834d209"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>average_entropy</b> = None</td></tr>
<tr class="separator:a2dc4b3f0d2006aac69a9846ba834d209"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2991e13a2c75e2e899f3f8f773d1d9b6"><td class="memItemLeft" align="right" valign="top"><a id="a2991e13a2c75e2e899f3f8f773d1d9b6"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>average_size</b> = None</td></tr>
<tr class="separator:a2991e13a2c75e2e899f3f8f773d1d9b6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac7ce087a6722a531555c005c6b37ae88"><td class="memItemLeft" align="right" valign="top"><a id="ac7ce087a6722a531555c005c6b37ae88"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>var_entropy</b> = None</td></tr>
<tr class="separator:ac7ce087a6722a531555c005c6b37ae88"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a27d1e4719b819c4e99e889904fbff5ed"><td class="memItemLeft" align="right" valign="top"><a id="a27d1e4719b819c4e99e889904fbff5ed"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>fullentropy</b> = None</td></tr>
<tr class="separator:a27d1e4719b819c4e99e889904fbff5ed"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7c21598e68d19819e523bb08af5a815d"><td class="memItemLeft" align="right" valign="top"><a id="a7c21598e68d19819e523bb08af5a815d"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>fullvarentropy</b> = None</td></tr>
<tr class="separator:a7c21598e68d19819e523bb08af5a815d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4c199498e2938fdd8521fb724e9f4e58"><td class="memItemLeft" align="right" valign="top"><a id="a4c199498e2938fdd8521fb724e9f4e58"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>fullGCcount</b> = None</td></tr>
<tr class="separator:a4c199498e2938fdd8521fb724e9f4e58"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6f99cfcc3643e72bf48a2479c81b7e03"><td class="memItemLeft" align="right" valign="top"><a id="a6f99cfcc3643e72bf48a2479c81b7e03"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>varGCcount</b> = None</td></tr>
<tr class="separator:a6f99cfcc3643e72bf48a2479c81b7e03"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a32834ad6a3a42b850438789ae1e7270b"><td class="memItemLeft" align="right" valign="top"><a id="a32834ad6a3a42b850438789ae1e7270b"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>meanGChomo</b> = None</td></tr>
<tr class="separator:a32834ad6a3a42b850438789ae1e7270b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5fdd0450239ab1bbe67a22e51caf15c6"><td class="memItemLeft" align="right" valign="top"><a id="a5fdd0450239ab1bbe67a22e51caf15c6"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>tRNAentropy</b> = None</td></tr>
<tr class="separator:a5fdd0450239ab1bbe67a22e51caf15c6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a283d26f1b90fe5a4afb4cb4d950b946f"><td class="memItemLeft" align="right" valign="top"><a id="a283d26f1b90fe5a4afb4cb4d950b946f"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>tot_homologies</b> = None</td></tr>
<tr class="separator:a283d26f1b90fe5a4afb4cb4d950b946f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af1c5763db7472083035dd6ccb6904a0c"><td class="memItemLeft" align="right" valign="top"><a id="af1c5763db7472083035dd6ccb6904a0c"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>meanecai</b> = None</td></tr>
<tr class="separator:af1c5763db7472083035dd6ccb6904a0c"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pri-methods"></a>
Private Member Functions</h2></td></tr>
<tr class="memitem:a4edfabc363a5d4be437afb94f5d03368"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1espece_1_1_espece.html#a4edfabc363a5d4be437afb94f5d03368">_dictify</a> (self)</td></tr>
<tr class="memdesc:a4edfabc363a5d4be437afb94f5d03368"><td class="mdescLeft">&#160;</td><td class="mdescRight">Used by the saving function.  <a href="#a4edfabc363a5d4be437afb94f5d03368">More...</a><br /></td></tr>
<tr class="separator:a4edfabc363a5d4be437afb94f5d03368"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>docstring for <a class="el" href="class_py_c_u_b_1_1espece_1_1_espece.html" title="docstring for Espece ">Espece</a> </p>
<p>This is an object that contains all required information of a species for PyCUB and some nice functions to interact for each species</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">code</td><td>a dict from gene_name to dna_seq string (deprecated) </td></tr>
    <tr><td class="paramname">metadata</td><td>a dict containing different metadata information that one can gather, preferentially boolean flags to classify the species for plottings and comparings </td></tr>
    <tr><td class="paramname">name</td><td>the full scientific name of the species is_stored, state if the data is stored in HD (deprecated) </td></tr>
    <tr><td class="paramname">link</td><td>the link to the ensembl genome </td></tr>
    <tr><td class="paramname">num_genes</td><td>the number of gene of this species </td></tr>
    <tr><td class="paramname">genome_size</td><td>the bp size of the coding genome </td></tr>
    <tr><td class="paramname">name</td><td>the name of the species </td></tr>
    <tr><td class="paramname">taxonid</td><td>the number assoxiated to this taxon </td></tr>
    <tr><td class="paramname">copynumbers</td><td>the approx. copynumbers if any of each tRNA known of this species </td></tr>
    <tr><td class="paramname">average_entropy</td><td>the mean CUB value for each amino acids (CUBD dimension) </td></tr>
    <tr><td class="paramname">average_size</td><td>the mean size of each homologies </td></tr>
    <tr><td class="paramname">var_entropy</td><td>the mean of fullvarentropy </td></tr>
    <tr><td class="paramname">fullentropy</td><td>the array containing all CUB values from the homologies of this species </td></tr>
    <tr><td class="paramname">fullvarentropy</td><td>the variance for each amino acids of the full CUB values of this species </td></tr>
    <tr><td class="paramname">fullGCcount</td><td>the GC content of the full coding genome </td></tr>
    <tr><td class="paramname">varGCcount</td><td>the variance of the GC content of the full coding genome </td></tr>
    <tr><td class="paramname">tRNAentropy</td><td>the entropy values of the copynumbers of the tRNAs if sufficient tRNAs exist </td></tr>
    <tr><td class="paramname">tot_homologies</td><td>the total number of homologies to cerevisiae </td></tr>
  </table>
  </dd>
</dl>
</div><h2 class="groupheader">Member Function Documentation</h2>
<a id="a4edfabc363a5d4be437afb94f5d03368"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a4edfabc363a5d4be437afb94f5d03368">&#9670;&nbsp;</a></span>_dictify()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.espece.Espece._dictify </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Used by the saving function. </p>
<p>transform the object into a dictionary that can be json serializable</p>
<dl class="section return"><dt>Returns</dt><dd>A dict holding every element to be jsonized </dd></dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a477c55065989b10bce8389c5edaf3322">PyCUB.pyCUB.PyCUB.save()</a>.</p>

</div>
</div>
<a id="afa35655bb969cbc02456aa74228a1b8d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#afa35655bb969cbc02456aa74228a1b8d">&#9670;&nbsp;</a></span>get_tRNAcopy()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.espece.Espece.get_tRNAcopy </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>by</em> = <code>&quot;entropy&quot;</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>setnans</em> = <code>False</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>kingdom</em> = <code>'fungi'</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>baseCNvalue</em> = <code>2</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Retrieves tRNA copy numbers from ensembl DB. </p>
<p>will print the number of tRNAs and the number of tRNAs with a knwon codons ( the usefull ones) will stop and set a trace for the user to inspect the data to do so: please write "dat" in the console. if you see something that should be corrected please do so from the console directly or from the code if there seems to be an error in the code if it is an error in the db that you can't do anything, like a mismatched codon and amino acid, you can't do much. resume the process by typing "c" in the console.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">species</td><td>string, the species from which you want the Trna copy number</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Will populate copynumbers. And tRNAentropy if by="entropy" Or will not do anything if the species is unavailable and will print it</dd></dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">AttributeError</td><td>this is a wrong argument try frequency or entropy </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1espece_1_1_espece.html#a4222c359e63e76ea34631a427b62fd60">PyCUB.espece.Espece.__str__()</a>.</p>

</div>
</div>
<a id="af716fcd7e7e34e12cd7786469d695422"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af716fcd7e7e34e12cd7786469d695422">&#9670;&nbsp;</a></span>gettaxons()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.espece.Espece.gettaxons </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>kingdom</em> = <code>'fungi'</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Pars the ensemblgenomes REST API to retrieve the taxons id. </p>
<p>for the species from which we would not have any (downloaded via Yun for example)</p>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">HTTPrequestError</td><td>not able to connect to the server </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1espece_1_1_espece.html#afa35655bb969cbc02456aa74228a1b8d">PyCUB.espece.Espece.get_tRNAcopy()</a>.</p>

</div>
</div>
<h2 class="groupheader">Member Data Documentation</h2>
<a id="a0a27801c75339436b18b7e865b890a00"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a0a27801c75339436b18b7e865b890a00">&#9670;&nbsp;</a></span>metadata</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">dictionary PyCUB.espece.Espece.metadata</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<b>Initial value:</b><div class="fragment"><div class="line">=  {</div><div class="line">        <span class="stringliteral">&quot;isplant_pathogen&quot;</span>: <span class="keyword">False</span>,</div><div class="line">        <span class="stringliteral">&quot;isanimal_pathogen&quot;</span>: <span class="keyword">False</span>,</div><div class="line">        <span class="stringliteral">&quot;isplant_symbiotic&quot;</span>: <span class="keyword">False</span>,  <span class="comment"># endophyte or mycorrhizal</span></div><div class="line">        <span class="stringliteral">&quot;isbrown_rot&quot;</span>: <span class="keyword">False</span>,</div><div class="line">        <span class="stringliteral">&quot;iswhite_rot&quot;</span>: <span class="keyword">False</span></div><div class="line">    }</div></div><!-- fragment -->
<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1espece_1_1_espece.html#a13a7403216d2c03d6e9ec81d6034196e">PyCUB.espece.Espece.__init__()</a>, <a class="el" href="class_py_c_u_b_1_1espece_1_1_espece.html#a4222c359e63e76ea34631a427b62fd60">PyCUB.espece.Espece.__str__()</a>, and <a class="el" href="class_py_c_u_b_1_1espece_1_1_espece.html#a4edfabc363a5d4be437afb94f5d03368">PyCUB.espece.Espece._dictify()</a>.</p>

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>PyCUB/espece.py</li>
</ul>
</div><!-- contents -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="navelem"><b>PyCUB</b></li><li class="navelem"><b>espece</b></li><li class="navelem"><a class="el" href="class_py_c_u_b_1_1espece_1_1_espece.html">Espece</a></li>
    <li class="footer">Generated by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.8.14 </li>
  </ul>
</div>
</body>
</html>
