<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.14"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>PyCUB: Machine exploration of the Codon Usage Bias: PyCUB.pyCUB.PyCUB Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
  $(document).ready(initResizable);
/* @license-end */</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    extensions: ["tex2jax.js"],
    jax: ["input/TeX","output/HTML-CSS"],
});
</script><script type="text/javascript" async src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.2/MathJax.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">PyCUB: Machine exploration of the Codon Usage Bias
   </div>
   <div id="projectbrief">A set of tools to explore the codon usage bias</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.14 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(document).ready(function(){initNavTree('class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html','');});
/* @license-end */
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-attribs">Public Attributes</a> &#124;
<a href="#pub-static-attribs">Static Public Attributes</a> &#124;
<a href="#pri-methods">Private Member Functions</a> &#124;
<a href="#pri-attribs">Private Attributes</a> &#124;
<a href="#pri-static-attribs">Static Private Attributes</a> &#124;
<a href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">PyCUB.pyCUB.PyCUB Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p><a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html" title="PyCUB is the main object of the project that allows the user to access most of the functions...">PyCUB</a> is the main object of the project that allows the user to access most of the functions.  
 <a href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#details">More...</a></p>
<div class="dynheader">
Inheritance diagram for PyCUB.pyCUB.PyCUB:</div>
<div class="dyncontent">
 <div class="center">
  <img src="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.png" usemap="#PyCUB.pyCUB.PyCUB_map" alt=""/>
  <map id="PyCUB.pyCUB.PyCUB_map" name="PyCUB.pyCUB.PyCUB_map">
</map>
 </div></div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a0a8de5f32a03de2c681c0582d9998f03"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a0a8de5f32a03de2c681c0582d9998f03">__init__</a> (self, species={}, _is_saved=False, _is_loaded=False, working_homoset=None, all_homoset=None, session='session1')</td></tr>
<tr class="memdesc:a0a8de5f32a03de2c681c0582d9998f03"><td class="mdescLeft">&#160;</td><td class="mdescRight">will initialize the object with the different values you might have from another project  <a href="#a0a8de5f32a03de2c681c0582d9998f03">More...</a><br /></td></tr>
<tr class="separator:a0a8de5f32a03de2c681c0582d9998f03"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae06a2843c0718f86396ee255dfc6ae6f"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#ae06a2843c0718f86396ee255dfc6ae6f">getHomologylist</a> (self, species='saccharomyces_cerevisiae', kingdom='fungi')</td></tr>
<tr class="memdesc:ae06a2843c0718f86396ee255dfc6ae6f"><td class="mdescLeft">&#160;</td><td class="mdescRight">A function to retrieve the homologies directly from a given species.  <a href="#ae06a2843c0718f86396ee255dfc6ae6f">More...</a><br /></td></tr>
<tr class="separator:ae06a2843c0718f86396ee255dfc6ae6f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac4d231f47432a18f84da1cef8673b50b"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#ac4d231f47432a18f84da1cef8673b50b">get_data</a> (self, From='yun', homonames=None, kingdom='fungi', sequence='cdna', additional='type=orthologues', saveonfiles=False, normalized=True, setnans=False, by=&quot;entropy&quot;, using=&quot;normal&quot;, tRNA=True, getCAI=True, first=20, inpar=True)</td></tr>
<tr class="memdesc:ac4d231f47432a18f84da1cef8673b50b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Download the data from somewhere on the web (Ensembl, Yun(with links))  <a href="#ac4d231f47432a18f84da1cef8673b50b">More...</a><br /></td></tr>
<tr class="separator:ac4d231f47432a18f84da1cef8673b50b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae561ba39b9c3919a83b51944f155e4a4"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#ae561ba39b9c3919a83b51944f155e4a4">get_metadata_Ensembl</a> (self, kingdoms)</td></tr>
<tr class="memdesc:ae561ba39b9c3919a83b51944f155e4a4"><td class="mdescLeft">&#160;</td><td class="mdescRight">download it and put it where it belongs in the Espece object  <a href="#ae561ba39b9c3919a83b51944f155e4a4">More...</a><br /></td></tr>
<tr class="separator:ae561ba39b9c3919a83b51944f155e4a4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1ff011ae5b64518d60bedcf1b98314ef"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a1ff011ae5b64518d60bedcf1b98314ef">get_mymetadata</a> (self, From='jerem', inpar=True)</td></tr>
<tr class="memdesc:a1ff011ae5b64518d60bedcf1b98314ef"><td class="mdescLeft">&#160;</td><td class="mdescRight">Go ahead and design your own metadata retrieval here.  <a href="#a1ff011ae5b64518d60bedcf1b98314ef">More...</a><br /></td></tr>
<tr class="separator:a1ff011ae5b64518d60bedcf1b98314ef"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:acd80f084f6671584c42542813874cd02"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#acd80f084f6671584c42542813874cd02">import_metadataTobias</a> (self)</td></tr>
<tr class="memdesc:acd80f084f6671584c42542813874cd02"><td class="mdescLeft">&#160;</td><td class="mdescRight">will import the metadata obtained from tobias for the fungi species affiliated to cerevisiae to each species for further diagnostics.  <a href="#acd80f084f6671584c42542813874cd02">More...</a><br /></td></tr>
<tr class="separator:acd80f084f6671584c42542813874cd02"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab471e823b0eba23af5a5322008ec09da"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#ab471e823b0eba23af5a5322008ec09da">load</a> (self, session=None, All=False, filename='first500', From=None, by='entropy', tRNA=True, inpar=True)</td></tr>
<tr class="memdesc:ab471e823b0eba23af5a5322008ec09da"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the data that is already present on a filename.  <a href="#ab471e823b0eba23af5a5322008ec09da">More...</a><br /></td></tr>
<tr class="separator:ab471e823b0eba23af5a5322008ec09da"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a16a1ab6be5dddcedf94852088936c1f2"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a16a1ab6be5dddcedf94852088936c1f2">loadmore</a> (self, filename='first500', by='entropyLocation')</td></tr>
<tr class="memdesc:a16a1ab6be5dddcedf94852088936c1f2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the data that is already present on a filename when you already have data.  <a href="#a16a1ab6be5dddcedf94852088936c1f2">More...</a><br /></td></tr>
<tr class="separator:a16a1ab6be5dddcedf94852088936c1f2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a477c55065989b10bce8389c5edaf3322"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a477c55065989b10bce8389c5edaf3322">save</a> (self, name, save_workspace=True, save_homo=True, add_homosets={}, cmdlinetozip=&quot;gzip&quot;)</td></tr>
<tr class="memdesc:a477c55065989b10bce8389c5edaf3322"><td class="mdescLeft">&#160;</td><td class="mdescRight">call to save your work.  <a href="#a477c55065989b10bce8389c5edaf3322">More...</a><br /></td></tr>
<tr class="separator:a477c55065989b10bce8389c5edaf3322"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1916ef696d74af1028a2aeff643727db"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a1916ef696d74af1028a2aeff643727db">get_working_homoset</a> (self, clusternb=None, species=None, homologies=None, cleanhomo=None)</td></tr>
<tr class="memdesc:a1916ef696d74af1028a2aeff643727db"><td class="mdescLeft">&#160;</td><td class="mdescRight">create a subset of all_homoset on which you would like to do further computation  <a href="#a1916ef696d74af1028a2aeff643727db">More...</a><br /></td></tr>
<tr class="separator:a1916ef696d74af1028a2aeff643727db"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9fab964d44e03b46d330b3754f5c39fb"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a9fab964d44e03b46d330b3754f5c39fb">get_subset</a> (self, homoset, withcopy=False, clusternb=None, species=None, homologies=None)</td></tr>
<tr class="memdesc:a9fab964d44e03b46d330b3754f5c39fb"><td class="mdescLeft">&#160;</td><td class="mdescRight">either changes or returns a subset of the provided homoset  <a href="#a9fab964d44e03b46d330b3754f5c39fb">More...</a><br /></td></tr>
<tr class="separator:a9fab964d44e03b46d330b3754f5c39fb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a53b60adfafbe019421d010c5fed04f5c"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a53b60adfafbe019421d010c5fed04f5c">get_full_genomes</a> (self, kingdom='fungi', seq='cds', avg=True, by=&quot;entropy&quot;, normalized=False)</td></tr>
<tr class="memdesc:a53b60adfafbe019421d010c5fed04f5c"><td class="mdescLeft">&#160;</td><td class="mdescRight">go trought all full genome fasta files in the ftp server of ensemblgenomes and  <a href="#a53b60adfafbe019421d010c5fed04f5c">More...</a><br /></td></tr>
<tr class="separator:a53b60adfafbe019421d010c5fed04f5c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4e9a2d43aacfc8d2fb1610140f2439f2"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a4e9a2d43aacfc8d2fb1610140f2439f2">get_taxons</a> (self)</td></tr>
<tr class="memdesc:a4e9a2d43aacfc8d2fb1610140f2439f2"><td class="mdescLeft">&#160;</td><td class="mdescRight">find the taxons of each referenced species (see PyCUB.Espece.gettaxons())  <a href="#a4e9a2d43aacfc8d2fb1610140f2439f2">More...</a><br /></td></tr>
<tr class="separator:a4e9a2d43aacfc8d2fb1610140f2439f2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af0d7bed125f4437bea6b8051f9922c1d"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#af0d7bed125f4437bea6b8051f9922c1d">get_evolutionary_distance</a> (self, display_tree=False, size=40)</td></tr>
<tr class="memdesc:af0d7bed125f4437bea6b8051f9922c1d"><td class="mdescLeft">&#160;</td><td class="mdescRight">uses metadata of the ancestry tree and computes a theoretical evolutionary distance matrix between each species  <a href="#af0d7bed125f4437bea6b8051f9922c1d">More...</a><br /></td></tr>
<tr class="separator:af0d7bed125f4437bea6b8051f9922c1d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a40a742a1a05fc21c7bced2c75fb93fc6"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a40a742a1a05fc21c7bced2c75fb93fc6">createRefCAI</a> (self, speciestocompare='saccharomyces_cerevisiae', kingdom='fungi', first=20)</td></tr>
<tr class="memdesc:a40a742a1a05fc21c7bced2c75fb93fc6"><td class="mdescLeft">&#160;</td><td class="mdescRight">do a compute CAI  <a href="#a40a742a1a05fc21c7bced2c75fb93fc6">More...</a><br /></td></tr>
<tr class="separator:a40a742a1a05fc21c7bced2c75fb93fc6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af3e280c265200f13946e4bdcb4cd9dd0"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#af3e280c265200f13946e4bdcb4cd9dd0">speciestable</a> (self)</td></tr>
<tr class="memdesc:af3e280c265200f13946e4bdcb4cd9dd0"><td class="mdescLeft">&#160;</td><td class="mdescRight">a copy of the utils.speciestable  <a href="#af3e280c265200f13946e4bdcb4cd9dd0">More...</a><br /></td></tr>
<tr class="separator:af3e280c265200f13946e4bdcb4cd9dd0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2ea40ac1dd5b4ad8847ec12c867d457f"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a2ea40ac1dd5b4ad8847ec12c867d457f">phylo_distances</a> (self)</td></tr>
<tr class="memdesc:a2ea40ac1dd5b4ad8847ec12c867d457f"><td class="mdescLeft">&#160;</td><td class="mdescRight">a copy of the phylodistances dataframe see (<a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#af0d7bed125f4437bea6b8051f9922c1d" title="uses metadata of the ancestry tree and computes a theoretical evolutionary distance matrix between ea...">get_evolutionary_distance()</a>)  <a href="#a2ea40ac1dd5b4ad8847ec12c867d457f">More...</a><br /></td></tr>
<tr class="separator:a2ea40ac1dd5b4ad8847ec12c867d457f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9526c0d10381daf5abc8ea120454e701"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a9526c0d10381daf5abc8ea120454e701">compute_averages</a> (self, homoset)</td></tr>
<tr class="memdesc:a9526c0d10381daf5abc8ea120454e701"><td class="mdescLeft">&#160;</td><td class="mdescRight">compute the average entropy  <a href="#a9526c0d10381daf5abc8ea120454e701">More...</a><br /></td></tr>
<tr class="separator:a9526c0d10381daf5abc8ea120454e701"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad7f24b316bc8c257719da20c0792ec01"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#ad7f24b316bc8c257719da20c0792ec01">compare_species</a> (self, showvar=True, reducer='tsne', perplexity=40, eps=0.3, size=10)</td></tr>
<tr class="memdesc:ad7f24b316bc8c257719da20c0792ec01"><td class="mdescLeft">&#160;</td><td class="mdescRight">compare the species according to their mean CUB,  <a href="#ad7f24b316bc8c257719da20c0792ec01">More...</a><br /></td></tr>
<tr class="separator:ad7f24b316bc8c257719da20c0792ec01"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af1cab3e8d69796c1a9a54a57a76228cd"><td class="memItemLeft" align="right" valign="top"><a id="af1cab3e8d69796c1a9a54a57a76228cd"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>compute_ages</b> (self, homoset, preserved=True, minpreserv=0.9, minsimi=0.85)</td></tr>
<tr class="separator:af1cab3e8d69796c1a9a54a57a76228cd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af37319783148fc8864d894b3e2f89078"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#af37319783148fc8864d894b3e2f89078">regress_on_species</a> (self, without=[&quot;&quot;], full=True, onlyhomo=False, perctrain=0.8, algo=&quot;lasso&quot;, eps=0.001, n_alphas=100)</td></tr>
<tr class="memdesc:af37319783148fc8864d894b3e2f89078"><td class="mdescLeft">&#160;</td><td class="mdescRight">Will fit a regression curve on the CUB values of the different species according to the metadatas available for each of them.  <a href="#af37319783148fc8864d894b3e2f89078">More...</a><br /></td></tr>
<tr class="separator:af37319783148fc8864d894b3e2f89078"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad350376cd68d4daacec78c694ed754c6"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#ad350376cd68d4daacec78c694ed754c6">compare_homologies</a> (self, homoset, homosapiens=False, mindistance=10, preserved=True, size=10, minpreserv=0.9, minsimi=0.9, showvar=True, eps=0.28, reducer='tsne', perplexity=40)</td></tr>
<tr class="memdesc:ad350376cd68d4daacec78c694ed754c6"><td class="mdescLeft">&#160;</td><td class="mdescRight">finds for species with a common ancester separated by a pseudo phylogenetic distance X,  <a href="#ad350376cd68d4daacec78c694ed754c6">More...</a><br /></td></tr>
<tr class="separator:ad350376cd68d4daacec78c694ed754c6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1b65efe7deb4ba5f65203c8be6fc7af2"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a1b65efe7deb4ba5f65203c8be6fc7af2">regress_on_genes</a> (self, homoset, full=True, without=['meanecai', meancai, perctrain=0.8, algo=&quot;lasso&quot;, eps=0.001, n_alphas=100)</td></tr>
<tr class="memdesc:a1b65efe7deb4ba5f65203c8be6fc7af2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Will fit a regression curve on the CUB values of the different homologies according to the metadatas available for each of them.  <a href="#a1b65efe7deb4ba5f65203c8be6fc7af2">More...</a><br /></td></tr>
<tr class="separator:a1b65efe7deb4ba5f65203c8be6fc7af2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a06d83a29ac506b4a9422ef7e64eef0b0"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a06d83a29ac506b4a9422ef7e64eef0b0">getRelation2G3DD</a> (self, species_name='saccharomyces_cerevisiae', kingdom='fungi', intrachromosome=&quot;utils/meta/3Dmodel/interactions_HindIII_fdr0.01_intra_cerevisiae.csv&quot;, interchromose=[&quot;utils/meta/3Dmodel/cerevisiae_inter1.csv&quot;, utils, meta, Dmodel, cerevisiae_inter2, csv, utils, meta, Dmodel, cerevisiae_inter3, csv, utils, meta, Dmodel, cerevisiae_inter4, csv, utils, meta, Dmodel, cerevisiae_inter5, csv, bins=2000, seq='cds', use='diament2')</td></tr>
<tr class="memdesc:a06d83a29ac506b4a9422ef7e64eef0b0"><td class="mdescLeft">&#160;</td><td class="mdescRight"><a href="https://www.nature.com/articles/ncomms6876">https://www.nature.com/articles/ncomms6876</a>  <a href="#a06d83a29ac506b4a9422ef7e64eef0b0">More...</a><br /></td></tr>
<tr class="separator:a06d83a29ac506b4a9422ef7e64eef0b0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4e13b55153bbe774e9203818f2b7f690"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a4e13b55153bbe774e9203818f2b7f690">plot_distances</a> (self, size=40)</td></tr>
<tr class="memdesc:a4e13b55153bbe774e9203818f2b7f690"><td class="mdescLeft">&#160;</td><td class="mdescRight">plot the phylogenetic distance matrix  <a href="#a4e13b55153bbe774e9203818f2b7f690">More...</a><br /></td></tr>
<tr class="separator:a4e13b55153bbe774e9203818f2b7f690"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac37364050a4ec9a7ddf3d6c1562e49a5"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#ac37364050a4ec9a7ddf3d6c1562e49a5">loadspeciestable</a> (self)</td></tr>
<tr class="memdesc:ac37364050a4ec9a7ddf3d6c1562e49a5"><td class="mdescLeft">&#160;</td><td class="mdescRight">short function to retrieve the speciestable from Disk  <a href="#ac37364050a4ec9a7ddf3d6c1562e49a5">More...</a><br /></td></tr>
<tr class="separator:ac37364050a4ec9a7ddf3d6c1562e49a5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8249834e1d2f7d061ae06ffa5a1adc32"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a8249834e1d2f7d061ae06ffa5a1adc32">savespeciestable</a> (self)</td></tr>
<tr class="memdesc:a8249834e1d2f7d061ae06ffa5a1adc32"><td class="mdescLeft">&#160;</td><td class="mdescRight">short function to put the speciestable on Disk  <a href="#a8249834e1d2f7d061ae06ffa5a1adc32">More...</a><br /></td></tr>
<tr class="separator:a8249834e1d2f7d061ae06ffa5a1adc32"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-attribs"></a>
Public Attributes</h2></td></tr>
<tr class="memitem:a60cc232d3d52f88fe8a80244b3fcddc8"><td class="memItemLeft" align="right" valign="top"><a id="a60cc232d3d52f88fe8a80244b3fcddc8"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>species</b></td></tr>
<tr class="separator:a60cc232d3d52f88fe8a80244b3fcddc8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a91b69c31f346c4983b40c1525739cf16"><td class="memItemLeft" align="right" valign="top"><a id="a91b69c31f346c4983b40c1525739cf16"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>homolist</b></td></tr>
<tr class="separator:a91b69c31f346c4983b40c1525739cf16"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab2fcbabe6d3870dbbaa0cfd11605b004"><td class="memItemLeft" align="right" valign="top"><a id="ab2fcbabe6d3870dbbaa0cfd11605b004"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>pent</b></td></tr>
<tr class="separator:ab2fcbabe6d3870dbbaa0cfd11605b004"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9f1b34d59aff65055117cde8a174a474"><td class="memItemLeft" align="right" valign="top"><a id="a9f1b34d59aff65055117cde8a174a474"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>pcub</b></td></tr>
<tr class="separator:a9f1b34d59aff65055117cde8a174a474"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0081f01dedc28de7864a2424c6760546"><td class="memItemLeft" align="right" valign="top"><a id="a0081f01dedc28de7864a2424c6760546"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>pcuf</b></td></tr>
<tr class="separator:a0081f01dedc28de7864a2424c6760546"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-static-attribs"></a>
Static Public Attributes</h2></td></tr>
<tr class="memitem:a34f579da0a139dffa0a01c3a90e9ef8d"><td class="memItemLeft" align="right" valign="top"><a id="a34f579da0a139dffa0a01c3a90e9ef8d"></a>
dictionary&#160;</td><td class="memItemRight" valign="bottom"><b>links</b></td></tr>
<tr class="separator:a34f579da0a139dffa0a01c3a90e9ef8d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a39f174e37479c070657a4e9e710a53ca"><td class="memItemLeft" align="right" valign="top"><a id="a39f174e37479c070657a4e9e710a53ca"></a>
dictionary&#160;</td><td class="memItemRight" valign="bottom"><b>species</b> = {}</td></tr>
<tr class="separator:a39f174e37479c070657a4e9e710a53ca"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad3b3799416b15243c9027230dcd68482"><td class="memItemLeft" align="right" valign="top"><a id="ad3b3799416b15243c9027230dcd68482"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>working_homoset</b> = None</td></tr>
<tr class="separator:ad3b3799416b15243c9027230dcd68482"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5e68f8b0897b4e75b69abd6f1ac1beed"><td class="memItemLeft" align="right" valign="top"><a id="a5e68f8b0897b4e75b69abd6f1ac1beed"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>all_homoset</b> = None</td></tr>
<tr class="separator:a5e68f8b0897b4e75b69abd6f1ac1beed"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a24fdc39b781fc1974625810704bd8140"><td class="memItemLeft" align="right" valign="top"><a id="a24fdc39b781fc1974625810704bd8140"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>session</b> = None</td></tr>
<tr class="separator:a24fdc39b781fc1974625810704bd8140"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abf0e792c29eb4ea90049ca996cad0022"><td class="memItemLeft" align="right" valign="top"><a id="abf0e792c29eb4ea90049ca996cad0022"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>coeffgenes</b> = None</td></tr>
<tr class="separator:abf0e792c29eb4ea90049ca996cad0022"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a70eae30eb6f055c7955969ec573a3f23"><td class="memItemLeft" align="right" valign="top"><a id="a70eae30eb6f055c7955969ec573a3f23"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>scoregenes</b> = None</td></tr>
<tr class="separator:a70eae30eb6f055c7955969ec573a3f23"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:acbae28b1a15a26eedfa82aed4c8ae24f"><td class="memItemLeft" align="right" valign="top"><a id="acbae28b1a15a26eedfa82aed4c8ae24f"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>scorespecies</b> = None</td></tr>
<tr class="separator:acbae28b1a15a26eedfa82aed4c8ae24f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a37440d3a799d825552f477fc4a44c54f"><td class="memItemLeft" align="right" valign="top"><a id="a37440d3a799d825552f477fc4a44c54f"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>coeffspecies</b> = None</td></tr>
<tr class="separator:a37440d3a799d825552f477fc4a44c54f"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pri-methods"></a>
Private Member Functions</h2></td></tr>
<tr class="memitem:acf1d1899b90f65a22eef8bc93b1735ab"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#acf1d1899b90f65a22eef8bc93b1735ab">_dictify</a> (self, save_workspace, save_homo, add_homosets)</td></tr>
<tr class="memdesc:acf1d1899b90f65a22eef8bc93b1735ab"><td class="mdescLeft">&#160;</td><td class="mdescRight">Used by the saving function.  <a href="#acf1d1899b90f65a22eef8bc93b1735ab">More...</a><br /></td></tr>
<tr class="separator:acf1d1899b90f65a22eef8bc93b1735ab"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8f60d259562234b7872cdc9cef03ce75"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a8f60d259562234b7872cdc9cef03ce75">_undictify</a> (self, data)</td></tr>
<tr class="memdesc:a8f60d259562234b7872cdc9cef03ce75"><td class="mdescLeft">&#160;</td><td class="mdescRight">same function but to retransform everything  <a href="#a8f60d259562234b7872cdc9cef03ce75">More...</a><br /></td></tr>
<tr class="separator:a8f60d259562234b7872cdc9cef03ce75"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pri-attribs"></a>
Private Attributes</h2></td></tr>
<tr class="memitem:aeb1012203f7072dafb43d297f76a438e"><td class="memItemLeft" align="right" valign="top"><a id="aeb1012203f7072dafb43d297f76a438e"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>_is_saved</b></td></tr>
<tr class="separator:aeb1012203f7072dafb43d297f76a438e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1cfa3e77ee07be4e8f1dd233fc97fea3"><td class="memItemLeft" align="right" valign="top"><a id="a1cfa3e77ee07be4e8f1dd233fc97fea3"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>_is_loaded</b></td></tr>
<tr class="separator:a1cfa3e77ee07be4e8f1dd233fc97fea3"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pri-static-attribs"></a>
Static Private Attributes</h2></td></tr>
<tr class="memitem:aeb1012203f7072dafb43d297f76a438e"><td class="memItemLeft" align="right" valign="top"><a id="aeb1012203f7072dafb43d297f76a438e"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><b>_is_saved</b></td></tr>
<tr class="separator:aeb1012203f7072dafb43d297f76a438e"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p><a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html" title="PyCUB is the main object of the project that allows the user to access most of the functions...">PyCUB</a> is the main object of the project that allows the user to access most of the functions. </p>
<p>When using it, please follow the documentation and examples on notebooks thought you can still use it as you please and use some of the nice tricks provided here and in python</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">species</td><td>dictionary of Espece objects from the name of the species. (see espece.py) </td></tr>
    <tr><td class="paramname">working_homoset</td><td>PyCUB.homoset object that stores a subset of the homologies you want to work on all_homoset PyCUB.homoset that stores the all the homologies </td></tr>
    <tr><td class="paramname">session</td><td>str the session name you want to use (will appear in the savings for example </td></tr>
    <tr><td class="paramname">_is_saved</td><td>bool trivial system only boolean </td></tr>
    <tr><td class="paramname">links</td><td>dict of all the links readily available in <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html" title="PyCUB is the main object of the project that allows the user to access most of the functions...">PyCUB</a>. for the project of Jeremie KALFON please use whatever datasets you may find usefull (you can also download from Ensembl)</td></tr>
    <tr><td class="paramname">coeffgenes</td><td>np.array regressing values for each attributes </td></tr>
    <tr><td class="paramname">scoregenes</td><td>the score of the regressor </td></tr>
    <tr><td class="paramname">scorespecies</td><td>the score of the regressor </td></tr>
    <tr><td class="paramname">coeffspecies</td><td>np.array regressing values for each attributes </td></tr>
  </table>
  </dd>
</dl>
</div><h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a id="a0a8de5f32a03de2c681c0582d9998f03"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a0a8de5f32a03de2c681c0582d9998f03">&#9670;&nbsp;</a></span>__init__()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.pyCUB.PyCUB.__init__ </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>species</em> = <code>{}</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>_is_saved</em> = <code>False</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>_is_loaded</em> = <code>False</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>working_homoset</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>all_homoset</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>session</em> = <code>'session1'</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>will initialize the object with the different values you might have from another project </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">species</td><td>dictionary of Espece objects from the name of the species. (see espece.py) </td></tr>
    <tr><td class="paramname">working_homoset</td><td>PyCUB.homoset object that stores a subset of the homologies you want to work on all_homoset PyCUB.homoset that stores the all the homologies </td></tr>
    <tr><td class="paramname">session</td><td>str the session name you want to use (will appear in the savings for example </td></tr>
    <tr><td class="paramname">_is_saved</td><td>bool trivial system only boolean </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a id="acf1d1899b90f65a22eef8bc93b1735ab"></a>
<h2 class="memtitle"><span class="permalink"><a href="#acf1d1899b90f65a22eef8bc93b1735ab">&#9670;&nbsp;</a></span>_dictify()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.pyCUB.PyCUB._dictify </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>save_workspace</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>save_homo</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>add_homosets</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Used by the saving function. </p>
<p>transform the workspace object into a dictionary that can be json serializable </p><pre class="fragment">   adding some params because else the object may be too big
</pre><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">save_workspace</td><td>bool to save working_homoset </td></tr>
    <tr><td class="paramname">save_homo</td><td>bool to save all_homoset </td></tr>
    <tr><td class="paramname">ass_homosets</td><td>PyCUB.homoset instances to add to this dict</td></tr>
  </table>
  </dd>
</dl>
<dl class="section user"><dt>Return</dt><dd>A dict holding every element to be jsonized </dd></dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a4e13b55153bbe774e9203818f2b7f690">PyCUB.pyCUB.PyCUB.plot_distances()</a>, and <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a477c55065989b10bce8389c5edaf3322">PyCUB.pyCUB.PyCUB.save()</a>.</p>

</div>
</div>
<a id="a8f60d259562234b7872cdc9cef03ce75"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a8f60d259562234b7872cdc9cef03ce75">&#9670;&nbsp;</a></span>_undictify()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.pyCUB.PyCUB._undictify </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>data</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>same function but to retransform everything </p>
<p>Here we don't use other classes undictify functions but we just recreate them by passing it to their init methods which is clearer.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">data</td><td>dict to undictify into the workspace object</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Other PyCUB.homosets that would have been saved as add_homosets </dd></dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#acf1d1899b90f65a22eef8bc93b1735ab">PyCUB.pyCUB.PyCUB._dictify()</a>, and <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#ab471e823b0eba23af5a5322008ec09da">PyCUB.pyCUB.PyCUB.load()</a>.</p>

</div>
</div>
<a id="ad350376cd68d4daacec78c694ed754c6"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad350376cd68d4daacec78c694ed754c6">&#9670;&nbsp;</a></span>compare_homologies()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.pyCUB.PyCUB.compare_homologies </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>homoset</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>homosapiens</em> = <code>False</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>mindistance</em> = <code>10</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>preserved</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>size</em> = <code>10</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>minpreserv</em> = <code>0.9</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>minsimi</em> = <code>0.9</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>showvar</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>eps</em> = <code>0.28</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>reducer</em> = <code>'tsne'</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>perplexity</em> = <code>40</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>finds for species with a common ancester separated by a pseudo phylogenetic distance X, </p>
<p>genes/functions that are novel to only a subset. plot the two with their differences the differences between the two also considers an homology as highly preserved if it is shared amongst most of the species and if the average similarity score is high amongst this homology. also shows if there is a relationship between the number of amino acids the sequence does not encode for and the codon usage bias We could have used the sequence dating of ensembl but it only works for homo sapiens for now maybe use it for homosapiens later</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">homosapiens</td><td>bool to true if we should use homosapiens dataset on gene dates </td></tr>
    <tr><td class="paramname">mindistance</td><td>int the minimal phylogenetic distance between in average in this homology to consider it highly conserved </td></tr>
    <tr><td class="paramname">preserved</td><td>bool to true if we should find highly preserved genes or not </td></tr>
    <tr><td class="paramname">size</td><td>the average size of the datapoints in the pointcloud representation of this dataset </td></tr>
    <tr><td class="paramname">minpreserv</td><td>float minimal percentage of homologous species that have this homology </td></tr>
    <tr><td class="paramname">minsimi</td><td>float minimal avg similarity between genes to consider them highly preserved </td></tr>
    <tr><td class="paramname">showvar</td><td>bool to true, show the mean variance in CUB values accros this homology as a variation in dot sizes </td></tr>
    <tr><td class="paramname">eps</td><td>float the hyperparamter of the clustering algorithm applied to this dataset </td></tr>
    <tr><td class="paramname">homoset</td><td>PyCUB.homoset the homoset to use </td></tr>
    <tr><td class="paramname">reducer</td><td>str the reducer to use 'tsne' or 'PCA' </td></tr>
    <tr><td class="paramname">perplexity</td><td>int the perplexity hyperparam for tSNE</td></tr>
  </table>
  </dd>
</dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">UnboundLocalError</td><td>"you need to compute the averages of the all_homoset. use PyCUB.compute_averages(homoset)" </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#af37319783148fc8864d894b3e2f89078">PyCUB.pyCUB.PyCUB.regress_on_species()</a>.</p>

</div>
</div>
<a id="ad7f24b316bc8c257719da20c0792ec01"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad7f24b316bc8c257719da20c0792ec01">&#9670;&nbsp;</a></span>compare_species()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.pyCUB.PyCUB.compare_species </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>showvar</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>reducer</em> = <code>'tsne'</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>perplexity</em> = <code>40</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>eps</em> = <code>0.3</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>size</em> = <code>10</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>compare the species according to their mean CUB, </p>
<p>plot the mean CUB to their full CUB, to their tRNA copy numbers, to the euclidean distance of their CUB to the one of their phylogenetic matrix. in this plot, the mean entropy value is plotted as a regular homology plot but each dot is a species thus we can compare them together, moreover, the size of the dots informs oneself of the variance in entropy per species. the color intensity informs on how much this is close to what is given by the tRNA values. (additional information such as name of the species, number of tRNA values,metadata and the point from its tRNA value is plotted when hovering the dot) then we can also compare the mean of homologies or else, the full entropy of the cdna sequence per species is also computed the euclidean distance amongst the species for the full entropy to see if a difference can be linked to some evolutionary for one codon</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">size</td><td>the average size of the datapoints in the pointcloud representation of this dataset </td></tr>
    <tr><td class="paramname">showvar</td><td>bool to true, show the mean variance in CUB values accros this homology as a variation in dot sizes </td></tr>
    <tr><td class="paramname">eps</td><td>float the hyperparamter of the clustering algorithm applied to this dataset </td></tr>
    <tr><td class="paramname">homoset</td><td>PyCUB.homoset the homoset to use </td></tr>
    <tr><td class="paramname">reducer</td><td>str the reducer to use 'tsne' or 'PCA' </td></tr>
    <tr><td class="paramname">perplexity</td><td>int the perplexity hyperparam for tSNE</td></tr>
  </table>
  </dd>
</dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">UnboundLocalError</td><td>"you need to compute the averages of the all_homoset. use PyCUB.compute_averages(homoset)" </td></tr>
    <tr><td class="paramname">UnboundLocalError</td><td>"you have to compute tRNA values, use PyCUB.get_tRNAcopy()" </td></tr>
    <tr><td class="paramname">AttributeError</td><td>"try avg or full" </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a9526c0d10381daf5abc8ea120454e701">PyCUB.pyCUB.PyCUB.compute_averages()</a>.</p>

</div>
</div>
<a id="a9526c0d10381daf5abc8ea120454e701"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a9526c0d10381daf5abc8ea120454e701">&#9670;&nbsp;</a></span>compute_averages()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.pyCUB.PyCUB.compute_averages </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>homoset</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>compute the average entropy </p>
<p>Will add species related averages gotten from this homoset in the species container, and in the homoset everytime you compute averages from a set, they will get erased !</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">homoset</td><td>PyCUB.homoset from which to compute the averages </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#af3e280c265200f13946e4bdcb4cd9dd0">PyCUB.pyCUB.PyCUB.speciestable()</a>.</p>

</div>
</div>
<a id="a40a742a1a05fc21c7bced2c75fb93fc6"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a40a742a1a05fc21c7bced2c75fb93fc6">&#9670;&nbsp;</a></span>createRefCAI()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.pyCUB.PyCUB.createRefCAI </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>speciestocompare</em> = <code>'saccharomyces_cerevisiae'</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>kingdom</em> = <code>'fungi'</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>first</em> = <code>20</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>do a compute CAI </p>
<p>where we get Tobias' data to find highly expressed genes and use them to compute codon frequency for the reference set and use it to compute the CAI and mean CAI for each ho mology.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">speciestocompare</td><td>str the name of the species to retrieve the genes from </td></tr>
    <tr><td class="paramname">kingdom</td><td>str the kingdom where we can find it </td></tr>
    <tr><td class="paramname">first</td><td>the number of highly expressed genes to retrieve </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#ac4d231f47432a18f84da1cef8673b50b">PyCUB.pyCUB.PyCUB.get_data()</a>, and <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#af0d7bed125f4437bea6b8051f9922c1d">PyCUB.pyCUB.PyCUB.get_evolutionary_distance()</a>.</p>

</div>
</div>
<a id="ac4d231f47432a18f84da1cef8673b50b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac4d231f47432a18f84da1cef8673b50b">&#9670;&nbsp;</a></span>get_data()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.pyCUB.PyCUB.get_data </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>From</em> = <code>'yun'</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>homonames</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>kingdom</em> = <code>'fungi'</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>sequence</em> = <code>'cdna'</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>additional</em> = <code>'type=orthologues'</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>saveonfiles</em> = <code>False</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>normalized</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>setnans</em> = <code>False</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>by</em> = <code>&quot;entropy&quot;</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>using</em> = <code>&quot;normal&quot;</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>tRNA</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>getCAI</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>first</em> = <code>20</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>inpar</em> = <code>True</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Download the data from somewhere on the web (Ensembl, Yun(with links)) </p>
<p>you can provide a lot of different values to scrape Ensembl's datasets it will compute from ensembl to retrieve a similar dataset as what yun's data is.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">From</td><td>str flag 'yun' or 'ensembl': </td></tr>
    <tr><td class="paramname">homonames</td><td>list[str] what particular homologies you want to scrap if 'all' and you have used the <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#ae06a2843c0718f86396ee255dfc6ae6f" title="A function to retrieve the homologies directly from a given species. ">getHomologylist()</a> function, will get the homologies from there </td></tr>
    <tr><td class="paramname">kingdom</td><td>str same for kingdoms </td></tr>
    <tr><td class="paramname">sequence</td><td>str the type of sequences you want to use </td></tr>
    <tr><td class="paramname">additional</td><td>str additional information about the scrapping </td></tr>
    <tr><td class="paramname">saveonfiles</td><td>bool save the unprocessed data before populating working homoset </td></tr>
    <tr><td class="paramname">normalized</td><td>bool if you want the values to be normalized by the length of the codons (lengths are always saved) </td></tr>
    <tr><td class="paramname">setnans</td><td>bool if you want to save the nans as metadata </td></tr>
    <tr><td class="paramname">by</td><td>str flag 'entropy', 'entropyLocation' (entropy location), 'frequency' </td></tr>
    <tr><td class="paramname">using</td><td>str flag 'random' 'normal' 'permutation' 'full' </td></tr>
    <tr><td class="paramname">inpar</td><td>bool or int for parallel computing and number of core </td></tr>
    <tr><td class="paramname">tRNA</td><td>bool whether or not to compute tRNA data </td></tr>
    <tr><td class="paramname">getCAI</td><td>bool flag to true to retrieve the CAI as well </td></tr>
    <tr><td class="paramname">first</td><td>int the first most expressed genes to compute the CAI ref statistics</td></tr>
  </table>
  </dd>
</dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">AttributeError</td><td>"you can't compute codon frequency with Yun's data...", 'not the right From' </td></tr>
    <tr><td class="paramname">UnboundLocalError</td><td>"you have to load the homologies first" </td></tr>
    <tr><td class="paramname">AttributeError</td><td>'not the right From' <pre class="fragment">   http://rest.ensemblgenomes.org/</pre> </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#ae06a2843c0718f86396ee255dfc6ae6f">PyCUB.pyCUB.PyCUB.getHomologylist()</a>.</p>

</div>
</div>
<a id="af0d7bed125f4437bea6b8051f9922c1d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af0d7bed125f4437bea6b8051f9922c1d">&#9670;&nbsp;</a></span>get_evolutionary_distance()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.pyCUB.PyCUB.get_evolutionary_distance </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>display_tree</em> = <code>False</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>size</em> = <code>40</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>uses metadata of the ancestry tree and computes a theoretical evolutionary distance matrix between each species </p>
<p>can optionaly take any hierarchical evolutionary file between a group of species will populate utils.phylo_distances with a pandas.df of the phylodistance and meandist with the average distance amongst species in the df, species are referenced by their taxon ids. you have to have taxons in your species. will also plot the distance matrix</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">display_tree</td><td>bool to true to print the phylogenetic tree as a txt (may be quite big) </td></tr>
    <tr><td class="paramname">size</td><td>int the x size of the plot</td></tr>
  </table>
  </dd>
</dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">EnvironmentError</td><td>"you need to have R installed to compute the distance" </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a4e9a2d43aacfc8d2fb1610140f2439f2">PyCUB.pyCUB.PyCUB.get_taxons()</a>.</p>

</div>
</div>
<a id="a53b60adfafbe019421d010c5fed04f5c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a53b60adfafbe019421d010c5fed04f5c">&#9670;&nbsp;</a></span>get_full_genomes()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.pyCUB.PyCUB.get_full_genomes </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>kingdom</em> = <code>'fungi'</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>seq</em> = <code>'cds'</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>avg</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>by</em> = <code>&quot;entropy&quot;</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>normalized</em> = <code>False</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>go trought all full genome fasta files in the ftp server of ensemblgenomes and </p>
<p>download then parse them to get the full entropy of the genome. usefull for futher comparison steps. will populate the fullentropy, fullvarentropy, fullGCcount, varGCcount of each species where the full sequence is known</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">kingdom</td><td>str flags the relevant kingdom of you current session [fungi,plants,bacteria, animals] </td></tr>
    <tr><td class="paramname">seq</td><td>str flags the type of sequence you consider the full genome is (coding or non coding or full) [cds, all, cda] </td></tr>
    <tr><td class="paramname">avg</td><td>bool to true if we average over each gene or get the full dna in one go. </td></tr>
    <tr><td class="paramname">by</td><td>str flags what type of computation should be done [entropy,frequency] </td></tr>
    <tr><td class="paramname">normalized</td><td>should we normalize the entorpy by length</td></tr>
  </table>
  </dd>
</dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">MemoryError</td><td>"this sequence is too long to be computed (&gt; 1 billion bp)" </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a9fab964d44e03b46d330b3754f5c39fb">PyCUB.pyCUB.PyCUB.get_subset()</a>.</p>

</div>
</div>
<a id="ae561ba39b9c3919a83b51944f155e4a4"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae561ba39b9c3919a83b51944f155e4a4">&#9670;&nbsp;</a></span>get_metadata_Ensembl()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.pyCUB.PyCUB.get_metadata_Ensembl </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>kingdoms</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>download it and put it where it belongs in the Espece object </p>
<p>parse the server <a href="https://fungi.ensembl.org/info/website/ftp/index.html">https://fungi.ensembl.org/info/website/ftp/index.html</a> will also get the metadata from the kingdoms that you are analysing</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">kingdoms</td><td>str flag the type of kingdoms you wanna have 'fungi' 'bacteria' 'plants' 'animals' </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#ac4d231f47432a18f84da1cef8673b50b">PyCUB.pyCUB.PyCUB.get_data()</a>.</p>

</div>
</div>
<a id="a1ff011ae5b64518d60bedcf1b98314ef"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a1ff011ae5b64518d60bedcf1b98314ef">&#9670;&nbsp;</a></span>get_mymetadata()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.pyCUB.PyCUB.get_mymetadata </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>From</em> = <code>'jerem'</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>inpar</em> = <code>True</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Go ahead and design your own metadata retrieval here. </p>
<p>obviously you woud need to change some other functions. for me it is mean protein abundances in cerevisiae cells.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">From</td><td>str flag designer of the function to load metadatas </td></tr>
    <tr><td class="paramname">inpar</td><td>bool for parallel processing </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#ae561ba39b9c3919a83b51944f155e4a4">PyCUB.pyCUB.PyCUB.get_metadata_Ensembl()</a>.</p>

</div>
</div>
<a id="a9fab964d44e03b46d330b3754f5c39fb"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a9fab964d44e03b46d330b3754f5c39fb">&#9670;&nbsp;</a></span>get_subset()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.pyCUB.PyCUB.get_subset </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>homoset</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>withcopy</em> = <code>False</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>clusternb</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>species</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>homologies</em> = <code>None</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>either changes or returns a subset of the provided homoset </p>
<p>To use once if you want to further refine a set of homologies</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">homoset</td><td>PyCUB.homoset to get a subset from </td></tr>
    <tr><td class="paramname">withcopy</td><td>bool to true if we don't want to change the homoset object but create a copy from it </td></tr>
    <tr><td class="paramname">clusternb</td><td>int set the cluster of the group you want to get need to be between 1 and homogroupnb </td></tr>
    <tr><td class="paramname">homologies</td><td>list[str] the subset as a list or a tuple of int </td></tr>
    <tr><td class="paramname">species</td><td>list[the subset as a list, or a list of int</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a HomoSet object (see homoset.py) </dd></dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a1916ef696d74af1028a2aeff643727db">PyCUB.pyCUB.PyCUB.get_working_homoset()</a>.</p>

</div>
</div>
<a id="a4e9a2d43aacfc8d2fb1610140f2439f2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a4e9a2d43aacfc8d2fb1610140f2439f2">&#9670;&nbsp;</a></span>get_taxons()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.pyCUB.PyCUB.get_taxons </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>find the taxons of each referenced species (see PyCUB.Espece.gettaxons()) </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">None</td><td></td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a53b60adfafbe019421d010c5fed04f5c">PyCUB.pyCUB.PyCUB.get_full_genomes()</a>.</p>

</div>
</div>
<a id="a1916ef696d74af1028a2aeff643727db"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a1916ef696d74af1028a2aeff643727db">&#9670;&nbsp;</a></span>get_working_homoset()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.pyCUB.PyCUB.get_working_homoset </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>clusternb</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>species</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>homologies</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>cleanhomo</em> = <code>None</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>create a subset of all_homoset on which you would like to do further computation </p>
<p>To use once you have clustered homology groups, else takes everything. Can also be used just to get a subset of the all homosets.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">clusternb</td><td>int set the cluster of the group you want to get need to be between 1 and homogroupnb </td></tr>
    <tr><td class="paramname">homologies</td><td>list[str] the subset as a list you want to get from all_homoset (can be additional to a clusternb) </td></tr>
    <tr><td class="paramname">species</td><td>list[str] the subset as a list you want to get from all_homoset (can be additional to a clusternb) </td></tr>
    <tr><td class="paramname">cleanhomo</td><td>float if the homology is only shared by less than this amount amongst the species present in this homoset, removes them.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a HomoSet object (see homoset.py)</dd></dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">UnboundLocalError</td><td>"you have not clusterized you 'all_homoset'. you want to just use 'order_from_matrix' on it." </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a477c55065989b10bce8389c5edaf3322">PyCUB.pyCUB.PyCUB.save()</a>.</p>

</div>
</div>
<a id="ae06a2843c0718f86396ee255dfc6ae6f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae06a2843c0718f86396ee255dfc6ae6f">&#9670;&nbsp;</a></span>getHomologylist()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.pyCUB.PyCUB.getHomologylist </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>species</em> = <code>'saccharomyces_cerevisiae'</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>kingdom</em> = <code>'fungi'</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>A function to retrieve the homologies directly from a given species. </p>
<p>(it is better to use one of the key species for the different kingdoms (sacharomyces, HS, Arabidopsis..))</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">specie</td><td>str the name of the specie to get the homology from </td></tr>
    <tr><td class="paramname">kingdom</td><td>str the kingdom where we can find this specie </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a0a8de5f32a03de2c681c0582d9998f03">PyCUB.pyCUB.PyCUB.__init__()</a>.</p>

</div>
</div>
<a id="a06d83a29ac506b4a9422ef7e64eef0b0"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a06d83a29ac506b4a9422ef7e64eef0b0">&#9670;&nbsp;</a></span>getRelation2G3DD()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.pyCUB.PyCUB.getRelation2G3DD </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>species_name</em> = <code>'saccharomyces_cerevisiae'</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>kingdom</em> = <code>'fungi'</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>intrachromosome</em> = <code>&quot;utils/meta/3Dmodel/interactions_HindIII_fdr0.01_intra_cerevisiae.csv&quot;</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>interchromose</em> = <code>[&quot;utils/meta/3Dmodel/cerevisiae_inter1.csv&quot;</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>utils</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>meta</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>Dmodel</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>cerevisiae_inter2</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>csv</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>utils</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>meta</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>Dmodel</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>cerevisiae_inter3</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>csv</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>utils</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>meta</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>Dmodel</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>cerevisiae_inter4</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>csv</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>utils</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>meta</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>Dmodel</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>cerevisiae_inter5</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>csv</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>bins</em> = <code>2000</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>seq</em> = <code>'cds'</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>use</em> = <code>'diament2'</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><a href="https://www.nature.com/articles/ncomms6876">https://www.nature.com/articles/ncomms6876</a> </p>
<p>retrieve the data for the species sacharomyces cerevisiae and Schizosaccharomyces pombe and find if similarity distances of CUB using entropy between genes of this species is predictive of closeness of genes in the nucleus.</p>
<p>Used to confirm a work on nature and see if we can have some similar results by only looking at the CUB</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">species_name</td><td>str the name of the species to look for </td></tr>
    <tr><td class="paramname">kingdom</td><td>str the kingdom in which to find the species </td></tr>
    <tr><td class="paramname">intrachromosome</td><td>str the location of the csv interaction data for intrachromosome respecting the format of the default file </td></tr>
    <tr><td class="paramname">interchromose</td><td>str the location of the csv interaction data for interchromose respecting the format of the default file </td></tr>
    <tr><td class="paramname">bins</td><td>int, the number of bin to use (a power of 2) </td></tr>
    <tr><td class="paramname">seq</td><td>the type of sequence to compare to. (to compute the CUB from) </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a1b65efe7deb4ba5f65203c8be6fc7af2">PyCUB.pyCUB.PyCUB.regress_on_genes()</a>.</p>

</div>
</div>
<a id="acd80f084f6671584c42542813874cd02"></a>
<h2 class="memtitle"><span class="permalink"><a href="#acd80f084f6671584c42542813874cd02">&#9670;&nbsp;</a></span>import_metadataTobias()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.pyCUB.PyCUB.import_metadataTobias </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>will import the metadata obtained from tobias for the fungi species affiliated to cerevisiae to each species for further diagnostics. </p>
<p>Populates metadata[num_genes, plant_pathogen, animal_pathogen, genome_size, plant_symbiotic, brown_rot, white_rot] for each species and weight, mRNA_abundance, is_secreted, protein_abundance, cys_elements, decay_rate for each homology</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">None</td><td></td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a1ff011ae5b64518d60bedcf1b98314ef">PyCUB.pyCUB.PyCUB.get_mymetadata()</a>.</p>

</div>
</div>
<a id="ab471e823b0eba23af5a5322008ec09da"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab471e823b0eba23af5a5322008ec09da">&#9670;&nbsp;</a></span>load()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.pyCUB.PyCUB.load </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>session</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>All</em> = <code>False</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>filename</em> = <code>'first500'</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>From</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>by</em> = <code>'entropy'</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>tRNA</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>inpar</em> = <code>True</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Get the data that is already present on a filename. </p>
<p>Either load from Yun's datasets or from an already saved session. Is being called by get_data. But you can call it to just use one of Yun's files as well</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">From</td><td>str if this flag is set to 'yun' it means that the filename is made of Yundata in which case we will create directly the homology map in the same time as the rest of the <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html" title="PyCUB is the main object of the project that allows the user to access most of the functions...">PyCUB</a> object. </td></tr>
    <tr><td class="paramname">All</td><td>bool set to true if load everything from Yun </td></tr>
    <tr><td class="paramname">by</td><td>str same flag as get_data (for Yun's files here). </td></tr>
    <tr><td class="paramname">filename</td><td>str the particular filename when not loading them all </td></tr>
    <tr><td class="paramname">session</td><td>str if a session name is provided, then will load a zip file from this session's folder</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>May return additionals if loading from a session where one decided to save more than the two All/working homologies. to be handled separately </dd></dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#ac4d231f47432a18f84da1cef8673b50b">PyCUB.pyCUB.PyCUB.get_data()</a>, and <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#acd80f084f6671584c42542813874cd02">PyCUB.pyCUB.PyCUB.import_metadataTobias()</a>.</p>

</div>
</div>
<a id="a16a1ab6be5dddcedf94852088936c1f2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a16a1ab6be5dddcedf94852088936c1f2">&#9670;&nbsp;</a></span>loadmore()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.pyCUB.PyCUB.loadmore </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>filename</em> = <code>'first500'</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>by</em> = <code>'entropyLocation'</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Get the data that is already present on a filename when you already have data. </p>
<p>is usefull to load more of Yun's datasets. is called when load is set to All</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">From</td><td>if this flag is set to 'yun' it means that the filename is made of Yundata in which case we will create directly the homology map in the same time as the rest of the <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html" title="PyCUB is the main object of the project that allows the user to access most of the functions...">PyCUB</a> object. Here it is the only available option. </td></tr>
    <tr><td class="paramname">filename</td><td>str the filename to additionaly load </td></tr>
    <tr><td class="paramname">by</td><td>flag same as before</td></tr>
  </table>
  </dd>
</dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">UnboundLocalError</td><td>"You should try load first, this object is empty" </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#ab471e823b0eba23af5a5322008ec09da">PyCUB.pyCUB.PyCUB.load()</a>.</p>

</div>
</div>
<a id="ac37364050a4ec9a7ddf3d6c1562e49a5"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac37364050a4ec9a7ddf3d6c1562e49a5">&#9670;&nbsp;</a></span>loadspeciestable()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.pyCUB.PyCUB.loadspeciestable </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>short function to retrieve the speciestable from Disk </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">None</td><td></td></tr>
  </table>
  </dd>
</dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">IOError</td><td>"no speciestable file" </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a8f60d259562234b7872cdc9cef03ce75">PyCUB.pyCUB.PyCUB._undictify()</a>.</p>

</div>
</div>
<a id="a2ea40ac1dd5b4ad8847ec12c867d457f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a2ea40ac1dd5b4ad8847ec12c867d457f">&#9670;&nbsp;</a></span>phylo_distances()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.pyCUB.PyCUB.phylo_distances </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>a copy of the phylodistances dataframe see (<a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#af0d7bed125f4437bea6b8051f9922c1d" title="uses metadata of the ancestry tree and computes a theoretical evolutionary distance matrix between ea...">get_evolutionary_distance()</a>) </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">None</td><td></td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a copy of the phylodistances dataframe see (<a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#af0d7bed125f4437bea6b8051f9922c1d" title="uses metadata of the ancestry tree and computes a theoretical evolutionary distance matrix between ea...">get_evolutionary_distance()</a>) </dd></dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#ad7f24b316bc8c257719da20c0792ec01">PyCUB.pyCUB.PyCUB.compare_species()</a>, and <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a40a742a1a05fc21c7bced2c75fb93fc6">PyCUB.pyCUB.PyCUB.createRefCAI()</a>.</p>

</div>
</div>
<a id="a4e13b55153bbe774e9203818f2b7f690"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a4e13b55153bbe774e9203818f2b7f690">&#9670;&nbsp;</a></span>plot_distances()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.pyCUB.PyCUB.plot_distances </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>size</em> = <code>40</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>plot the phylogenetic distance matrix </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">size</td><td>int the x size of the plot</td></tr>
  </table>
  </dd>
</dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">UnboundLocalError</td><td>"compute the phylo distance matrix first (look at the doc)" </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#af0d7bed125f4437bea6b8051f9922c1d">PyCUB.pyCUB.PyCUB.get_evolutionary_distance()</a>, and <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a06d83a29ac506b4a9422ef7e64eef0b0">PyCUB.pyCUB.PyCUB.getRelation2G3DD()</a>.</p>

</div>
</div>
<a id="a1b65efe7deb4ba5f65203c8be6fc7af2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a1b65efe7deb4ba5f65203c8be6fc7af2">&#9670;&nbsp;</a></span>regress_on_genes()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.pyCUB.PyCUB.regress_on_genes </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>homoset</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>full</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>without</em> = <code>['meanecai'</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>meancai</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>perctrain</em> = <code>0.8</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>algo</em> = <code>&quot;lasso&quot;</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>eps</em> = <code>0.001</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>n_alphas</em> = <code>100</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Will fit a regression curve on the CUB values of the different homologies according to the metadatas available for each of them. </p>
<p>It will try to see if there is enough information in the metadata to retrieve CUB values. and if there is, how much for each metadata (if we constraint the number of regressors) is it better for entropy values, mean entropy or ECAI values or raw frequency, should we remove some data</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">without</td><td>list[str] of flags [similarity_scores, KaKs_Scores, nans, lenmat, GCcount, weight, <div class="fragment"><div class="line">protein_abundance, mRNA_abundance, decay_rate, cys_elements, tot_volume, mean_hydrophobicity,</div></div><!-- fragment --> glucose_cost, synthesis_steps, is_recent, meanecai] </td></tr>
    <tr><td class="paramname">full</td><td>bool flags to true to use full CUB values or meanCUB values, as regressee </td></tr>
    <tr><td class="paramname">homoset</td><td>PyCUB.homoset the homoset to use </td></tr>
    <tr><td class="paramname">perctrain</td><td>the percentage of training set to total set ( the rest is used as test set) </td></tr>
    <tr><td class="paramname">algo</td><td>str flag to lasso or nn to use either Lasso with Cross Validation, or a 2 layer neural net </td></tr>
    <tr><td class="paramname">eps</td><td>the eps value for the Lasso </td></tr>
    <tr><td class="paramname">n_alphas</td><td>the number of alphas for the lasso</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd></dd>
<dd>
scoregenes float, the score of the regression performed </dd>
<dd>
coeffgenes the coefficient applied to each category (for each CUB value if using full) </dd>
<dd>
attrlist the corresponding list[str] of attribute used</dd></dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">UnboundLocalError</td><td>"wrong params" </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#ad350376cd68d4daacec78c694ed754c6">PyCUB.pyCUB.PyCUB.compare_homologies()</a>.</p>

</div>
</div>
<a id="af37319783148fc8864d894b3e2f89078"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af37319783148fc8864d894b3e2f89078">&#9670;&nbsp;</a></span>regress_on_species()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.pyCUB.PyCUB.regress_on_species </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>without</em> = <code>[&quot;&quot;]</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>full</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>onlyhomo</em> = <code>False</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>perctrain</em> = <code>0.8</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>algo</em> = <code>&quot;lasso&quot;</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>eps</em> = <code>0.001</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>n_alphas</em> = <code>100</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Will fit a regression curve on the CUB values of the different species according to the metadatas available for each of them. </p>
<p>It will try to see if there is enough information in the metadata to retrieve CUB values. and if there is, how much for each metadata (if we constraint the number of regressors) is it better for mean homology CUB or full genome CUB ? or raw frequency, should we remove some data?</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">without</td><td>list[str] of flags [similarity_scores, KaKs_Scores, nans, lenmat, GCcount, weight, <div class="fragment"><div class="line">protein_abundance, mRNA_abundance, decay_rate, cys_elements, tot_volume, mean_hydrophobicity,</div></div><!-- fragment --> glucose_cost, synthesis_steps, is_recent, meanecai] </td></tr>
    <tr><td class="paramname">full</td><td>bool flags to true to use full CUB values or meanCUB values, as regressee </td></tr>
    <tr><td class="paramname">homoset</td><td>PyCUB.homoset the homoset to use </td></tr>
    <tr><td class="paramname">perctrain</td><td>the percentage of training set to total set ( the rest is used as test set) </td></tr>
    <tr><td class="paramname">algo</td><td>str flag to lasso or nn to use either Lasso with Cross Validation, or a 2 layer neural net </td></tr>
    <tr><td class="paramname">eps</td><td>the eps value for the Lasso </td></tr>
    <tr><td class="paramname">n_alphas</td><td>the number of alphas for the lasso</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd></dd>
<dd>
scoregenes float, the score of the regression performed </dd>
<dd>
coeffgenes the coefficient applied to each category (for each CUB value if using full) </dd>
<dd>
attrlist the corresponding list[str] of attribute used</dd></dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">UnboundLocalError</td><td>"wrong params" </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#ad7f24b316bc8c257719da20c0792ec01">PyCUB.pyCUB.PyCUB.compare_species()</a>.</p>

</div>
</div>
<a id="a477c55065989b10bce8389c5edaf3322"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a477c55065989b10bce8389c5edaf3322">&#9670;&nbsp;</a></span>save()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.pyCUB.PyCUB.save </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>name</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>save_workspace</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>save_homo</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>add_homosets</em> = <code>{}</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>cmdlinetozip</em> = <code>&quot;gzip&quot;</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>call to save your work. </p>
<p>you should call save on specific data structure if this is what you want to save. </p><pre class="fragment">   Will call other object's save, will transform all the variable into dict and save the dicts
   as json files. will save the df also as json files. PyCUB and homoset have
   their own json file.
   adding some params because else the object may be too big
</pre><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">name</td><td>str the name of the particular save on this session </td></tr>
    <tr><td class="paramname">save_workspace</td><td>bool to fale not to save working_homoset </td></tr>
    <tr><td class="paramname">save_homo</td><td>bool to false not to save all_homoset </td></tr>
    <tr><td class="paramname">cmdlinetozip</td><td>str you need to tell the platform how to zip on your system uses gzip by default but it needs to be installed </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#ad350376cd68d4daacec78c694ed754c6">PyCUB.pyCUB.PyCUB.compare_homologies()</a>, <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#ad7f24b316bc8c257719da20c0792ec01">PyCUB.pyCUB.PyCUB.compare_species()</a>, and <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a16a1ab6be5dddcedf94852088936c1f2">PyCUB.pyCUB.PyCUB.loadmore()</a>.</p>

</div>
</div>
<a id="a8249834e1d2f7d061ae06ffa5a1adc32"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a8249834e1d2f7d061ae06ffa5a1adc32">&#9670;&nbsp;</a></span>savespeciestable()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.pyCUB.PyCUB.savespeciestable </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>short function to put the speciestable on Disk </p>
<p>This is done since there may be some memory leakage, probably due to some autoreloading behavior of the global data stored on utils.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">None</td><td></td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a8f60d259562234b7872cdc9cef03ce75">PyCUB.pyCUB.PyCUB._undictify()</a>.</p>

</div>
</div>
<a id="af3e280c265200f13946e4bdcb4cd9dd0"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af3e280c265200f13946e4bdcb4cd9dd0">&#9670;&nbsp;</a></span>speciestable()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.pyCUB.PyCUB.speciestable </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>a copy of the utils.speciestable </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">None</td><td></td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a copy of the utils.speciestable (dict[int,str] of species to their <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html" title="PyCUB is the main object of the project that allows the user to access most of the functions...">PyCUB</a> coded value </dd></dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a40a742a1a05fc21c7bced2c75fb93fc6">PyCUB.pyCUB.PyCUB.createRefCAI()</a>.</p>

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>PyCUB/pyCUB.py</li>
</ul>
</div><!-- contents -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="navelem"><b>PyCUB</b></li><li class="navelem"><b>pyCUB</b></li><li class="navelem"><a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html">PyCUB</a></li>
    <li class="footer">Generated by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.8.14 </li>
  </ul>
</div>
</body>
</html>
