<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.14"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>PyCUB: Machine exploration of the Codon Usage Bias: PyCUB.homoset.HomoSet Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
  $(document).ready(initResizable);
/* @license-end */</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    extensions: ["tex2jax.js"],
    jax: ["input/TeX","output/HTML-CSS"],
});
</script><script type="text/javascript" async src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.2/MathJax.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">PyCUB: Machine exploration of the Codon Usage Bias
   </div>
   <div id="projectbrief">A set of tools to explore the codon usage bias</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.14 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(document).ready(function(){initNavTree('class_py_c_u_b_1_1homoset_1_1_homo_set.html','');});
/* @license-end */
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-attribs">Public Attributes</a> &#124;
<a href="#pub-static-attribs">Static Public Attributes</a> &#124;
<a href="#pri-methods">Private Member Functions</a> &#124;
<a href="class_py_c_u_b_1_1homoset_1_1_homo_set-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">PyCUB.homoset.HomoSet Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p><a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html" title="HomoSet is the object containing evrey homology as a dictionnary according to thie rhomology code...">HomoSet</a> is the object containing evrey homology as a dictionnary according to thie rhomology code.  
 <a href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#details">More...</a></p>
<div class="dynheader">
Inheritance diagram for PyCUB.homoset.HomoSet:</div>
<div class="dyncontent">
 <div class="center">
  <img src="class_py_c_u_b_1_1homoset_1_1_homo_set.png" usemap="#PyCUB.homoset.HomoSet_map" alt=""/>
  <map id="PyCUB.homoset.HomoSet_map" name="PyCUB.homoset.HomoSet_map">
</map>
 </div></div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:adbe4d9e4087fe5b53d861476c011329d"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#adbe4d9e4087fe5b53d861476c011329d">__init__</a> (self, kwargs)</td></tr>
<tr class="memdesc:adbe4d9e4087fe5b53d861476c011329d"><td class="mdescLeft">&#160;</td><td class="mdescRight">will initialize the object with the different values you might have from another project use the data dictionnary to add any type of data  <a href="#adbe4d9e4087fe5b53d861476c011329d">More...</a><br /></td></tr>
<tr class="separator:adbe4d9e4087fe5b53d861476c011329d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab7ac33ea3033b440816e179cb4a5e3c0"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#ab7ac33ea3033b440816e179cb4a5e3c0">__getitem__</a> (self, key)</td></tr>
<tr class="memdesc:ab7ac33ea3033b440816e179cb4a5e3c0"><td class="mdescLeft">&#160;</td><td class="mdescRight">get the homology at the corresponding key  <a href="#ab7ac33ea3033b440816e179cb4a5e3c0">More...</a><br /></td></tr>
<tr class="separator:ab7ac33ea3033b440816e179cb4a5e3c0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a42a7e7dc7ddb4b5717084cb093f90774"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a42a7e7dc7ddb4b5717084cb093f90774">__setitem__</a> (self, key, value)</td></tr>
<tr class="memdesc:a42a7e7dc7ddb4b5717084cb093f90774"><td class="mdescLeft">&#160;</td><td class="mdescRight">add an item at the corresponding key  <a href="#a42a7e7dc7ddb4b5717084cb093f90774">More...</a><br /></td></tr>
<tr class="separator:a42a7e7dc7ddb4b5717084cb093f90774"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5211ffec9061ca49c771d070048524a1"><td class="memItemLeft" align="right" valign="top"><a id="a5211ffec9061ca49c771d070048524a1"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>__delitem__</b> (self, key)</td></tr>
<tr class="separator:a5211ffec9061ca49c771d070048524a1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2d35b1436a2bbdaea69454ffbf648bf0"><td class="memItemLeft" align="right" valign="top"><a id="a2d35b1436a2bbdaea69454ffbf648bf0"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>__iter__</b> (self)</td></tr>
<tr class="separator:a2d35b1436a2bbdaea69454ffbf648bf0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aefca94ad0e64bcad582edcc163bc21ad"><td class="memItemLeft" align="right" valign="top"><a id="aefca94ad0e64bcad582edcc163bc21ad"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>iteritems</b> (self)</td></tr>
<tr class="separator:aefca94ad0e64bcad582edcc163bc21ad"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ada72dc2741d84a8c617dc23449154fc2"><td class="memItemLeft" align="right" valign="top"><a id="ada72dc2741d84a8c617dc23449154fc2"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>__len__</b> (self)</td></tr>
<tr class="separator:ada72dc2741d84a8c617dc23449154fc2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8cf76800f047841bf63d507fb1b54048"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a8cf76800f047841bf63d507fb1b54048">update</a> (self, val)</td></tr>
<tr class="memdesc:a8cf76800f047841bf63d507fb1b54048"><td class="mdescLeft">&#160;</td><td class="mdescRight">a function to update the dictionnary that homoset is  <a href="#a8cf76800f047841bf63d507fb1b54048">More...</a><br /></td></tr>
<tr class="separator:a8cf76800f047841bf63d507fb1b54048"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af57de86adab3635645d684b89636cd5c"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#af57de86adab3635645d684b89636cd5c">plot_all</a> (self, With='tsne', perplexity=60, interactive=False, bins=100, offset=20, iteration=400, redo=False, bypasstsne=False, dotsize=7, inpar=True)</td></tr>
<tr class="memdesc:af57de86adab3635645d684b89636cd5c"><td class="mdescLeft">&#160;</td><td class="mdescRight">will plot all the homologies in the full_homo_matrix (and compute it)  <a href="#af57de86adab3635645d684b89636cd5c">More...</a><br /></td></tr>
<tr class="separator:af57de86adab3635645d684b89636cd5c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5029dbdf4e7f61bd38b6273f4fde9ead"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a5029dbdf4e7f61bd38b6273f4fde9ead">loadfullhomo</a> (self)</td></tr>
<tr class="memdesc:a5029dbdf4e7f61bd38b6273f4fde9ead"><td class="mdescLeft">&#160;</td><td class="mdescRight">function to concatenate all the homologies in one big array(practicle for certain computations)  <a href="#a5029dbdf4e7f61bd38b6273f4fde9ead">More...</a><br /></td></tr>
<tr class="separator:a5029dbdf4e7f61bd38b6273f4fde9ead"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5a56a99b5bf87afb418a1bbc5ec1c10b"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a5a56a99b5bf87afb418a1bbc5ec1c10b">loadhashomo</a> (self, withnames=False)</td></tr>
<tr class="memdesc:a5a56a99b5bf87afb418a1bbc5ec1c10b"><td class="mdescLeft">&#160;</td><td class="mdescRight">function to compute the matrix of bool saying wether species X has a gene or more in homology Y  <a href="#a5a56a99b5bf87afb418a1bbc5ec1c10b">More...</a><br /></td></tr>
<tr class="separator:a5a56a99b5bf87afb418a1bbc5ec1c10b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a000511647e3f85d8abadcdddaddfb04d"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a000511647e3f85d8abadcdddaddfb04d">size</a> (self)</td></tr>
<tr class="memdesc:a000511647e3f85d8abadcdddaddfb04d"><td class="mdescLeft">&#160;</td><td class="mdescRight">the size of the homoset (number of genes)  <a href="#a000511647e3f85d8abadcdddaddfb04d">More...</a><br /></td></tr>
<tr class="separator:a000511647e3f85d8abadcdddaddfb04d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a05935a3f03f2e2b61b2f3e7e0222e2a9"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a05935a3f03f2e2b61b2f3e7e0222e2a9">add_random_homology</a> (self)</td></tr>
<tr class="memdesc:a05935a3f03f2e2b61b2f3e7e0222e2a9"><td class="mdescLeft">&#160;</td><td class="mdescRight">a function to populate a homology with random values  <a href="#a05935a3f03f2e2b61b2f3e7e0222e2a9">More...</a><br /></td></tr>
<tr class="separator:a05935a3f03f2e2b61b2f3e7e0222e2a9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3f689e03251e7b4fe262a86d08946998"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a3f689e03251e7b4fe262a86d08946998">preprocessing</a> (self, withtaxons=False, withnames=True)</td></tr>
<tr class="memdesc:a3f689e03251e7b4fe262a86d08946998"><td class="mdescLeft">&#160;</td><td class="mdescRight">will compute the full list of names,  <a href="#a3f689e03251e7b4fe262a86d08946998">More...</a><br /></td></tr>
<tr class="separator:a3f689e03251e7b4fe262a86d08946998"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa39455620288b4912e5540491913e792"><td class="memItemLeft" align="right" valign="top"><a id="aa39455620288b4912e5540491913e792"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>preprocessing_taxons</b> (self)</td></tr>
<tr class="separator:aa39455620288b4912e5540491913e792"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a36cf71a7bc24b788c4d526591a8ec376"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a36cf71a7bc24b788c4d526591a8ec376">preprocessing_names</a> (self)</td></tr>
<tr class="memdesc:a36cf71a7bc24b788c4d526591a8ec376"><td class="mdescLeft">&#160;</td><td class="mdescRight">same as preprocessing_taxons() but admiting there is no taxon information (Yun's data for example)  <a href="#a36cf71a7bc24b788c4d526591a8ec376">More...</a><br /></td></tr>
<tr class="separator:a36cf71a7bc24b788c4d526591a8ec376"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4060c440127aa861c17b6157c81e2d27"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a4060c440127aa861c17b6157c81e2d27">preprocessing_namelist</a> (self)</td></tr>
<tr class="memdesc:a4060c440127aa861c17b6157c81e2d27"><td class="mdescLeft">&#160;</td><td class="mdescRight">same as <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a36cf71a7bc24b788c4d526591a8ec376" title="same as preprocessing_taxons() but admiting there is no taxon information (Yun&#39;s data for example) ...">preprocessing_names()</a> but without preprocessing the names of each homologies only updating the species_namelist  <a href="#a4060c440127aa861c17b6157c81e2d27">More...</a><br /></td></tr>
<tr class="separator:a4060c440127aa861c17b6157c81e2d27"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a59a0c4c70f6dbd2dd156e2b2c9b00421"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a59a0c4c70f6dbd2dd156e2b2c9b00421">compute_ages</a> (self, preserved=True, minpreserv=0.9, minsimi=0.85)</td></tr>
<tr class="memdesc:a59a0c4c70f6dbd2dd156e2b2c9b00421"><td class="mdescLeft">&#160;</td><td class="mdescRight">will compute whether or not a coding gene is highly preserved and if not how recent it is  <a href="#a59a0c4c70f6dbd2dd156e2b2c9b00421">More...</a><br /></td></tr>
<tr class="separator:a59a0c4c70f6dbd2dd156e2b2c9b00421"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a78ab39ddfa288661f5e809ea0a455187"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a78ab39ddfa288661f5e809ea0a455187">compute_entropyloc</a> (self, using='computejerem')</td></tr>
<tr class="memdesc:a78ab39ddfa288661f5e809ea0a455187"><td class="mdescLeft">&#160;</td><td class="mdescRight">called if need entropy location and used ensembl data.  <a href="#a78ab39ddfa288661f5e809ea0a455187">More...</a><br /></td></tr>
<tr class="separator:a78ab39ddfa288661f5e809ea0a455187"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af19ffd1cceb4b222e642713456b300a1"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#af19ffd1cceb4b222e642713456b300a1">remove</a> (self, species)</td></tr>
<tr class="memdesc:af19ffd1cceb4b222e642713456b300a1"><td class="mdescLeft">&#160;</td><td class="mdescRight">remove this list of species from the homologies  <a href="#af19ffd1cceb4b222e642713456b300a1">More...</a><br /></td></tr>
<tr class="separator:af19ffd1cceb4b222e642713456b300a1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2062c7f9c3c956d43172e1688504826a"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a2062c7f9c3c956d43172e1688504826a">clean_species</a> (self, thresh=0.3)</td></tr>
<tr class="memdesc:a2062c7f9c3c956d43172e1688504826a"><td class="mdescLeft">&#160;</td><td class="mdescRight">will remove a species from all the homologies of this homoset  <a href="#a2062c7f9c3c956d43172e1688504826a">More...</a><br /></td></tr>
<tr class="separator:a2062c7f9c3c956d43172e1688504826a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2b4bf6cef1b2545373be7ef11c329796"><td class="memItemLeft" align="right" valign="top"><a id="a2b4bf6cef1b2545373be7ef11c329796"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a2b4bf6cef1b2545373be7ef11c329796">plot_homoperspecies</a> (self)</td></tr>
<tr class="memdesc:a2b4bf6cef1b2545373be7ef11c329796"><td class="mdescLeft">&#160;</td><td class="mdescRight">will plot the number of homology per spcies in this homology group <br /></td></tr>
<tr class="separator:a2b4bf6cef1b2545373be7ef11c329796"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2cd71ff76ae8cd4ed1bdc206346942d4"><td class="memItemLeft" align="right" valign="top"><a id="a2cd71ff76ae8cd4ed1bdc206346942d4"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a2cd71ff76ae8cd4ed1bdc206346942d4">plot_speciesperhomo</a> (self)</td></tr>
<tr class="memdesc:a2cd71ff76ae8cd4ed1bdc206346942d4"><td class="mdescLeft">&#160;</td><td class="mdescRight">will plot the number of species per homologies in this homology group <br /></td></tr>
<tr class="separator:a2cd71ff76ae8cd4ed1bdc206346942d4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2d543ce742d6a4e968e14b1471319287"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a2d543ce742d6a4e968e14b1471319287">cluster_homologies</a> (self, clustering='kmeans', byspecie=False, order=True, plot_ordering=True, homogroupnb=2, findnb=False)</td></tr>
<tr class="memdesc:a2d543ce742d6a4e968e14b1471319287"><td class="mdescLeft">&#160;</td><td class="mdescRight">Compute an homology group :  <a href="#a2d543ce742d6a4e968e14b1471319287">More...</a><br /></td></tr>
<tr class="separator:a2d543ce742d6a4e968e14b1471319287"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abebd6a9804b73a10fc726974630e5d26"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#abebd6a9804b73a10fc726974630e5d26">orderfromclust</a> (self, homogroupnb, clust, byspecie=False, plot_ordering=True)</td></tr>
<tr class="memdesc:abebd6a9804b73a10fc726974630e5d26"><td class="mdescLeft">&#160;</td><td class="mdescRight">creates an ordering of every elements  <a href="#abebd6a9804b73a10fc726974630e5d26">More...</a><br /></td></tr>
<tr class="separator:abebd6a9804b73a10fc726974630e5d26"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa41e2430673cf29f96a522e03040ca1d"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#aa41e2430673cf29f96a522e03040ca1d">get_clusterstats</a> (self, sort=True, interactive=True, redo=False)</td></tr>
<tr class="memdesc:aa41e2430673cf29f96a522e03040ca1d"><td class="mdescLeft">&#160;</td><td class="mdescRight">will find the number of cluster i per homologies and per species  <a href="#aa41e2430673cf29f96a522e03040ca1d">More...</a><br /></td></tr>
<tr class="separator:aa41e2430673cf29f96a522e03040ca1d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae429c30197127c5d6248c94dbb6dfc46"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#ae429c30197127c5d6248c94dbb6dfc46">compare_clusters</a> (self, cubdistance_matrix=True, plot=True, interactive=True, <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a000511647e3f85d8abadcdddaddfb04d">size</a>=40)</td></tr>
<tr class="memdesc:ae429c30197127c5d6248c94dbb6dfc46"><td class="mdescLeft">&#160;</td><td class="mdescRight">for each clusters in homologies, will compare them with a similarity matrix and a distance matrix  <a href="#ae429c30197127c5d6248c94dbb6dfc46">More...</a><br /></td></tr>
<tr class="separator:ae429c30197127c5d6248c94dbb6dfc46"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a81b40822ff0d1d3e0cfccd6c3c54cdd6"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a81b40822ff0d1d3e0cfccd6c3c54cdd6">find_clusters</a> (self, clustering='dbscan', homogroupnb=None, assess=True, eps=0.8, best_eps=True, trainingset=30, epoch=20, ranges=[0.2, <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a000511647e3f85d8abadcdddaddfb04d">size</a>=10, redo=False)</td></tr>
<tr class="memdesc:a81b40822ff0d1d3e0cfccd6c3c54cdd6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Finds, for each homologies in the working homoset, groups that are part of compact clusters.  <a href="#a81b40822ff0d1d3e0cfccd6c3c54cdd6">More...</a><br /></td></tr>
<tr class="separator:a81b40822ff0d1d3e0cfccd6c3c54cdd6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a36e6e64142f4cc63425f145a1754d1d6"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a36e6e64142f4cc63425f145a1754d1d6">findbest_eps</a> (self, trainingset=400, clustering=&quot;dbscan&quot;, epoch=20, ranges=[0.2, <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a000511647e3f85d8abadcdddaddfb04d">size</a>=10, redo=False)</td></tr>
<tr class="memdesc:a36e6e64142f4cc63425f145a1754d1d6"><td class="mdescLeft">&#160;</td><td class="mdescRight">will find the best eps hyperparameter (the one that minimizes the evolutionary distance within its clusters)  <a href="#a36e6e64142f4cc63425f145a1754d1d6">More...</a><br /></td></tr>
<tr class="separator:a36e6e64142f4cc63425f145a1754d1d6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8ab62bae7be17abf4b13ad37d5c881eb"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a8ab62bae7be17abf4b13ad37d5c881eb">plot_hashomo</a> (self, invert=False, <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a000511647e3f85d8abadcdddaddfb04d">size</a>=40, interactive=False, rangeto=None)</td></tr>
<tr class="memdesc:a8ab62bae7be17abf4b13ad37d5c881eb"><td class="mdescLeft">&#160;</td><td class="mdescRight">plot the has homo matrix  <a href="#a8ab62bae7be17abf4b13ad37d5c881eb">More...</a><br /></td></tr>
<tr class="separator:a8ab62bae7be17abf4b13ad37d5c881eb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a87fcd4602c943c98543e2bc552502663"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a87fcd4602c943c98543e2bc552502663">plot_simiclust</a> (self, interactive=True, <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a000511647e3f85d8abadcdddaddfb04d">size</a>=40)</td></tr>
<tr class="memdesc:a87fcd4602c943c98543e2bc552502663"><td class="mdescLeft">&#160;</td><td class="mdescRight">plot the similarity matrix of each homologies from its clusters  <a href="#a87fcd4602c943c98543e2bc552502663">More...</a><br /></td></tr>
<tr class="separator:a87fcd4602c943c98543e2bc552502663"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa385d8acea32381efae12c219c24a331"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#aa385d8acea32381efae12c219c24a331">plot_distcub</a> (self, interactive=False, <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a000511647e3f85d8abadcdddaddfb04d">size</a>=40)</td></tr>
<tr class="memdesc:aa385d8acea32381efae12c219c24a331"><td class="mdescLeft">&#160;</td><td class="mdescRight">plot the distance matrix of each homologies from the average of their CUB distances  <a href="#aa385d8acea32381efae12c219c24a331">More...</a><br /></td></tr>
<tr class="separator:aa385d8acea32381efae12c219c24a331"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-attribs"></a>
Public Attributes</h2></td></tr>
<tr class="memitem:acbfe58dd06d4c6ca61a528ddfabee3d6"><td class="memItemLeft" align="right" valign="top"><a id="acbfe58dd06d4c6ca61a528ddfabee3d6"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>homo_namelist</b></td></tr>
<tr class="separator:acbfe58dd06d4c6ca61a528ddfabee3d6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a141708f3582bf009ec08b111124589b1"><td class="memItemLeft" align="right" valign="top"><a id="a141708f3582bf009ec08b111124589b1"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>species_namelist</b></td></tr>
<tr class="separator:a141708f3582bf009ec08b111124589b1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7e0bb4c43fb89bb5165f509048a5025b"><td class="memItemLeft" align="right" valign="top"><a id="a7e0bb4c43fb89bb5165f509048a5025b"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>homogroupnb</b></td></tr>
<tr class="separator:a7e0bb4c43fb89bb5165f509048a5025b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1e216b6dd586b4d71d48db2bd9e82fd1"><td class="memItemLeft" align="right" valign="top"><a id="a1e216b6dd586b4d71d48db2bd9e82fd1"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>clusters</b></td></tr>
<tr class="separator:a1e216b6dd586b4d71d48db2bd9e82fd1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7b5672c7d4faf1bec55d7afffc13ff2a"><td class="memItemLeft" align="right" valign="top"><a id="a7b5672c7d4faf1bec55d7afffc13ff2a"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>datatype</b></td></tr>
<tr class="separator:a7b5672c7d4faf1bec55d7afffc13ff2a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a28336c404d04628639692b2254c4c7f4"><td class="memItemLeft" align="right" valign="top"><a id="a28336c404d04628639692b2254c4c7f4"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>phylo_distances</b></td></tr>
<tr class="separator:a28336c404d04628639692b2254c4c7f4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5835f6f6921772dd599c4deee9e01b35"><td class="memItemLeft" align="right" valign="top"><a id="a5835f6f6921772dd599c4deee9e01b35"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>best_eps</b></td></tr>
<tr class="separator:a5835f6f6921772dd599c4deee9e01b35"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ade27fdf472815f4fb2eb1d1ef552ad42"><td class="memItemLeft" align="right" valign="top"><a id="ade27fdf472815f4fb2eb1d1ef552ad42"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>wasclusterized</b></td></tr>
<tr class="separator:ade27fdf472815f4fb2eb1d1ef552ad42"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a25c653d5b412d60a7c3663e327b222cf"><td class="memItemLeft" align="right" valign="top"><a id="a25c653d5b412d60a7c3663e327b222cf"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>stats</b></td></tr>
<tr class="separator:a25c653d5b412d60a7c3663e327b222cf"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad09211d828ad602a1e3806829eb96f04"><td class="memItemLeft" align="right" valign="top"><a id="ad09211d828ad602a1e3806829eb96f04"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>homodict</b></td></tr>
<tr class="separator:ad09211d828ad602a1e3806829eb96f04"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae0885604920c67cd0bf29e93a7cf4a5a"><td class="memItemLeft" align="right" valign="top"><a id="ae0885604920c67cd0bf29e93a7cf4a5a"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>cluster_similarity</b></td></tr>
<tr class="separator:ae0885604920c67cd0bf29e93a7cf4a5a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a66fe1cab79f6c84cd93ba95a9b233f0e"><td class="memItemLeft" align="right" valign="top"><a id="a66fe1cab79f6c84cd93ba95a9b233f0e"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>cub_similarity</b></td></tr>
<tr class="separator:a66fe1cab79f6c84cd93ba95a9b233f0e"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-static-attribs"></a>
Static Public Attributes</h2></td></tr>
<tr class="memitem:a7c35df6a1db3f4b64a91be2417610c9e"><td class="memItemLeft" align="right" valign="top"><a id="a7c35df6a1db3f4b64a91be2417610c9e"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>hashomo_matrix</b> = None</td></tr>
<tr class="separator:a7c35df6a1db3f4b64a91be2417610c9e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a092f90e457bce969aa9c012cd971fbb4"><td class="memItemLeft" align="right" valign="top"><a id="a092f90e457bce969aa9c012cd971fbb4"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>homo_matrix</b> = None</td></tr>
<tr class="separator:a092f90e457bce969aa9c012cd971fbb4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afb5e078e3db98fc9e1383a5f542fc9af"><td class="memItemLeft" align="right" valign="top"><a id="afb5e078e3db98fc9e1383a5f542fc9af"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>homo_matrixnames</b> = None</td></tr>
<tr class="separator:afb5e078e3db98fc9e1383a5f542fc9af"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a44e879b2c7395a60f106e22e1df04d69"><td class="memItemLeft" align="right" valign="top"><a id="a44e879b2c7395a60f106e22e1df04d69"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>fulleng</b> = None</td></tr>
<tr class="separator:a44e879b2c7395a60f106e22e1df04d69"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab759c3c5aedf63d26a9eb767ee351b36"><td class="memItemLeft" align="right" valign="top"><a id="ab759c3c5aedf63d26a9eb767ee351b36"></a>
dictionary&#160;</td><td class="memItemRight" valign="bottom"><b>homodict</b> = {}</td></tr>
<tr class="separator:ab759c3c5aedf63d26a9eb767ee351b36"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4c886336a64e5bf641ab01ef71aa0cfe"><td class="memItemLeft" align="right" valign="top"><a id="a4c886336a64e5bf641ab01ef71aa0cfe"></a>
list&#160;</td><td class="memItemRight" valign="bottom"><b>homo_namelist</b> = []</td></tr>
<tr class="separator:a4c886336a64e5bf641ab01ef71aa0cfe"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aead1ed5c51a2567ce8b0c5f05370e375"><td class="memItemLeft" align="right" valign="top"><a id="aead1ed5c51a2567ce8b0c5f05370e375"></a>
list&#160;</td><td class="memItemRight" valign="bottom"><b>species_namelist</b> = []</td></tr>
<tr class="separator:aead1ed5c51a2567ce8b0c5f05370e375"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aea3ec877da461f16485a2f4a0fb4b4e5"><td class="memItemLeft" align="right" valign="top"><a id="aea3ec877da461f16485a2f4a0fb4b4e5"></a>
list&#160;</td><td class="memItemRight" valign="bottom"><b>clusters</b> = []</td></tr>
<tr class="separator:aea3ec877da461f16485a2f4a0fb4b4e5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8139ce28113d3f89df1c2d33491ce387"><td class="memItemLeft" align="right" valign="top"><a id="a8139ce28113d3f89df1c2d33491ce387"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>homogroupnb</b> = 2</td></tr>
<tr class="separator:a8139ce28113d3f89df1c2d33491ce387"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abead19765d92567bdc6d8d2892b72436"><td class="memItemLeft" align="right" valign="top"><a id="abead19765d92567bdc6d8d2892b72436"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>red_homomatrix</b> = None</td></tr>
<tr class="separator:abead19765d92567bdc6d8d2892b72436"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9437a1fb1e9cd3fd024d89ea7869c26f"><td class="memItemLeft" align="right" valign="top"><a id="a9437a1fb1e9cd3fd024d89ea7869c26f"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><b>wasclusterized</b> = False</td></tr>
<tr class="separator:a9437a1fb1e9cd3fd024d89ea7869c26f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5e37288c128509ca765ac3dfa40d24bc"><td class="memItemLeft" align="right" valign="top"><a id="a5e37288c128509ca765ac3dfa40d24bc"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>homo_clusters</b> = None</td></tr>
<tr class="separator:a5e37288c128509ca765ac3dfa40d24bc"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa0852fc13f3dee9f3f2ebda89a3a4b93"><td class="memItemLeft" align="right" valign="top"><a id="aa0852fc13f3dee9f3f2ebda89a3a4b93"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>datatype</b> = ''</td></tr>
<tr class="separator:aa0852fc13f3dee9f3f2ebda89a3a4b93"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8cb746f5ac4c3fb02bdfa1d7dc996727"><td class="memItemLeft" align="right" valign="top"><a id="a8cb746f5ac4c3fb02bdfa1d7dc996727"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>averagehomo_matrix</b> = None</td></tr>
<tr class="separator:a8cb746f5ac4c3fb02bdfa1d7dc996727"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a64970a91783342ed77e4bc577d6c0e1f"><td class="memItemLeft" align="right" valign="top"><a id="a64970a91783342ed77e4bc577d6c0e1f"></a>
dictionary&#160;</td><td class="memItemRight" valign="bottom"><b>stats</b> = {}</td></tr>
<tr class="separator:a64970a91783342ed77e4bc577d6c0e1f"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pri-methods"></a>
Private Member Functions</h2></td></tr>
<tr class="memitem:af6e8927ae33e2785ab21293a59628946"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#af6e8927ae33e2785ab21293a59628946">_barplot</a> (self, interactive=True)</td></tr>
<tr class="memdesc:af6e8927ae33e2785ab21293a59628946"><td class="mdescLeft">&#160;</td><td class="mdescRight">called by statistics function to plot a barplot of the proportion of different cluster per homologies and per species  <a href="#af6e8927ae33e2785ab21293a59628946">More...</a><br /></td></tr>
<tr class="separator:af6e8927ae33e2785ab21293a59628946"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afd51ea6276ecae41960d426ce72755b0"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#afd51ea6276ecae41960d426ce72755b0">_dictify</a> (self, savehomos=False)</td></tr>
<tr class="memdesc:afd51ea6276ecae41960d426ce72755b0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Used by the saving function.  <a href="#afd51ea6276ecae41960d426ce72755b0">More...</a><br /></td></tr>
<tr class="separator:afd51ea6276ecae41960d426ce72755b0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a795cfd3187026bde3c065aab9fe09642"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a795cfd3187026bde3c065aab9fe09642">_plot_clust</a> (self, mat, orderedmat)</td></tr>
<tr class="memdesc:a795cfd3187026bde3c065aab9fe09642"><td class="mdescLeft">&#160;</td><td class="mdescRight">will plot the correlation matrix of the has_homomatrix before and after ordering allowing one to show its effect  <a href="#a795cfd3187026bde3c065aab9fe09642">More...</a><br /></td></tr>
<tr class="separator:a795cfd3187026bde3c065aab9fe09642"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p><a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html" title="HomoSet is the object containing evrey homology as a dictionnary according to thie rhomology code...">HomoSet</a> is the object containing evrey homology as a dictionnary according to thie rhomology code. </p>
<p>from Homoset you can do much computation that requires the set of homologies Object where we store an homology group basically where we do our entire Computation from. Be aware that even if you use str, the keys will be stored as unicode as jsonized dict will output unicode in python &lt; 3</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">hashomo_matrix</td><td>a np.array[boolean] (species, homologies) that store the matrix of gene presence in species </td></tr>
    <tr><td class="paramname">homo_matrix</td><td>np.array[float] (homologies*species(inthehomologies),aminoacids) but containing the codon entropy vectors instead </td></tr>
    <tr><td class="paramname">homodict</td><td>dictionnary of homology object containing codon usage per species for the gene coresponding to this homology </td></tr>
    <tr><td class="paramname">homo_namelist</td><td>list[str] of all the homology names </td></tr>
    <tr><td class="paramname">species_namelist</td><td>list[str] of all the species names </td></tr>
    <tr><td class="paramname">clusters</td><td>list[int] of clusters for the homoset clustering can be of size of nb of species or of nb of homologies </td></tr>
    <tr><td class="paramname">homogroupnb</td><td>int of group for the homology clustering </td></tr>
    <tr><td class="paramname">red_homomatrix</td><td>np.array[float] (homologies*species(inthehomologies),x*y)the reduced 2D version of the homology matrix for the full homology matrix </td></tr>
    <tr><td class="paramname">wasclusterized</td><td>bool if the homologies have been clustered or not. usefull for processing requiring clusters </td></tr>
    <tr><td class="paramname">homocluster</td><td>np.array[int] homologies*species of int of cluster number </td></tr>
    <tr><td class="paramname">homo_matrixnames</td><td>np.array[int] corresponding to names in PyCUB.utils.speciestable </td></tr>
    <tr><td class="paramname">fulleng</td><td>np.array[int] number of coding for each amino for each gene for each homology </td></tr>
    <tr><td class="paramname">datatype</td><td>str of the CUB value type [entropy, A_value(entropyLocation), frequency] </td></tr>
    <tr><td class="paramname">averagehomo_matrix</td><td>np.array[float] of the avg CUB values per homologies </td></tr>
    <tr><td class="paramname">stats</td><td>dict of statistics on the clusterings (see <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#aa41e2430673cf29f96a522e03040ca1d" title="will find the number of cluster i per homologies and per species ">get_clusterstats()</a>) </td></tr>
  </table>
  </dd>
</dl>
</div><h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a id="adbe4d9e4087fe5b53d861476c011329d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#adbe4d9e4087fe5b53d861476c011329d">&#9670;&nbsp;</a></span>__init__()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homoset.HomoSet.__init__ </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>kwargs</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>will initialize the object with the different values you might have from another project use the data dictionnary to add any type of data </p>
<p>a dictionary to any values present in the homoset </p>

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a id="ab7ac33ea3033b440816e179cb4a5e3c0"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab7ac33ea3033b440816e179cb4a5e3c0">&#9670;&nbsp;</a></span>__getitem__()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homoset.HomoSet.__getitem__ </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>key</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>get the homology at the corresponding key </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">key</td><td>str, unicode the key at which to get the homology</td></tr>
  </table>
  </dd>
</dl>
<pre class="fragment">        TypeError "the type you should enter is int or unicode or str"
</pre> <dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">KeyError</td><td></td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#adbe4d9e4087fe5b53d861476c011329d">PyCUB.homoset.HomoSet.__init__()</a>.</p>

</div>
</div>
<a id="a42a7e7dc7ddb4b5717084cb093f90774"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a42a7e7dc7ddb4b5717084cb093f90774">&#9670;&nbsp;</a></span>__setitem__()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homoset.HomoSet.__setitem__ </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>key</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>value</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>add an item at the corresponding key </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">key</td><td>str, unicode the key at which to add </td></tr>
    <tr><td class="paramname">val</td><td>PyCUB.homology to add at this key</td></tr>
  </table>
  </dd>
</dl>
<pre class="fragment">        TypeError "the type you should enter is int or unicode or str"
</pre> <dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">KeyError</td><td></td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#ab7ac33ea3033b440816e179cb4a5e3c0">PyCUB.homoset.HomoSet.__getitem__()</a>.</p>

</div>
</div>
<a id="af6e8927ae33e2785ab21293a59628946"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af6e8927ae33e2785ab21293a59628946">&#9670;&nbsp;</a></span>_barplot()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homoset.HomoSet._barplot </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>interactive</em> = <code>True</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>called by statistics function to plot a barplot of the proportion of different cluster per homologies and per species </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">interactive</td><td>bool to true if you want to use the bokeh interactive version </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the barplot holoviews object (will directly render if in a notebook) </dd></dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a36e6e64142f4cc63425f145a1754d1d6">PyCUB.homoset.HomoSet.findbest_eps()</a>, and <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#aa41e2430673cf29f96a522e03040ca1d">PyCUB.homoset.HomoSet.get_clusterstats()</a>.</p>

</div>
</div>
<a id="afd51ea6276ecae41960d426ce72755b0"></a>
<h2 class="memtitle"><span class="permalink"><a href="#afd51ea6276ecae41960d426ce72755b0">&#9670;&nbsp;</a></span>_dictify()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homoset.HomoSet._dictify </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>savehomos</em> = <code>False</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Used by the saving function. </p>
<p>transform the object into a dictionary that can be json serializable</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">savehomos</td><td>bool to true if you consider this homology as containing all the information (the other ones only have references to a subset of the data of this one)</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the dictionnary of all the data in this Object in the correct format to be jsonized </dd></dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#af6e8927ae33e2785ab21293a59628946">PyCUB.homoset.HomoSet._barplot()</a>, and <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a477c55065989b10bce8389c5edaf3322">PyCUB.pyCUB.PyCUB.save()</a>.</p>

</div>
</div>
<a id="a795cfd3187026bde3c065aab9fe09642"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a795cfd3187026bde3c065aab9fe09642">&#9670;&nbsp;</a></span>_plot_clust()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homoset.HomoSet._plot_clust </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>mat</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>orderedmat</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>will plot the correlation matrix of the has_homomatrix before and after ordering allowing one to show its effect </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">mat</td><td>np.array[bool] the current homology matrix </td></tr>
    <tr><td class="paramname">orderedmat</td><td>np.array[bool] the ordered homology matrix </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#abebd6a9804b73a10fc726974630e5d26">PyCUB.homoset.HomoSet.orderfromclust()</a>, and <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#aa385d8acea32381efae12c219c24a331">PyCUB.homoset.HomoSet.plot_distcub()</a>.</p>

</div>
</div>
<a id="a05935a3f03f2e2b61b2f3e7e0222e2a9"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a05935a3f03f2e2b61b2f3e7e0222e2a9">&#9670;&nbsp;</a></span>add_random_homology()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homoset.HomoSet.add_random_homology </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>a function to populate a homology with random values </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">None</td><td></td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the name of the random homology (str) </dd></dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a000511647e3f85d8abadcdddaddfb04d">PyCUB.homoset.HomoSet.size()</a>.</p>

</div>
</div>
<a id="a2062c7f9c3c956d43172e1688504826a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a2062c7f9c3c956d43172e1688504826a">&#9670;&nbsp;</a></span>clean_species()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homoset.HomoSet.clean_species </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>thresh</em> = <code>0.3</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>will remove a species from all the homologies of this homoset </p>
<p>Warning, as the all/working homosets share the ref to homologies, deleting some species in working_homoset will result in removing some species in some homologies of all_homoset</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">thresh</td><td>float the threshold of avg presence in homologies below which the species are remove </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#af19ffd1cceb4b222e642713456b300a1">PyCUB.homoset.HomoSet.remove()</a>.</p>

</div>
</div>
<a id="a2d543ce742d6a4e968e14b1471319287"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a2d543ce742d6a4e968e14b1471319287">&#9670;&nbsp;</a></span>cluster_homologies()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homoset.HomoSet.cluster_homologies </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>clustering</em> = <code>'kmeans'</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>byspecie</em> = <code>False</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>order</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>plot_ordering</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>homogroupnb</em> = <code>2</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>findnb</em> = <code>False</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Compute an homology group : </p>
<p>from matrix computation using the homo_matrix (or from network computation in homologize_from_network) Can be computed many times and will updata homoset with the most recent homoset found if homoset exists, it will save it.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">clustering</td><td>str flags to 'kmeans', 'kmodes', 'fast' to use different sk-learn algorithms </td></tr>
    <tr><td class="paramname">plot</td><td>bool flags to true for the function to output ploting of the affinity matrix with and without the clusters </td></tr>
    <tr><td class="paramname">homogroupnb</td><td>int nb of groups you want to extract </td></tr>
    <tr><td class="paramname">byspecie</td><td>bool to true if we cluster by species instead of homologies </td></tr>
    <tr><td class="paramname">order</td><td>bool whether or not to order </td></tr>
    <tr><td class="paramname">plot_ordering</td><td>bool to true to plot this ordering </td></tr>
    <tr><td class="paramname">findnb</td><td>bool to true to find the right number of clusters (homogroupnb)</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>if findnb, will return the clusters for each homogroupnb up to 9</dd></dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">AttributeError</td><td>"you entered a wrong clustering algorithm" </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a2cd71ff76ae8cd4ed1bdc206346942d4">PyCUB.homoset.HomoSet.plot_speciesperhomo()</a>.</p>

</div>
</div>
<a id="ae429c30197127c5d6248c94dbb6dfc46"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae429c30197127c5d6248c94dbb6dfc46">&#9670;&nbsp;</a></span>compare_clusters()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homoset.HomoSet.compare_clusters </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>cubdistance_matrix</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>plot</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>interactive</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>size</em> = <code>40</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>for each clusters in homologies, will compare them with a similarity matrix and a distance matrix </p>
<p>compare amongst the working homoset homologies, the clusters together, by what species they contains by creating a new vector of species presence in each cluster and plotting the similarity matrix of each of those vectors. &ndash;&gt; create a compare function in homoset of homologies clusters similarity matrix and ordering. basically the distance should be nan if it has not the species, -1 if outlier to other and 1 if one cluster to another and zeros if the same to the same</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">cubdistance_matrix</td><td>bool to true if want to compute the matrix of the averageCUB value distances summed for each cluster amongst the homologies </td></tr>
    <tr><td class="paramname">plot</td><td>bool to true to plot </td></tr>
    <tr><td class="paramname">size</td><td>int the size of the plot </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#aa41e2430673cf29f96a522e03040ca1d">PyCUB.homoset.HomoSet.get_clusterstats()</a>.</p>

</div>
</div>
<a id="a59a0c4c70f6dbd2dd156e2b2c9b00421"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a59a0c4c70f6dbd2dd156e2b2c9b00421">&#9670;&nbsp;</a></span>compute_ages()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homoset.HomoSet.compute_ages </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>preserved</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>minpreserv</em> = <code>0.9</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>minsimi</em> = <code>0.85</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>will compute whether or not a coding gene is highly preserved and if not how recent it is </p>
<p>it uses the phylogenetic distances and the similarities amongst homlogies to try to find a good proxy</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">preserved</td><td>bool to true if we should find highly preserved genes or not </td></tr>
    <tr><td class="paramname">minpreserv</td><td>float minimal percentage of homologous species that have this homology </td></tr>
    <tr><td class="paramname">minsimi</td><td>float minimal avg similarity between genes to consider them highly preserved</td></tr>
  </table>
  </dd>
</dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">UnboundLocalError</td><td>"you need to have the similarity_scores" </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a4060c440127aa861c17b6157c81e2d27">PyCUB.homoset.HomoSet.preprocessing_namelist()</a>.</p>

</div>
</div>
<a id="a78ab39ddfa288661f5e809ea0a455187"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a78ab39ddfa288661f5e809ea0a455187">&#9670;&nbsp;</a></span>compute_entropyloc()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homoset.HomoSet.compute_entropyloc </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>using</em> = <code>'computejerem'</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>called if need entropy location and used ensembl data. </p>
<p>you can always compute entropy location from entropy data. </p><pre class="fragment">   Will be much faster than doing it directly when calling
   ensembl's data as it computes the partition function
   only one for each lengths
</pre><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">using</td><td>str flags the partition algorithm to use</td></tr>
  </table>
  </dd>
</dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">UnboundLocalError</td><td>"you need to have your CUB values as entropy" </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a59a0c4c70f6dbd2dd156e2b2c9b00421">PyCUB.homoset.HomoSet.compute_ages()</a>.</p>

</div>
</div>
<a id="a81b40822ff0d1d3e0cfccd6c3c54cdd6"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a81b40822ff0d1d3e0cfccd6c3c54cdd6">&#9670;&nbsp;</a></span>find_clusters()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homoset.HomoSet.find_clusters </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>clustering</em> = <code>'dbscan'</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>homogroupnb</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>assess</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>eps</em> = <code>0.8</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>best_eps</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>trainingset</em> = <code>30</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>epoch</em> = <code>20</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>ranges</em> = <code>[0.2</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>size</em> = <code>10</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>redo</em> = <code>False</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Finds, for each homologies in the working homoset, groups that are part of compact clusters. </p>
<p>it will be using gaussian mixture clustering or DBSCAN and order them according to the density of each cluster (we are interested in the densest ones) and assess the quality using 3 criterion:BIC, number of outliers, also find if we are close to ancestry tree, here we need to represent a comparison of the closeness in a phylogenetic tree to a cluster of species &ndash;&gt; given a grouping of phylogenetic tree, what cluster is the most similar to it</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">clustering</td><td>method (DBSCAN, gaussian mixture) </td></tr>
    <tr><td class="paramname">homogroupnb</td><td>the number of groups can be a number or else will look for the better number of cluster according to assessments. </td></tr>
    <tr><td class="paramname">assess</td><td>plot or not the assessments </td></tr>
    <tr><td class="paramname">eps</td><td>the hyperparams </td></tr>
    <tr><td class="paramname">best_eps</td><td>bool to true wether or not to do an hyperparam greedy search </td></tr>
    <tr><td class="paramname">trainingset</td><td>int if best_eps, the size of the training ser </td></tr>
    <tr><td class="paramname">epoch</td><td>int the number of increasing trials </td></tr>
    <tr><td class="paramname">ranges</td><td>tuple[float] the two min and max values to use </td></tr>
    <tr><td class="paramname">size</td><td>int the x size of the plot </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#ae429c30197127c5d6248c94dbb6dfc46">PyCUB.homoset.HomoSet.compare_clusters()</a>.</p>

</div>
</div>
<a id="a36e6e64142f4cc63425f145a1754d1d6"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a36e6e64142f4cc63425f145a1754d1d6">&#9670;&nbsp;</a></span>findbest_eps()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homoset.HomoSet.findbest_eps </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>trainingset</em> = <code>400</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>clustering</em> = <code>&quot;dbscan&quot;</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>epoch</em> = <code>20</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>ranges</em> = <code>[0.2</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>size</em> = <code>10</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>redo</em> = <code>False</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>will find the best eps hyperparameter (the one that minimizes the evolutionary distance within its clusters) </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">trainingset</td><td>int the number of homologies in your training set (should be 20% of the total) </td></tr>
    <tr><td class="paramname">clustering</td><td>str flags the clustering algorithm from which to find the best hyperparam </td></tr>
    <tr><td class="paramname">epoch</td><td>int the number of increasing trials </td></tr>
    <tr><td class="paramname">ranges</td><td>tuple[float] the two min and max values to use </td></tr>
    <tr><td class="paramname">size</td><td>int the x size of the plot </td></tr>
    <tr><td class="paramname">redo</td><td>wether or not to recompute everyhting if it has already been computed once</td></tr>
  </table>
  </dd>
</dl>
<div class="fragment"><div class="line">Returns;</div></div><!-- fragment --><p> the best value (often a float)</p>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">LookupError</td><td>"you need to compute the phylogenetic distances first to have some form of labels" </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a81b40822ff0d1d3e0cfccd6c3c54cdd6">PyCUB.homoset.HomoSet.find_clusters()</a>.</p>

</div>
</div>
<a id="aa41e2430673cf29f96a522e03040ca1d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aa41e2430673cf29f96a522e03040ca1d">&#9670;&nbsp;</a></span>get_clusterstats()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homoset.HomoSet.get_clusterstats </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>sort</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>interactive</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>redo</em> = <code>False</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>will find the number of cluster i per homologies and per species </p>
<p>plot for each species, how much its genes are outliers, how much are belonging to a secondary cluster and how much are belonging to the principal cluster. &ndash;&gt; create a long stacked bar plot with these values</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">sort</td><td>bool to true to sort everyhting according to the statistics differences </td></tr>
    <tr><td class="paramname">interactive</td><td>bool to true to have an interactive barplot </td></tr>
    <tr><td class="paramname">redo</td><td>bool to true not to reused cached data</td></tr>
  </table>
  </dd>
</dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">UnboundLocalError</td><td>"you need to find clusters first" </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#abebd6a9804b73a10fc726974630e5d26">PyCUB.homoset.HomoSet.orderfromclust()</a>.</p>

</div>
</div>
<a id="a5029dbdf4e7f61bd38b6273f4fde9ead"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a5029dbdf4e7f61bd38b6273f4fde9ead">&#9670;&nbsp;</a></span>loadfullhomo()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homoset.HomoSet.loadfullhomo </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>function to concatenate all the homologies in one big array(practicle for certain computations) </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">None</td><td></td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a78ab39ddfa288661f5e809ea0a455187">PyCUB.homoset.HomoSet.compute_entropyloc()</a>, and <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#af57de86adab3635645d684b89636cd5c">PyCUB.homoset.HomoSet.plot_all()</a>.</p>

</div>
</div>
<a id="a5a56a99b5bf87afb418a1bbc5ec1c10b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a5a56a99b5bf87afb418a1bbc5ec1c10b">&#9670;&nbsp;</a></span>loadhashomo()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homoset.HomoSet.loadhashomo </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>withnames</em> = <code>False</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>function to compute the matrix of bool saying wether species X has a gene or more in homology Y </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">None</td><td></td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a5029dbdf4e7f61bd38b6273f4fde9ead">PyCUB.homoset.HomoSet.loadfullhomo()</a>, and <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a000511647e3f85d8abadcdddaddfb04d">PyCUB.homoset.HomoSet.size()</a>.</p>

</div>
</div>
<a id="abebd6a9804b73a10fc726974630e5d26"></a>
<h2 class="memtitle"><span class="permalink"><a href="#abebd6a9804b73a10fc726974630e5d26">&#9670;&nbsp;</a></span>orderfromclust()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homoset.HomoSet.orderfromclust </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>homogroupnb</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>clust</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>byspecie</em> = <code>False</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>plot_ordering</em> = <code>True</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>creates an ordering of every elements </p>
<p>(names, homologies according to the found clusters) from an ordered cluster of species or of homologies self.hashomomatrix should reflect this orientation as well</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">homogroupnb</td><td>int the number of clusters </td></tr>
    <tr><td class="paramname">clust</td><td>np.array(int) the clusters </td></tr>
    <tr><td class="paramname">byspecie</td><td>bool to true to order by species </td></tr>
    <tr><td class="paramname">plot_ordering</td><td>bool to true to plot the ordering</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the ordered hashomomatrix (np.array[bool]) </dd></dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a2d543ce742d6a4e968e14b1471319287">PyCUB.homoset.HomoSet.cluster_homologies()</a>.</p>

</div>
</div>
<a id="af57de86adab3635645d684b89636cd5c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af57de86adab3635645d684b89636cd5c">&#9670;&nbsp;</a></span>plot_all()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homoset.HomoSet.plot_all </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>With</em> = <code>'tsne'</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>perplexity</em> = <code>60</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>interactive</em> = <code>False</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>bins</em> = <code>100</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>offset</em> = <code>20</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>iteration</em> = <code>400</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>redo</em> = <code>False</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>bypasstsne</em> = <code>False</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>dotsize</em> = <code>7</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>inpar</em> = <code>True</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>will plot all the homologies in the full_homo_matrix (and compute it) </p>
<p>(sometimes around 800 000 datapoints) to look at any kind of relationships as there is too much datapoints, the plots are density ones.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">With</td><td>flag the dim reduction algorithm to use (tsne: need &gt;16gigs of RAM,now use another version of tsne for large datasets )(PCA: works well)(lsta/hessian:untested) </td></tr>
    <tr><td class="paramname">perplexity</td><td>ints of basic tsne hyperparams </td></tr>
    <tr><td class="paramname">interactive</td><td>bool if true should use bokeh else matplotlib</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the desired plot if the size is high and we are interactive</dd></dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">AttributeError</td><td>"not a good red algorithm" </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a8cf76800f047841bf63d507fb1b54048">PyCUB.homoset.HomoSet.update()</a>.</p>

</div>
</div>
<a id="aa385d8acea32381efae12c219c24a331"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aa385d8acea32381efae12c219c24a331">&#9670;&nbsp;</a></span>plot_distcub()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homoset.HomoSet.plot_distcub </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>interactive</em> = <code>False</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>size</em> = <code>40</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>plot the distance matrix of each homologies from the average of their CUB distances </p>
<p>the interactive version allows you to see each particular datapoint with much more precision</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">interactive</td><td>bool to use the bokeh interactive version </td></tr>
    <tr><td class="paramname">size</td><td>size of the matrix </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#ae429c30197127c5d6248c94dbb6dfc46">PyCUB.homoset.HomoSet.compare_clusters()</a>, and <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a87fcd4602c943c98543e2bc552502663">PyCUB.homoset.HomoSet.plot_simiclust()</a>.</p>

</div>
</div>
<a id="a8ab62bae7be17abf4b13ad37d5c881eb"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a8ab62bae7be17abf4b13ad37d5c881eb">&#9670;&nbsp;</a></span>plot_hashomo()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homoset.HomoSet.plot_hashomo </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>invert</em> = <code>False</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>size</em> = <code>40</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>interactive</em> = <code>False</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>rangeto</em> = <code>None</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>plot the has homo matrix </p>
<p>the interactive version allows you to see each particular datapoint with much more precision</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">interactive</td><td>bool to use the bokeh interactive version </td></tr>
    <tr><td class="paramname">size</td><td>int size of the matrix </td></tr>
    <tr><td class="paramname">invert</td><td>bool flag to true to invert the plot </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#afd51ea6276ecae41960d426ce72755b0">PyCUB.homoset.HomoSet._dictify()</a>.</p>

</div>
</div>
<a id="a87fcd4602c943c98543e2bc552502663"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a87fcd4602c943c98543e2bc552502663">&#9670;&nbsp;</a></span>plot_simiclust()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homoset.HomoSet.plot_simiclust </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>interactive</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>size</em> = <code>40</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>plot the similarity matrix of each homologies from its clusters </p>
<p>the interactive version allows you to see each particular datapoint with much more precision</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">interactive</td><td>bool to use the bokeh interactive version </td></tr>
    <tr><td class="paramname">size</td><td>size of the matrix </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#ae429c30197127c5d6248c94dbb6dfc46">PyCUB.homoset.HomoSet.compare_clusters()</a>, and <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a8ab62bae7be17abf4b13ad37d5c881eb">PyCUB.homoset.HomoSet.plot_hashomo()</a>.</p>

</div>
</div>
<a id="a3f689e03251e7b4fe262a86d08946998"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3f689e03251e7b4fe262a86d08946998">&#9670;&nbsp;</a></span>preprocessing()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homoset.HomoSet.preprocessing </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>withtaxons</em> = <code>False</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>withnames</em> = <code>True</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>will compute the full list of names, </p>
<p>find doublons, and set the names to ints instead of strings. called after loading from ensembl and associate namelist in each homologies to a number in utils.speciestable</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">withtaxons</td><td>bool to true calls preprocessing_taxons() else one of the other two </td></tr>
    <tr><td class="paramname">withnames</td><td>bool to true to call <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a36cf71a7bc24b788c4d526591a8ec376" title="same as preprocessing_taxons() but admiting there is no taxon information (Yun&#39;s data for example) ...">preprocessing_names()</a> else calls <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a4060c440127aa861c17b6157c81e2d27" title="same as preprocessing_names() but without preprocessing the names of each homologies only updating th...">preprocessing_namelist()</a></td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd></dd>
<dd>
taxons list, if the names contains an additional list of taxon ids. and the corresponding species in the same order only if withtaxons </dd></dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a05935a3f03f2e2b61b2f3e7e0222e2a9">PyCUB.homoset.HomoSet.add_random_homology()</a>.</p>

</div>
</div>
<a id="a4060c440127aa861c17b6157c81e2d27"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a4060c440127aa861c17b6157c81e2d27">&#9670;&nbsp;</a></span>preprocessing_namelist()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homoset.HomoSet.preprocessing_namelist </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>same as <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a36cf71a7bc24b788c4d526591a8ec376" title="same as preprocessing_taxons() but admiting there is no taxon information (Yun&#39;s data for example) ...">preprocessing_names()</a> but without preprocessing the names of each homologies only updating the species_namelist </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">None</td><td></td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a3f689e03251e7b4fe262a86d08946998">PyCUB.homoset.HomoSet.preprocessing()</a>, and <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a36cf71a7bc24b788c4d526591a8ec376">PyCUB.homoset.HomoSet.preprocessing_names()</a>.</p>

</div>
</div>
<a id="a36cf71a7bc24b788c4d526591a8ec376"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a36cf71a7bc24b788c4d526591a8ec376">&#9670;&nbsp;</a></span>preprocessing_names()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homoset.HomoSet.preprocessing_names </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>same as preprocessing_taxons() but admiting there is no taxon information (Yun's data for example) </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">None</td><td></td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a3f689e03251e7b4fe262a86d08946998">PyCUB.homoset.HomoSet.preprocessing()</a>.</p>

</div>
</div>
<a id="af19ffd1cceb4b222e642713456b300a1"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af19ffd1cceb4b222e642713456b300a1">&#9670;&nbsp;</a></span>remove()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homoset.HomoSet.remove </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>species</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>remove this list of species from the homologies </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">species</td><td>list[str] the species to remove from all homologies </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a2062c7f9c3c956d43172e1688504826a">PyCUB.homoset.HomoSet.clean_species()</a>, and <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a78ab39ddfa288661f5e809ea0a455187">PyCUB.homoset.HomoSet.compute_entropyloc()</a>.</p>

</div>
</div>
<a id="a000511647e3f85d8abadcdddaddfb04d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a000511647e3f85d8abadcdddaddfb04d">&#9670;&nbsp;</a></span>size()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homoset.HomoSet.size </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>the size of the homoset (number of genes) </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">None</td><td></td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>int the number of genes </dd></dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a5a56a99b5bf87afb418a1bbc5ec1c10b">PyCUB.homoset.HomoSet.loadhashomo()</a>.</p>

</div>
</div>
<a id="a8cf76800f047841bf63d507fb1b54048"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a8cf76800f047841bf63d507fb1b54048">&#9670;&nbsp;</a></span>update()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homoset.HomoSet.update </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>val</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>a function to update the dictionnary that homoset is </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">val</td><td>the dict to append tot this one </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#adbe4d9e4087fe5b53d861476c011329d">PyCUB.homoset.HomoSet.__init__()</a>, <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a42a7e7dc7ddb4b5717084cb093f90774">PyCUB.homoset.HomoSet.__setitem__()</a>, and <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a05935a3f03f2e2b61b2f3e7e0222e2a9">PyCUB.homoset.HomoSet.add_random_homology()</a>.</p>

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>PyCUB/homoset.py</li>
</ul>
</div><!-- contents -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="navelem"><b>PyCUB</b></li><li class="navelem"><b>homoset</b></li><li class="navelem"><a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html">HomoSet</a></li>
    <li class="footer">Generated by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.8.14 </li>
  </ul>
</div>
</body>
</html>
