<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.14"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>PyCUB: Machine exploration of the Codon Usage Bias: PyCUB.homology.homology Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
  $(document).ready(initResizable);
/* @license-end */</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    extensions: ["tex2jax.js"],
    jax: ["input/TeX","output/HTML-CSS"],
});
</script><script type="text/javascript" async src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.2/MathJax.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">PyCUB: Machine exploration of the Codon Usage Bias
   </div>
   <div id="projectbrief">A set of tools to explore the codon usage bias</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.14 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(document).ready(function(){initNavTree('class_py_c_u_b_1_1homology_1_1homology.html','');});
/* @license-end */
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-attribs">Public Attributes</a> &#124;
<a href="#pub-static-attribs">Static Public Attributes</a> &#124;
<a href="#pri-methods">Private Member Functions</a> &#124;
<a href="class_py_c_u_b_1_1homology_1_1homology-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">PyCUB.homology.homology Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p>in homology we store an homology with all its related data,  
 <a href="class_py_c_u_b_1_1homology_1_1homology.html#details">More...</a></p>
<div class="dynheader">
Inheritance diagram for PyCUB.homology.homology:</div>
<div class="dyncontent">
 <div class="center">
  <img src="class_py_c_u_b_1_1homology_1_1homology.png" usemap="#PyCUB.homology.homology_map" alt=""/>
  <map id="PyCUB.homology.homology_map" name="PyCUB.homology.homology_map">
</map>
 </div></div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a441a70cdc5814708841f5d3c29548203"><td class="memItemLeft" align="right" valign="top"><a id="a441a70cdc5814708841f5d3c29548203"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>__init__</b> (self, kwargs)</td></tr>
<tr class="separator:a441a70cdc5814708841f5d3c29548203"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af27d98c4ced0023053744c7ba038b33a"><td class="memItemLeft" align="right" valign="top"><a id="af27d98c4ced0023053744c7ba038b33a"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>__str__</b> (self)</td></tr>
<tr class="separator:af27d98c4ced0023053744c7ba038b33a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a32017c483bad9a5c4498640279a79634"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homology_1_1homology.html#a32017c483bad9a5c4498640279a79634">remove</a> (self, species)</td></tr>
<tr class="memdesc:a32017c483bad9a5c4498640279a79634"><td class="mdescLeft">&#160;</td><td class="mdescRight">removes the list of species from this homology if it exists there  <a href="#a32017c483bad9a5c4498640279a79634">More...</a><br /></td></tr>
<tr class="separator:a32017c483bad9a5c4498640279a79634"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad430899a462da76cdd6e1fc6ee73e0ec"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homology_1_1homology.html#ad430899a462da76cdd6e1fc6ee73e0ec">nb_unique_species</a> (self)</td></tr>
<tr class="memdesc:ad430899a462da76cdd6e1fc6ee73e0ec"><td class="mdescLeft">&#160;</td><td class="mdescRight">compute the number of unique species in this homologies  <a href="#ad430899a462da76cdd6e1fc6ee73e0ec">More...</a><br /></td></tr>
<tr class="separator:ad430899a462da76cdd6e1fc6ee73e0ec"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a513d82c775ba47678304c97572eaf932"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homology_1_1homology.html#a513d82c775ba47678304c97572eaf932">order</a> (self, withtaxons=False)</td></tr>
<tr class="memdesc:a513d82c775ba47678304c97572eaf932"><td class="mdescLeft">&#160;</td><td class="mdescRight">order the names by numerical increasing order  <a href="#a513d82c775ba47678304c97572eaf932">More...</a><br /></td></tr>
<tr class="separator:a513d82c775ba47678304c97572eaf932"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aaeb164cd217a4746ff0082da1a92fc5e"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homology_1_1homology.html#aaeb164cd217a4746ff0082da1a92fc5e">compute_averages</a> (self)</td></tr>
<tr class="memdesc:aaeb164cd217a4746ff0082da1a92fc5e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Computes the mean, var and mean of the homology.  <a href="#aaeb164cd217a4746ff0082da1a92fc5e">More...</a><br /></td></tr>
<tr class="separator:aaeb164cd217a4746ff0082da1a92fc5e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3db21ba20b7720362cc19cc5b204eeab"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homology_1_1homology.html#a3db21ba20b7720362cc19cc5b204eeab">reduce_dim</a> (self, alg='tsne', n=2, perplexity=40)</td></tr>
<tr class="memdesc:a3db21ba20b7720362cc19cc5b204eeab"><td class="mdescLeft">&#160;</td><td class="mdescRight">reduce the dimensionality of your gene dataset to a defined dimension using the t-SNE algorithm  <a href="#a3db21ba20b7720362cc19cc5b204eeab">More...</a><br /></td></tr>
<tr class="separator:a3db21ba20b7720362cc19cc5b204eeab"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1a3802095c2ea6f426e803b055a2aea9"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homology_1_1homology.html#a1a3802095c2ea6f426e803b055a2aea9">plot</a> (self, reducer=&quot;tsne&quot;, per=40, interactive=True, D=2, size=20)</td></tr>
<tr class="memdesc:a1a3802095c2ea6f426e803b055a2aea9"><td class="mdescLeft">&#160;</td><td class="mdescRight">plot the dimensionality reduced datapoint of the homology matrix  <a href="#a1a3802095c2ea6f426e803b055a2aea9">More...</a><br /></td></tr>
<tr class="separator:a1a3802095c2ea6f426e803b055a2aea9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7e7748ba85bf571bb04df7a26e2c682c"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homology_1_1homology.html#a7e7748ba85bf571bb04df7a26e2c682c">clusterize_</a> (self, clustering='gaussian', eps=0.8, homogroupnb=None, assess=True, verbose=True)</td></tr>
<tr class="memdesc:a7e7748ba85bf571bb04df7a26e2c682c"><td class="mdescLeft">&#160;</td><td class="mdescRight">will clusterize the homology using gaussian mixture clustering or DBSCAN  <a href="#a7e7748ba85bf571bb04df7a26e2c682c">More...</a><br /></td></tr>
<tr class="separator:a7e7748ba85bf571bb04df7a26e2c682c"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-attribs"></a>
Public Attributes</h2></td></tr>
<tr class="memitem:ad0fa49f698ad9e7ae29f624d45462c4d"><td class="memItemLeft" align="right" valign="top"><a id="ad0fa49f698ad9e7ae29f624d45462c4d"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>metrics</b></td></tr>
<tr class="separator:ad0fa49f698ad9e7ae29f624d45462c4d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1eb731bc0f1f864f9bdc1b00e2038dc6"><td class="memItemLeft" align="right" valign="top"><a id="a1eb731bc0f1f864f9bdc1b00e2038dc6"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>homocode</b></td></tr>
<tr class="separator:a1eb731bc0f1f864f9bdc1b00e2038dc6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a74a80cf95e01b8ac1f8547507481ab46"><td class="memItemLeft" align="right" valign="top"><a id="a74a80cf95e01b8ac1f8547507481ab46"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>proteinids</b></td></tr>
<tr class="separator:a74a80cf95e01b8ac1f8547507481ab46"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0438db8bca39f5964292ea3e5f3ae519"><td class="memItemLeft" align="right" valign="top"><a id="a0438db8bca39f5964292ea3e5f3ae519"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>protein_abundance</b></td></tr>
<tr class="separator:a0438db8bca39f5964292ea3e5f3ae519"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae230e5a5e755ffeb7166ceef23cfaa90"><td class="memItemLeft" align="right" valign="top"><a id="ae230e5a5e755ffeb7166ceef23cfaa90"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>weight</b></td></tr>
<tr class="separator:ae230e5a5e755ffeb7166ceef23cfaa90"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0290845a014043ef7cef5994da0ea14f"><td class="memItemLeft" align="right" valign="top"><a id="a0290845a014043ef7cef5994da0ea14f"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>mRNA_abundance</b></td></tr>
<tr class="separator:a0290845a014043ef7cef5994da0ea14f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a41ea8ca7536b4e7db86d9481f14c4c89"><td class="memItemLeft" align="right" valign="top"><a id="a41ea8ca7536b4e7db86d9481f14c4c89"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>cys_elements</b></td></tr>
<tr class="separator:a41ea8ca7536b4e7db86d9481f14c4c89"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6318ac761caa41be6b784340bc3e9d7c"><td class="memItemLeft" align="right" valign="top"><a id="a6318ac761caa41be6b784340bc3e9d7c"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>is_secreted</b></td></tr>
<tr class="separator:a6318ac761caa41be6b784340bc3e9d7c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6bfbe5b4868039efca584fb13e3e17b4"><td class="memItemLeft" align="right" valign="top"><a id="a6bfbe5b4868039efca584fb13e3e17b4"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>decay_rate</b></td></tr>
<tr class="separator:a6bfbe5b4868039efca584fb13e3e17b4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab5b1b81d50e5bca0eac0f6bee539c4d7"><td class="memItemLeft" align="right" valign="top"><a id="ab5b1b81d50e5bca0eac0f6bee539c4d7"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>cluster</b></td></tr>
<tr class="separator:ab5b1b81d50e5bca0eac0f6bee539c4d7"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-static-attribs"></a>
Static Public Attributes</h2></td></tr>
<tr class="memitem:a5220c03b38deb65d0e682bd1ad85802b"><td class="memItemLeft" align="right" valign="top"><a id="a5220c03b38deb65d0e682bd1ad85802b"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>full</b> = None</td></tr>
<tr class="separator:a5220c03b38deb65d0e682bd1ad85802b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af6fbc0a12aaec1fd2c4e2231d8de5520"><td class="memItemLeft" align="right" valign="top"><a id="af6fbc0a12aaec1fd2c4e2231d8de5520"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>var</b> = None</td></tr>
<tr class="separator:af6fbc0a12aaec1fd2c4e2231d8de5520"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aabb8f64472865c658c9089b6d5ff51fd"><td class="memItemLeft" align="right" valign="top"><a id="aabb8f64472865c658c9089b6d5ff51fd"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>mean</b> = None</td></tr>
<tr class="separator:aabb8f64472865c658c9089b6d5ff51fd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:acf7c487a7f77ef947afc193d36becd61"><td class="memItemLeft" align="right" valign="top"><a id="acf7c487a7f77ef947afc193d36becd61"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>clusters</b> = None</td></tr>
<tr class="separator:acf7c487a7f77ef947afc193d36becd61"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7553f27ba7af2b61b973688bf1c72a7b"><td class="memItemLeft" align="right" valign="top"><a id="a7553f27ba7af2b61b973688bf1c72a7b"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>centroids</b> = None</td></tr>
<tr class="separator:a7553f27ba7af2b61b973688bf1c72a7b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aae3f58df1a8aa3032b3bb648b5f9788a"><td class="memItemLeft" align="right" valign="top"><a id="aae3f58df1a8aa3032b3bb648b5f9788a"></a>
dictionary&#160;</td><td class="memItemRight" valign="bottom"><b>metrics</b> = {}</td></tr>
<tr class="separator:aae3f58df1a8aa3032b3bb648b5f9788a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af59d023da4ec847ead6845216b1c02f0"><td class="memItemLeft" align="right" valign="top"><a id="af59d023da4ec847ead6845216b1c02f0"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>homocode</b> = 'None'</td></tr>
<tr class="separator:af59d023da4ec847ead6845216b1c02f0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:acbb6a5d373b79e9a783eb9efaaaecb97"><td class="memItemLeft" align="right" valign="top"><a id="acbb6a5d373b79e9a783eb9efaaaecb97"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>nans</b> = None</td></tr>
<tr class="separator:acbb6a5d373b79e9a783eb9efaaaecb97"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a56a1f0e4e56e373d69c3d39712937278"><td class="memItemLeft" align="right" valign="top"><a id="a56a1f0e4e56e373d69c3d39712937278"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>lenmat</b> = None</td></tr>
<tr class="separator:a56a1f0e4e56e373d69c3d39712937278"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae8979e24b36f8ca79963aa1f03db1718"><td class="memItemLeft" align="right" valign="top"><a id="ae8979e24b36f8ca79963aa1f03db1718"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>names</b> = None</td></tr>
<tr class="separator:ae8979e24b36f8ca79963aa1f03db1718"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adb1c7c9de932faa407921b746fea951d"><td class="memItemLeft" align="right" valign="top"><a id="adb1c7c9de932faa407921b746fea951d"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>doub</b> = None</td></tr>
<tr class="separator:adb1c7c9de932faa407921b746fea951d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a862b28ceb170090b9c9959d0ec234ecf"><td class="memItemLeft" align="right" valign="top"><a id="a862b28ceb170090b9c9959d0ec234ecf"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>GCcount</b> = None</td></tr>
<tr class="separator:a862b28ceb170090b9c9959d0ec234ecf"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af5cc9c64f61bf6154788e601e68ac56f"><td class="memItemLeft" align="right" valign="top"><a id="af5cc9c64f61bf6154788e601e68ac56f"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>reduced</b> = None</td></tr>
<tr class="separator:af5cc9c64f61bf6154788e601e68ac56f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1332a6816c31c207bcf9b950b9befdd2"><td class="memItemLeft" align="right" valign="top"><a id="a1332a6816c31c207bcf9b950b9befdd2"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>reduced_algo</b> = None</td></tr>
<tr class="separator:a1332a6816c31c207bcf9b950b9befdd2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3d898f87bb402fff974e77a3813c91b9"><td class="memItemLeft" align="right" valign="top"><a id="a3d898f87bb402fff974e77a3813c91b9"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>KaKs_Scores</b> = None</td></tr>
<tr class="separator:a3d898f87bb402fff974e77a3813c91b9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a456033ba31e423f2844163b02441f253"><td class="memItemLeft" align="right" valign="top"><a id="a456033ba31e423f2844163b02441f253"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>similarity_scores</b> = None</td></tr>
<tr class="separator:a456033ba31e423f2844163b02441f253"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8d77dcd1deecff3b3114e89d50bd08d7"><td class="memItemLeft" align="right" valign="top"><a id="a8d77dcd1deecff3b3114e89d50bd08d7"></a>
list&#160;</td><td class="memItemRight" valign="bottom"><b>proteinids</b> = []</td></tr>
<tr class="separator:a8d77dcd1deecff3b3114e89d50bd08d7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a25b698233f4fb598400fb725c00e2c99"><td class="memItemLeft" align="right" valign="top"><a id="a25b698233f4fb598400fb725c00e2c99"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>isrecent</b> = None</td></tr>
<tr class="separator:a25b698233f4fb598400fb725c00e2c99"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af0f123770c60d2444ffb139acaaf5442"><td class="memItemLeft" align="right" valign="top"><a id="af0f123770c60d2444ffb139acaaf5442"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>ishighpreserved</b> = None</td></tr>
<tr class="separator:af0f123770c60d2444ffb139acaaf5442"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:add9343a07f66053ea6ee9dd5ebc275dc"><td class="memItemLeft" align="right" valign="top"><a id="add9343a07f66053ea6ee9dd5ebc275dc"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>geneids</b> = None</td></tr>
<tr class="separator:add9343a07f66053ea6ee9dd5ebc275dc"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa710cdd07ec15477bb995f17f14f0ac5"><td class="memItemLeft" align="right" valign="top"><a id="aa710cdd07ec15477bb995f17f14f0ac5"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>ref</b> = None</td></tr>
<tr class="separator:aa710cdd07ec15477bb995f17f14f0ac5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3aeac1174f5e74f3291d35bfbeb0ae56"><td class="memItemLeft" align="right" valign="top"><a id="a3aeac1174f5e74f3291d35bfbeb0ae56"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>refprot</b> = None</td></tr>
<tr class="separator:a3aeac1174f5e74f3291d35bfbeb0ae56"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a549fd390f70453a9676ee4d26c84fe04"><td class="memItemLeft" align="right" valign="top"><a id="a549fd390f70453a9676ee4d26c84fe04"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>refgene</b> = None</td></tr>
<tr class="separator:a549fd390f70453a9676ee4d26c84fe04"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7f37b98057afffafaeae13285490b229"><td class="memItemLeft" align="right" valign="top"><a id="a7f37b98057afffafaeae13285490b229"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>ecai</b> = None</td></tr>
<tr class="separator:a7f37b98057afffafaeae13285490b229"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a42eb34bc6c1e774b237e479f83cd6bf7"><td class="memItemLeft" align="right" valign="top"><a id="a42eb34bc6c1e774b237e479f83cd6bf7"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>meanecai</b> = None</td></tr>
<tr class="separator:a42eb34bc6c1e774b237e479f83cd6bf7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9dea84f934449466e561988ea1d2031d"><td class="memItemLeft" align="right" valign="top"><a id="a9dea84f934449466e561988ea1d2031d"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>cai</b> = None</td></tr>
<tr class="separator:a9dea84f934449466e561988ea1d2031d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a97977b754401226826b71af612e6d33e"><td class="memItemLeft" align="right" valign="top"><a id="a97977b754401226826b71af612e6d33e"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>meancai</b> = None</td></tr>
<tr class="separator:a97977b754401226826b71af612e6d33e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5a3df5b52429c04c5bed49bd5e4a3b20"><td class="memItemLeft" align="right" valign="top"><a id="a5a3df5b52429c04c5bed49bd5e4a3b20"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>protein_abundance</b> = 0.</td></tr>
<tr class="separator:a5a3df5b52429c04c5bed49bd5e4a3b20"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aeb597a006a4b74e0c4332079191fddc9"><td class="memItemLeft" align="right" valign="top"><a id="aeb597a006a4b74e0c4332079191fddc9"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>weight</b> = 0</td></tr>
<tr class="separator:aeb597a006a4b74e0c4332079191fddc9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa51ccde91bb3ca90cb11ae263627665a"><td class="memItemLeft" align="right" valign="top"><a id="aa51ccde91bb3ca90cb11ae263627665a"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>conservation</b> = None</td></tr>
<tr class="separator:aa51ccde91bb3ca90cb11ae263627665a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a62e7776334a677b8634faede236529de"><td class="memItemLeft" align="right" valign="top"><a id="a62e7776334a677b8634faede236529de"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>mRNA_abundance</b> = 0.</td></tr>
<tr class="separator:a62e7776334a677b8634faede236529de"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5492c796d1e811df903eb6ac5e6c5873"><td class="memItemLeft" align="right" valign="top"><a id="a5492c796d1e811df903eb6ac5e6c5873"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>cys_elements</b> = 0</td></tr>
<tr class="separator:a5492c796d1e811df903eb6ac5e6c5873"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0db815d3136df0d8057d6aa207e3a9d7"><td class="memItemLeft" align="right" valign="top"><a id="a0db815d3136df0d8057d6aa207e3a9d7"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><b>is_secreted</b> = False</td></tr>
<tr class="separator:a0db815d3136df0d8057d6aa207e3a9d7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a86cb7804faa6ffae20b7155897d88971"><td class="memItemLeft" align="right" valign="top"><a id="a86cb7804faa6ffae20b7155897d88971"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>decay_rate</b> = 0.</td></tr>
<tr class="separator:a86cb7804faa6ffae20b7155897d88971"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7b73b4cd404f7432302ce682970c04c1"><td class="memItemLeft" align="right" valign="top"><a id="a7b73b4cd404f7432302ce682970c04c1"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>tot_volume</b> = None</td></tr>
<tr class="separator:a7b73b4cd404f7432302ce682970c04c1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2fcdeaac4c8dcda15c2faaa26d565c2b"><td class="memItemLeft" align="right" valign="top"><a id="a2fcdeaac4c8dcda15c2faaa26d565c2b"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>mean_hydrophobicity</b> = None</td></tr>
<tr class="separator:a2fcdeaac4c8dcda15c2faaa26d565c2b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3935ee71c6588e71ed71ca843dabb275"><td class="memItemLeft" align="right" valign="top"><a id="a3935ee71c6588e71ed71ca843dabb275"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>glucose_cost</b> = None</td></tr>
<tr class="separator:a3935ee71c6588e71ed71ca843dabb275"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af6987d4b9223fb55cfe60e0456478b71"><td class="memItemLeft" align="right" valign="top"><a id="af6987d4b9223fb55cfe60e0456478b71"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>synthesis_steps</b> = None</td></tr>
<tr class="separator:af6987d4b9223fb55cfe60e0456478b71"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab3cc7d820ffac1600aea6dbc58995364"><td class="memItemLeft" align="right" valign="top"><a id="ab3cc7d820ffac1600aea6dbc58995364"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>isoelectricpoint</b> = None</td></tr>
<tr class="separator:ab3cc7d820ffac1600aea6dbc58995364"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0fee75b305ed6adb93ed111cce23218a"><td class="memItemLeft" align="right" valign="top"><a id="a0fee75b305ed6adb93ed111cce23218a"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>othercods</b> = None</td></tr>
<tr class="separator:a0fee75b305ed6adb93ed111cce23218a"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pri-methods"></a>
Private Member Functions</h2></td></tr>
<tr class="memitem:af7f186d30cb3469d73782da85f9d06bc"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_py_c_u_b_1_1homology_1_1homology.html#af7f186d30cb3469d73782da85f9d06bc">_dictify</a> (self)</td></tr>
<tr class="memdesc:af7f186d30cb3469d73782da85f9d06bc"><td class="mdescLeft">&#160;</td><td class="mdescRight">Used by the saving function.  <a href="#af7f186d30cb3469d73782da85f9d06bc">More...</a><br /></td></tr>
<tr class="separator:af7f186d30cb3469d73782da85f9d06bc"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>in homology we store an homology with all its related data, </p>
<p>it reduced matrix with dim reduction and its clusters for example it is supposed to be store in a dictionary of homologies an homology is a set of genes from different species related by a common ancester gene and generally a common function. the unique metadatas are generally from the reference species/genome</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">names</td><td>list of int corresponding to names in utils.speciestable </td></tr>
    <tr><td class="paramname">full</td><td>np.array[float] (species, amino) of one homology with entropy value vector per species </td></tr>
    <tr><td class="paramname">reduced</td><td>np.array[float] (species, x*y) of one homology dimensionality reduced </td></tr>
    <tr><td class="paramname">clusters</td><td>list[int] of cluster val for each species </td></tr>
    <tr><td class="paramname">centroids</td><td>np.array[float] the position in aminoacid# Dimension of each centroids (number of centroid == number of cluster), dimension is reduced when plotted </td></tr>
    <tr><td class="paramname">metrics</td><td>a dict of metrics names and values for the cluster of this homology </td></tr>
    <tr><td class="paramname">nans</td><td>np.array[bool] wether or not this position is a nan </td></tr>
    <tr><td class="paramname">lenmat</td><td>a np.array[int] species amino of int of length of each amino acids (number) </td></tr>
    <tr><td class="paramname">doub</td><td>np.array[bool] of wether or not this position is a doublon ( a copy number of the gene for the species) </td></tr>
    <tr><td class="paramname">reduced_algo</td><td>str dimensionality reduction algorithm used on this homology </td></tr>
    <tr><td class="paramname">var</td><td>np.array of the variance in the CUB value for each datapoint </td></tr>
    <tr><td class="paramname">mean</td><td>np.array[float] of the mean CUB for each datapoint </td></tr>
    <tr><td class="paramname">homocode</td><td>str the code of the homology </td></tr>
    <tr><td class="paramname">GCcount</td><td>np.array[float] GC bias for each datapoint </td></tr>
    <tr><td class="paramname">KaKs_Scores</td><td>np.array[float] a form of similarity score between genes/datapoints </td></tr>
    <tr><td class="paramname">similarity_scores</td><td>np.array[float] similarity score between genes/datapoints </td></tr>
    <tr><td class="paramname">proteinids</td><td>list[str] the protein names for each datapoints </td></tr>
    <tr><td class="paramname">isrecent</td><td>float a proxy for wether or not this homology has appeated recently </td></tr>
    <tr><td class="paramname">ishighpreserved</td><td>bool a proxy for wether or not this homology is conserved throughout evolution </td></tr>
    <tr><td class="paramname">geneids</td><td>list[str] the name of the genes </td></tr>
    <tr><td class="paramname">ref</td><td>np.array[float] the reference CUB value of cerevisiae gene </td></tr>
    <tr><td class="paramname">refprot</td><td>str the reference protein name of cerevisiae gene </td></tr>
    <tr><td class="paramname">refgene</td><td>str the reference gene name of cerevisiae gene </td></tr>
    <tr><td class="paramname">ecai</td><td>np.array[float] the codon adaptation index of each gene </td></tr>
    <tr><td class="paramname">meanecai</td><td>float the mean ecai of the ecai </td></tr>
    <tr><td class="paramname">protein_abundance</td><td>float the average abundance of the protein enoded by this gene in cerevisiae cells </td></tr>
    <tr><td class="paramname">weight</td><td>int the molecular weight of this protein </td></tr>
    <tr><td class="paramname">mRNA_abundance</td><td>float the average abundance of the messenger RNA of this coding gene in cerevisiae cells </td></tr>
    <tr><td class="paramname">cys_elements</td><td>int the number of cys regulatory elements known for cerevisiae on this gene (the value is only referenced for secreted genes) </td></tr>
    <tr><td class="paramname">is_secreted</td><td>bool wether or not this protein is secreted out of the cell </td></tr>
    <tr><td class="paramname">decay_rate</td><td>float the half life in minute of this protein </td></tr>
    <tr><td class="paramname">tot_volume</td><td>int a proxy of the molecular volume of this protein </td></tr>
    <tr><td class="paramname">mean_hydrophobicity</td><td>a very distant proxy of the hydrophobicity of this protein </td></tr>
    <tr><td class="paramname">glucose_cost</td><td>the glucose cost of creating the amino acids required to build up this protein </td></tr>
    <tr><td class="paramname">synthesis_steps</td><td>the number of steps required by the cell to build up the amino acids of this protein </td></tr>
    <tr><td class="paramname">isoelectricpoint</td><td>float, a proxy of the Pi of this protein. </td></tr>
    <tr><td class="paramname">conservation</td><td>float, the total conservation of each amino of the corresponding protein </td></tr>
    <tr><td class="paramname">othercods</td><td>float, the average number of codons other than the ones of the 18 amino acids we are looking at per species on the homology </td></tr>
  </table>
  </dd>
</dl>
</div><h2 class="groupheader">Member Function Documentation</h2>
<a id="af7f186d30cb3469d73782da85f9d06bc"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af7f186d30cb3469d73782da85f9d06bc">&#9670;&nbsp;</a></span>_dictify()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homology.homology._dictify </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Used by the saving function. </p>
<p>transform the object into a dictionary that can be json serializable</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">None</td><td></td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A dict holding every element to be jsonized </dd></dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1homology_1_1homology.html#a7e7748ba85bf571bb04df7a26e2c682c">PyCUB.homology.homology.clusterize_()</a>, and <a class="el" href="class_py_c_u_b_1_1py_c_u_b_1_1_py_c_u_b.html#a16a1ab6be5dddcedf94852088936c1f2">PyCUB.pyCUB.PyCUB.loadmore()</a>.</p>

</div>
</div>
<a id="a7e7748ba85bf571bb04df7a26e2c682c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a7e7748ba85bf571bb04df7a26e2c682c">&#9670;&nbsp;</a></span>clusterize_()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homology.homology.clusterize_ </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>clustering</em> = <code>'gaussian'</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>eps</em> = <code>0.8</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>homogroupnb</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>assess</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>verbose</em> = <code>True</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>will clusterize the homology using gaussian mixture clustering or DBSCAN </p>
<p>and order them according to the density of each cluster (we are interested in the dense ones) and assess the quality using 3 criterion: BIC, AIC ,silhouette, cal_hara, phylodistance.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">clustering</td><td>str flag the clustering algorithm [gaussian, dbscan] </td></tr>
    <tr><td class="paramname">eps</td><td>float, hyperparam of the max size of the nsphere of each cluster </td></tr>
    <tr><td class="paramname">homogroupnb</td><td>int hyperparam of gaussian for the number of clusters </td></tr>
    <tr><td class="paramname">assess</td><td>wether or not to assess the quality of the clustering </td></tr>
    <tr><td class="paramname">verbose</td><td>wether or not to show clustering quality information</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The clusters for each datapoint of the homology as a list[int]</dd></dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">AttributeError</td><td>"Hey, please use gaussian or dbscan" </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1homology_1_1homology.html#a1a3802095c2ea6f426e803b055a2aea9">PyCUB.homology.homology.plot()</a>.</p>

</div>
</div>
<a id="aaeb164cd217a4746ff0082da1a92fc5e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aaeb164cd217a4746ff0082da1a92fc5e">&#9670;&nbsp;</a></span>compute_averages()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homology.homology.compute_averages </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Computes the mean, var and mean of the homology. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">None</td><td></td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1homology_1_1homology.html#a513d82c775ba47678304c97572eaf932">PyCUB.homology.homology.order()</a>.</p>

</div>
</div>
<a id="ad430899a462da76cdd6e1fc6ee73e0ec"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad430899a462da76cdd6e1fc6ee73e0ec">&#9670;&nbsp;</a></span>nb_unique_species()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homology.homology.nb_unique_species </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>compute the number of unique species in this homologies </p>
<p>(basically count the number of doub)</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">None</td><td></td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The number of unique species in this homology </dd></dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1homology_1_1homology.html#a32017c483bad9a5c4498640279a79634">PyCUB.homology.homology.remove()</a>.</p>

</div>
</div>
<a id="a513d82c775ba47678304c97572eaf932"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a513d82c775ba47678304c97572eaf932">&#9670;&nbsp;</a></span>order()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homology.homology.order </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>withtaxons</em> = <code>False</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>order the names by numerical increasing order </p>
<p>and sorts every representations as well according to this ordering</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">withtaxons</td><td>bool to true if there is taxonomic data (present before preprocessing) </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1homology_1_1homology.html#ad430899a462da76cdd6e1fc6ee73e0ec">PyCUB.homology.homology.nb_unique_species()</a>.</p>

</div>
</div>
<a id="a1a3802095c2ea6f426e803b055a2aea9"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a1a3802095c2ea6f426e803b055a2aea9">&#9670;&nbsp;</a></span>plot()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homology.homology.plot </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>reducer</em> = <code>&quot;tsne&quot;</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>per</em> = <code>40</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>interactive</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>D</em> = <code>2</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>size</em> = <code>20</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>plot the dimensionality reduced datapoint of the homology matrix </p>
<p>colors represents the different clusters you can set the interactivity to gain deeper knowledge of the dataset. It can dim reduce the data and will also save the figure in utils/save. Moreover, it will print some interesting data about the homology</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">reducer</td><td>str flag the algorithm to reduce the matrix of utils.cubD to D D </td></tr>
    <tr><td class="paramname">per</td><td>int he perplexity when using tsne algorithm </td></tr>
    <tr><td class="paramname">size</td><td>int the size of the plot </td></tr>
    <tr><td class="paramname">interactive</td><td>(recommended) wether to use bokeh or not to represend the data will show name of the species when hovering over datapoint will show a gradient of evolutionary distance of the species of the datapoint currently hovered to each other species </td></tr>
    <tr><td class="paramname">D</td><td>int the goal dimension (2-3-4) when using matplotlib</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The plot interactive or not and some informations (as prints)</dd></dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">UnboundLocalError</td><td>"you need to have dimensionality reduced to 2D to have a right plotting of the centroids" </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1homology_1_1homology.html#a3db21ba20b7720362cc19cc5b204eeab">PyCUB.homology.homology.reduce_dim()</a>.</p>

</div>
</div>
<a id="a3db21ba20b7720362cc19cc5b204eeab"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3db21ba20b7720362cc19cc5b204eeab">&#9670;&nbsp;</a></span>reduce_dim()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homology.homology.reduce_dim </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>alg</em> = <code>'tsne'</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>n</em> = <code>2</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>perplexity</em> = <code>40</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>reduce the dimensionality of your gene dataset to a defined dimension using the t-SNE algorithm </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">alg</td><td>a matrix of gene codon usage per species </td></tr>
    <tr><td class="paramname">n</td><td>the desired dimension </td></tr>
    <tr><td class="paramname">perplexity</td><td>an optional value when you know about tsne</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The reduced dataset </dd></dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1homology_1_1homology.html#aaeb164cd217a4746ff0082da1a92fc5e">PyCUB.homology.homology.compute_averages()</a>, and <a class="el" href="class_py_c_u_b_1_1homology_1_1homology.html#a1a3802095c2ea6f426e803b055a2aea9">PyCUB.homology.homology.plot()</a>.</p>

</div>
</div>
<a id="a32017c483bad9a5c4498640279a79634"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a32017c483bad9a5c4498640279a79634">&#9670;&nbsp;</a></span>remove()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyCUB.homology.homology.remove </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>species</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>removes the list of species from this homology if it exists there </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">species</td><td>list[str] of species to remove </td></tr>
  </table>
  </dd>
</dl>

<p class="reference">Referenced by <a class="el" href="class_py_c_u_b_1_1homoset_1_1_homo_set.html#a78ab39ddfa288661f5e809ea0a455187">PyCUB.homoset.HomoSet.compute_entropyloc()</a>.</p>

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>PyCUB/homology.py</li>
</ul>
</div><!-- contents -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="navelem"><a class="el" href="namespace_py_c_u_b.html">PyCUB</a></li><li class="navelem"><b>homology</b></li><li class="navelem"><a class="el" href="class_py_c_u_b_1_1homology_1_1homology.html">homology</a></li>
    <li class="footer">Generated by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.8.14 </li>
  </ul>
</div>
</body>
</html>
